---
title: "Intergenic transcription pipeline"
author: |
  | Federico Agostini <federico.agostini@crick.ac.uk>
date: "5 February 2018"
output:
  html_notebook:
    code_folding: hide
    collapsed: yes
    fig_caption: yes
editor_options: 
  chunk_output_type: console
---

The first part of the analysis, including pre-processing, alignment of the sample to the human genome and generation of the individual transcriptome assembly has been performed with the [RNA-seq-pipeline](https://github.com/luslab/RNA-seq-pipeline).

```{bash}
for j in `ls "./stringtie/" | grep "StringTie.gtf"`; do
   if [[ ! -s ${j%%.gtf}".sorted.gtf" ]]; then
      igvtools sort "./stringtie/"$j "./stringtie/"${j%%.gtf}".sorted.gtf"
      igvtools index "./stringtie/"${j%%.gtf}".sorted.gtf"
   fi
done

stringtie --merge -G "./gencode.v27.annotation.sorted.gtf" -o "./stringtie.merged.gtf" `ls "./stringtie/"*".sorted.gtf" | grep "StringTie.sorted.gtf"$ | grep -e "Untreated" -e "WildType" -e "siLuc" -e "Bcells" -e "Unmodified" | grep -v "UV"`

igvtools sort "./stringtie.merged.gtf" "./stringtie.merged.sorted.gtf"

igvtools index "./stringtie.merged.sorted.gtf"

gffcompare -C -T -r "./gencode.v27.annotation.sorted.gtf" -o "./gffcompare" "./stringtie.merged.sorted.gtf"

igvtools sort "./gffcompare.combined.gtf" "./gffcompare.combined.sorted.gtf"

igvtools index "./gffcompare.combined.sorted.gtf"
```

******

## R session

```{r knitr, include=FALSE}
require("knitr")
```

```{r setup, echo=FALSE}
working.folder    = getwd()
annotation.folder = file.path(working.folder, "GencodeReference")

if( !dir.exists("RData") )
   dir.create("RData")
if( !dir.exists("Figures_and_Tables") )
   dir.create("Figures_and_Tables")
if( !dir.exists("RDS") )
   dir.create("RDS")
if( !dir.exists("annotatedRegions") )
   dir.create("annotatedRegions")

setwd(working.folder)
opts_chunk$set(root.dir = working.folder, dev = 'png', echo = TRUE)

cat(sessionInfo()$R.version$version.string, fill=TRUE)
cat(paste("Platform", sessionInfo()$platform), fill=TRUE)
cat(paste("Running under", sessionInfo()$running), fill=TRUE)
cat(paste("Last knitted on", date()), fill=TRUE)
cat(paste0("Working directory set to ", working.folder), fill=TRUE)
cat(paste0("Annotation directory set to ", annotation.folder), fill=TRUE)
```

******

## Set global parameters

```{r setParameters, cache=FALSE, echo=FALSE}
# Whether or not to regenerate all the objects
REBUILD = FALSE

# Define the species
species = "Homo_sapiens"

# Define the distance to consider two regions and separated
max.gapwidth = 1e3 # was 2e2

# Define the distance to consider a region distal rather than proximal
min.gapwidth = 1e4

# Define the minimum width for a novel region
min.frag.width = 1e3
min.feature.width = 1e3

# Define the minimum TPM expression threshold
min.expr = 1 

# Define the quantile threshold for i) Removing lowly expressed fragments (frags) and ii) Adding unassigned regions (gaps) 
qnt_th = 0.1

# Define the factor levels for the subset
factSubset = c("GencodeGene", "UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "< 10Kb", "> 10Kb")

# Define the biotype levels
factSubset2 = c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "< 10Kb", "> 10Kb",
                "Protein_coding (expressed)", "LincRNA", "Protein_coding (not expressed)")

# Define the biotype levels
factBiotype = c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "< 10Kb", "> 10Kb",
                "Protein_coding", "Long_non-coding", "Non-coding", "Other annotated")

# Define the biotype levels
factBiotype2 = c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "< 10Kb", "> 10Kb",
                 "protein_coding", "long_ncRNA", "ncRNA", "other annotated")

# Define the colour set
colSet = setNames(c("#4E79A7", "#F28E2B", "#E15759", "#76B7B2", "#59A14F", "#EDC948"),
                  c("Protein coding", "Long ncRNA", "Small ncRNA", "Other annotated", "Gene-associated TU", "Independent TU"))

# Define the colour set
colSet2 = setNames(c("#4E79A7", "#F28E2B", "#E15759", "#76B7B2", "#59A14F", "#59A14F", "#59A14F", "#EDC948"),
                  c("Protein coding", "Long ncRNA", "Small ncRNA", "Other annotated",
                    "UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "Independent TU"))
```

- Maximum gap width (max.gapwidth) for same transcriptional unit fragments: `r paste0(max.gapwidth, "nt")`
- Minimum gap width (max.gapwidth) for distal transcriptional units : `r paste0(max.gapwidth, "nt")`
- Minimum novel fragment/transcriptional unit width (minWidth): `r paste0(min.frag.width, "nt")`

******

## Packages and functions

### Load required R packages {-}

```{r loadPackages, echo=FALSE, message=FALSE, warning=FALSE, results="hide"}
require("GenomicFeatures")
require("rtracklayer")
require("data.table")
require("ggplot2")
require("ggpubr")
require("cowplot")
require("ggthemes")
require("ggsci")
require("ggforce")
require("ggExtra")
require("ggrepel")
require("scales")
require("scales")
require("DT")
require("circlize")
require("BSgenome.Hsapiens.UCSC.hg38")
require("ggbio")
require("phastCons100way.UCSC.hg38")
require("GenomicAlignments")
require("genomation")
require("VennDiagram")
require("viridis")

theme_set(theme_light()) # set the default theme for ggplot (cowplot changes the default)
```

## Define custom functions

```{r sava_and_plot, echo=FALSE}
# Function to save the plot (and separately its data) to file and show the plot in the notebook
save_and_plot <- function(x, bname, width, height, plot=FALSE){
   ggsave(x, filename=paste0(working.folder, bname, ".pdf"), width=width, height=height, units="in")
   ggsave(x, filename=paste0(working.folder, bname, ".png"), width=width, height=height, units="in", dpi=300)
   if( !is.list(x$data) )
      fwrite(x$data, file=paste0(working.folder, bname, ".tsv"), quote=FALSE, sep="\t", row.names=FALSE, showProgress=FALSE)
   if( plot )
      plot(x)
}
```

```{r count_to_tpm, echo=FALSE}
# Function to transform the raw counts to Transcripts Per Million mapped reads
countToTPM <- function(countVector, regionSet){
    if (class(regionSet)[1] == "GRangesList") {
        TscLength = sum(width(regionSet))
    }
    else {
        TscLength = width(regionSet)
    }
    normTscCount = sum(countVector/TscLength)
    tpm = (countVector * 1e+06)/(normTscCount * TscLength)
    return(tpm)
}
```

```{r count_overlaps_union, echo=FALSE}
# Function to calculate the overlaps between objects containing genomic ranges using a procedure similar to the "union" method of QoRTs and HTSeq-count
countOverlapsUnion <- function(query, subject, ignore.strand = FALSE, opposite.strand = FALSE, remove.overlaps = NULL, grouping = NULL, type = "any"){
    if (opposite.strand) {
        if (class(query)[1] == "GRangesList") {
            tmp = unlist(query)
            tmp = invertStrand(tmp)
            query = relist(tmp, query)
        }
        else {
            query = invertStrand(query)
        }
        if (!is.null(remove.overlaps)) 
            subject = subject[!overlapsAny(subject, remove.overlaps, 
                ignore.strand = FALSE), ]
    }
    overlaps = as.data.table(findOverlaps(query, subject, ignore.strand = ignore.strand, 
        type = type))
    if (class(query) == "GRangesList") {
        overlaps[, `:=`(id, names(query[overlaps[, queryHits]]))]
        grouping = c(grouping, "id")
    }
    if (!is.null(grouping)) {
        if (class(query) == "GRangesList") {
            overlaps[, `:=`(id, names(query[overlaps[, queryHits]]))]
            setkeyv(overlaps, c("subjectHits", "id"))
        }
        else if (length(grouping) == 1) {
            overlaps = cbind(overlaps, as.data.table(copy(mcols(query)[overlaps[, 
                queryHits], grouping])))
            setnames(overlaps, "V1", grouping)
            setkeyv(overlaps, c("subjectHits", grouping))
        }
        else {
            overlaps = cbind(overlaps, as.data.table(copy(mcols(query)[overlaps[, 
                queryHits], grouping])))
            setkeyv(overlaps, c("subjectHits", grouping))
        }
    }
    else {
        setkey(overlaps, subjectHits)
    }
    temp = overlaps[, .N, by = key(overlaps)]
    if (any(temp[, N] > 1)) {
        sel = overlaps[, subjectHits] %in% temp[temp[, N] == 
            1, subjectHits]
        subject = subject[overlaps[sel, subjectHits]]
    }
    counts = countOverlaps(query, subject, ignore.strand = ignore.strand, 
        type = type)
    if (interactive()) 
        print(paste("Mapped", sum(counts), "reads"))
    return(counts)
}
```

```{r genomation_profile_wig, echo=FALSE}
# Function to create the metaprofile data for a set of GRanges using coverage objects
genomationProfileWig <- function(regionsList, posReadsList, negReadsList, upstream = NULL, downstream = NULL, bins = NULL, step = NULL, normalise = c("none", "max"), normGroup = NULL, start = TRUE, stranded = TRUE, quiet = FALSE){
    if (stranded) 
        stopifnot(identical(names(posReadsList), names(negReadsList)))
    stopifnot(!(is.null(step) & is.null(bins)))
    bins = ifelse(is.null(bins), (upstream + downstream)/step, 
        bins)
    if (!quiet) {
        if (is.null(upstream) | is.null(downstream)) {
            message("Creating profiles in 'whole-region' mode.")
        }
        else {
            message("Creating profiles in 'extend-region' mode.")
            message(paste("    Upstream (nt):", upstream))
            message(paste("    Downstream (nt):", downstream))
            message(paste("    Reference: region", ifelse(start, 
                "start", "end")))
        }
        message(paste("    Number of bins:", bins))
        message(paste("    Output profile:", ifelse(stranded, 
            "stranded", "unstranded")))
        message(paste("    Type of normalisation:", normalise))
    }
    tmp = lapply(seq_along(posReadsList), function(x) {
        mat = lapply(regionsList, function(y) {
            if (is.null(upstream) | is.null(downstream)) {
                y = y[width(y) >= bins, ]
                if (stranded & !is.null(negReadsList)) {
                  sm = ScoreMatrixBin(target = posReadsList[[x]], 
                    windows = y[strand(y) == "+"], bin.num = bins, 
                    bin.op = "mean", strand.aware = TRUE)
                  sn = ScoreMatrixBin(target = negReadsList[[x]], 
                    windows = y[strand(y) == "-"], bin.num = bins, 
                    bin.op = "mean", strand.aware = TRUE)
                  as(rbind(sm, sn), "ScoreMatrix")
                }
                else {
                  as(ScoreMatrixBin(target = posReadsList[[x]], 
                    windows = y, bin.num = bins, bin.op = "mean", 
                    strand.aware = TRUE), "ScoreMatrix")
                }
            }
            else {
                if (start) {
                  if (stranded & !is.null(negReadsList)) {
                    sm = ScoreMatrixBin(target = posReadsList[[x]], 
                      windows = promoters(y[strand(y) == "+"], 
                        upstream = upstream, downstream = downstream), 
                      bin.num = bins, bin.op = "mean", strand.aware = TRUE)
                    sn = ScoreMatrixBin(target = negReadsList[[x]], 
                      windows = promoters(y[strand(y) == "-"], 
                        upstream = upstream, downstream = downstream), 
                      bin.num = bins, bin.op = "mean", strand.aware = TRUE)
                    as(rbind(sm, sn), "ScoreMatrix")
                  }
                  else {
                    as(ScoreMatrixBin(target = posReadsList[[x]], 
                      windows = promoters(y, upstream = upstream, 
                        downstream = downstream), bin.num = bins, 
                      bin.op = "mean", strand.aware = TRUE), 
                      "ScoreMatrix")
                  }
                }
                else {
                  if (stranded & !is.null(negReadsList)) {
                    sm = ScoreMatrixBin(target = posReadsList[[x]], 
                      windows = invertStrand(promoters(invertStrand(y[strand(y) == 
                        "+"]), upstream = downstream, downstream = upstream)), 
                      bin.num = bins, bin.op = "mean", strand.aware = TRUE)
                    sn = ScoreMatrixBin(target = negReadsList[[x]], 
                      windows = invertStrand(promoters(invertStrand(y[strand(y) == 
                        "-"]), upstream = downstream, downstream = upstream)), 
                      bin.num = bins, bin.op = "mean", strand.aware = TRUE)
                    as(rbind(sm, sn), "ScoreMatrix")
                  }
                  else {
                    as(ScoreMatrixBin(target = posReadsList[[x]], 
                      windows = invertStrand(promoters(invertStrand(y), 
                        upstream = downstream, downstream = upstream)), 
                      bin.num = bins, bin.op = "mean", strand.aware = TRUE), 
                      "ScoreMatrix")
                  }
                }
            }
        })
        names(mat) = names(regionsList)
        mat = as(mat, "ScoreMatrixList")
        dt = lapply(mat, function(x) data.table(as(x, "matrix")))
        if (is.null(normGroup)) {
            normGroup = names(mat)
        }
        for (i in seq_along(normGroup)) {
            if (normalise == "max") {
                nFact = rowMax(sapply(dt[normGroup[[i]]], function(z) rowMax(as.matrix(z))))
            }
            else {
                nFact = 1
            }
            for (ii in normGroup[[i]]) {
                dt[[ii]] = dt[[ii]]/nFact
                mat2 = as.matrix(dt[[ii]])
                mat2[!is.finite(mat2)] = 0
                dt[[ii]] = as.data.table(mat2)
            }
        }
        dt = rbindlist(dt, idcol = "Unit")
        dt = melt.data.table(dt, id.vars = "Unit", variable.name = "Position")
        dt[, `:=`(Position, as.numeric(gsub("V", "", Position)))]
        setkeyv(dt, c("Unit", "Position"))
        dt[, list(Value = mean(value)), by = key(dt)]
    })
    names(tmp) = names(posReadsList)
    tmp = rbindlist(tmp, idcol = "Sample")
    if (is.null(upstream) | is.null(downstream)) {
        labels = data.table(Label = seq(0, 100, length.out = 201))
        labels[, `:=`(Position, seq_along(Label) - 1)]
    }
    else {
        labels = data.table(Label = seq(-upstream, downstream, 
            step)/1000)
        labels[, `:=`(Position, seq_along(Label))]
    }
    setkey(labels, Position)
    setkey(tmp, Position)
    tmp = tmp[labels, `:=`(Position, Label)]
    return(tmp)
}
```

## Load the reference annotation objects

```{r loadAnnotation, echo=FALSE, message=FALSE}
# Define the set of chromosome to use
chromosomes = genomeStyles()[[species]]$UCSC
chromosomes = setdiff(chromosomes, c("chrY", "chrM"))
chromosomes.lengths = seqlengths(Hsapiens)
chromosomes.lengths = chromosomes.lengths[names(chromosomes.lengths)%in%chromosomes]

# Load the pre-processed GRanges and print the list
load(file.path(annotation.folder, "hg38_Gencode27_annotations.RData"))
load(file.path(annotation.folder, "hg38_Gencode27_annotations.all.genes.transcript.regions.RData"))

rm(exonsByTxs, txs, txs.metadata, txs.longest.pc.rna.granges, txs.longest.nc.rna.granges)

exons.pc.granges = updateObject(exons.pc.granges, verbose=TRUE)
exons.nc.granges = updateObject(exons.nc.granges, verbose=TRUE)

# Get the introns
introns = genes
names(introns) = NULL
introns = split(introns, introns$gene_id)
exons = c(exons.pc.granges, exons.nc.granges)
introns = GenomicRanges::setdiff(introns[match(names(exons), names(introns))],
                                 c(exons.pc.granges, exons.nc.granges))

lnc.gff = import.gff3(file.path(annotation.folder, "gencode.v27.long_noncoding_RNAs.gff3.gz"))
lnc.gff = keepSeqlevels(lnc.gff, chromosomes, pruning.mode="coarse")

# Remove the duplicated genes on the chromosome Y
x.genes = unique(lnc.gff[seqnames(lnc.gff)%in%"chrX"]$gene_id)
y.genes = unique(lnc.gff[seqnames(lnc.gff)%in%"chrY"]$gene_id)
y.genes = y.genes[y.genes%in%x.genes]
lnc.gff = lnc.gff[!(seqnames(lnc.gff)%in%"chrY" & lnc.gff$gene_id%in%y.genes)]

# Group gene biotypes into classes
protein.coding = gene.metadata[gene_type%in%c("protein_coding", paste(rep(c("IG","TR"),each=4), c("C","D","J","V"), "gene", sep="_"), "IG_LV_gene"), gene_id]
long.ncRNA = gene.metadata[gene_type%in%unique(names(sort(-table(lnc.gff$gene_type)))), gene_id]
ncRNA = c(setdiff(gene.metadata[grepl("RNA", gene_type), gene_id], long.ncRNA), "ribozyme")
other = setdiff(gene.metadata[, gene_id], c(ncRNA, long.ncRNA, protein.coding))

rm(lnc.gff)

tab.biotype = data.table(Gene_id = c(protein.coding, long.ncRNA, ncRNA, other),
                         Biotype = rep(c("protein_coding", "long_ncRNA", "ncRNA", "other annotated"),
                                       c(length(protein.coding), length(long.ncRNA), length(ncRNA), length(other))),
                         key="Gene_id")

entrez.genes = fread(cmd=paste("gunzip -c", file.path(annotation.folder, "gencode.v27.metadata.EntrezGene.gz")))

geneset = keepSeqlevels(genes, chromosomes, pruning.mode="coarse")
```

## Gene expression

### Evaluate quantification statistics

```{r readStatsQoRTs}
# Get the list of 'strand' files
files = list.files(paste0(working.folder, "counts"), pattern="*StrandCheck.out.tab", full.names=TRUE)

# Create a data.table containing the information on the library strand
allTable = lapply(files, function(x) fread(x, header=FALSE, col.names="library"))
names(allTable) = gsub("\\..*$", "", basename(files))
allTable = rbindlist(allTable, idcol="name")
allTable[, file := gsub("StrandCheck", "ReadsPerGene", files)]

allTable[, group := paste(tstrsplit(name, "_")[[1]], gsub("-.*$", "", tstrsplit(name, "_")[[2]]), tstrsplit(name, "_")[[4]], sep="_")]

# Keep only the WildType/Untreated samples
gffTable = allTable[grepl("_(Untreated|WildType|siLuc|Bcells|Unmodified-XRN2)_", name) & !grepl("_(UV)_", name),]

# Import the count files into a data.table
seqStats = do.call(cbind,
                 lapply(gffTable[, name],
                        function(x)
                           fread(paste0("counts/", x, "/QC.summary.txt"), col.names=as.character(x), select=2)))
df = fread(paste0("counts/", gffTable[1, name],"/QC.summary.txt"), stringsAsFactors=FALSE, header=TRUE)

seqStats = data.frame(row.names=df[, FIELD], seqStats)

fields.genes = c("Genes_WithNonzeroCounts")
fields.reads = c("ReadPairs_UniqueGene_CDS", "ReadPairs_UniqueGene_UTR", "ReadPairs_NoGene_Intron",
                 "ReadPairs_NoGene_OneKbFromGene", "ReadPairs_NoGene_TenKbFromGene", "ReadPairs_NoGene_MiddleOfNowhere")
fields.splicing = c("SpliceLoci_Known", "SpliceLoci_Novel", "SpliceEvents_KnownLoci", "SpliceEvents_NovelLoci")

seqStats.genes = data.table(seqStats[rownames(seqStats)%in%fields.genes,], keep.rownames=TRUE)
seqStats.reads = data.table(seqStats[rownames(seqStats)%in%fields.reads,], keep.rownames=TRUE)
seqStats.splicing = data.table(seqStats[rownames(seqStats)%in%fields.splicing,], keep.rownames=TRUE)

seqStats.reads[, c("Group","Subgroup") := tstrsplit(seqStats.reads[, rn], "_", keep=2:3)]
seqStats.reads[, rn := NULL]
seqStats.reads[Subgroup%in%"Intron", Group := "UniqueGene"]

seqStats.reads.merged = melt(seqStats.reads, id.vars=c("Group", "Subgroup"))
seqStats.reads.merged = seqStats.reads.merged[, variable := gsub("_rep.*$", "", variable)][, list(value = sum(value)), by=c("Group", "Subgroup", "variable")]

seqStats.reads = seqStats.reads[, c(.(Group, Subgroup), lapply(.SD, function(x) x/sum(x))), .SDcols = gsub("-", ".", gffTable[, name])]

seqStats.reads = melt(seqStats.reads, id.vars=c("V1","V2"))
seqStats.reads[, `:=`(c("V1", "V2"),
                      list(factor(V1, levels=c("UniqueGene", "NoGene")),
                           factor(V2, levels=c("CDS","UTR", "Intron", "OneKbFromGene","TenKbFromGene","MiddleOfNowhere"))))]
seqStats.reads[, c("percentage", "position") := list(round(value*100, 1), cumsum(value) - 0.5*value), by=c("variable")]
```

```{r plot_stats_1, fig.align="center", fig.width=26, fig.height=14}
seqStats.reads[, c("Sequencing", "Group", "Condition", "Cell", "Replicate", "Library") := tstrsplit(variable, "_")]

tab = seqStats.reads[, list(percentage = sum(value)), by=c("V1", "variable")]
tab = tab[, list(percentage=round(percentage*100, 1), position = cumsum(percentage) - 0.5*percentage), by=c("variable")]
tab[, c("Sequencing", "Group", "Condition", "Cell", "Replicate", "Library") := tstrsplit(variable, "_")]

myColours = setNames(ptol_pal()(8), c(levels(seqStats.reads[, V1]), levels(seqStats.reads[,V2])))
names(myColours) = factor(names(myColours), levels=names(myColours))

seqStats.reads[, `:=`(V1 = factor(V1, levels=names(myColours)), V2 = factor(V2, levels=names(myColours)))]

seqStats.reads[, Group := factor(Group,
                              levels=unlist(sapply(c("Chromatin", "Nuclear", "Nucleoplasm", "Cytoplasm"),
                                                   function(g)
                                                      unique(grep(g, Group, value=TRUE)))))]
tab[, Group := factor(Group,
                      levels=unlist(sapply(c("Chromatin", "Nuclear", "Nucleoplasm", "Cytoplasm"),
                                           function(g)
                                              unique(grep(g, Group, value=TRUE)))))]

seqStats.reads[Condition%in%"WildType", Condition := "Untreated"]
tab[Condition%in%"WildType", Condition := "Untreated"]

# gg0 = ggplot() +
#    geom_bar(data=seqStats.reads, aes(fill=V1, x=as.factor(1), y=value), width=0.9, stat="identity") +
#    geom_bar(data=seqStats.reads, aes(fill=V2, x=as.factor(2), y=value), width=0.9, stat="identity") +
#    geom_text(data=tab, aes(x = 1.1, y = 1-position, label = paste0(percentage, "%")), col="white", size = 3) +
#    # geom_text(aes(x = rep(c(2.1, 2.1, 2.1, 2.6, 2.8, 2.6), length(replicates)), y = 1-position, label = paste0(percentage, "%")), col="black", size = 3) +
#    coord_polar(theta = "y", direction=-1) +
#    scale_fill_manual(name = "Mapped regions", values = myColours, drop=FALSE) +
#    facet_grid(Condition+Replicate~Group+Cell) +
#    theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.title=element_blank(),
#          panel.grid  = element_blank(), panel.background = element_blank(),
#          strip.background = element_blank(), strip.text.x = element_text(size = 10), legend.position="bottom")
# save_and_plot(x=gg0, bname="Figures_and_Tables/00_Statistics_on_the_mapped_reads", width=26, height=14)
```

```{r plot_stats_2, fig.align="center", fig.width=8, fig.height=6}
seqStats.reads = seqStats.reads[Cell%in%"HeLa" & Condition%in%c("Untreated", "WildType") & !grepl("pA", Group)]
seqStats.reads = seqStats.reads[, list(value=mean(value)), by=c("Group", "V1", "V2")]

seqStats.reads[, c("percentage", "position") := list(round(value*100, 1), cumsum(value) - 0.5*value), by=c("Group")]
seqStats.reads[, Group := factor(Group, levels=c("Nuclear", "Cytoplasm", "Chromatin", "Nucleoplasm"))]

tab = seqStats.reads[, list(percentage = sum(value)), by=c("V1", "Group")]
tab = tab[, list(percentage=round(percentage*100, 1), position = cumsum(percentage) - 0.5*percentage), by=c("Group")]

seqStats.reads[, `:=`(V1 = factor(V1, levels=names(myColours)), V2 = factor(V2, levels=names(myColours)))]

gg0.1 = ggplot() +
   geom_bar(data=seqStats.reads, aes(fill=V1, col=V1, x=as.factor(1), y=value), width=0.9, stat="identity") +
   geom_bar(data=seqStats.reads, aes(fill=V2, col=V2, x=as.factor(2), y=value), width=0.9, stat="identity") +
   geom_text(data=tab, aes(x = 1.1, y = 1-position, label = paste0(percentage, "%")), col="white", size = 4) +
   # geom_text(aes(x = rep(c(2.1, 2.1, 2.1, 2.6, 2.8, 2.6), length(replicates)), y = 1-position, label = paste0(percentage, "%")), col="black", size = 3) +
   coord_polar(theta = "y", direction=-1) +
   scale_fill_manual(name = "Mapped regions", values = myColours, drop=FALSE) +
   scale_colour_manual(name = "Mapped regions", values = myColours, drop=FALSE) +
   facet_wrap(~Group, ncol=2) +
   theme(axis.text = element_blank(), axis.ticks = element_blank(), axis.title=element_blank(),
         panel.grid  = element_blank(), panel.background = element_blank(),
         strip.background = element_blank(), strip.text.x = element_text(size = 12), legend.position="right")
save_and_plot(x=gg0.1, bname="Figures_and_Tables/Supplementary_1A", width=8, height=6, plot=FALSE)
```

![](./Figures_and_Tables/Supplementary_1A.png)

### Import QoRTs

```{r readCountsQoRTs}
# Read the count files
annoCountsSplit = lapply(c("COUNT", "intronic"),
                         function(y)
                            do.call(cbind,
                                    lapply(gffTable[, name],
                                           function(x){
                                              tmp = fread(cmd=paste("gunzip -c", file.path("counts", x, "QC.geneCounts.detailed.txt.gz")))
                                              tmp[GENEID%like%"ENSG", (as.character(y)), with=FALSE]
                                           })))
names(annoCountsSplit) = c("Exonic", "Intronic")
colnames(annoCountsSplit[[1]]) = gffTable[, name]
colnames(annoCountsSplit[[2]]) = gffTable[, name]
# Create the Exon+Intron counts
annoCountsSplit[["Genic"]] = annoCountsSplit[[1]]+annoCountsSplit[[2]]

# Read the first file to assign gene names
df = fread(cmd=paste("gunzip -c", file.path("counts", gffTable[1, name], "QC.geneCounts.detailed.txt.gz")), stringsAsFactors=FALSE, header=TRUE)
annoCountsSplit = lapply(annoCountsSplit,
                         function(x)
                            data.frame(row.names=df[grepl("ENSG", GENEID), GENEID], x))

# Select which count to use for the genes (Exonic, intronic or genic)
countTable = annoCountsSplit[["Genic"]]

rm(annoCountsSplit)
# Re-order the counts to match the GRanges order
countTable = countTable[match(genes$gene_id, rownames(countTable)),]

# Calculate the TPM values
countTPM = apply(countTable, 2, function(x) countToTPM(x, genes))
```

## Build the annotation

### Load gffCompare (StringTie merged) output

```{r loadStringTie}
# Import the StringTie (or rather gffcompare) GFF output
gff = fread(paste0(working.folder, "gffcompare.combined.sorted.gtf"), sel=c(1,3:5,7,9),
            col.names=c("seqname", "feature", "start", "end", "strand", "attribute"), showProgress=FALSE)

# Filter and select the entries
gff = gff[seqname%in%chromosomes & feature%in%"transcript",]
gff[, attribute := gsub("\"|;", "", attribute)]
gff[, gene_id := sub(".*gene_id\\s(\\S+).*", "\\1", attribute)]
gff[, transcript_id := sub(".*transcript_id\\s(\\S+).*", "\\1", attribute)]
gff[, code := sub(".*class_code\\s(\\S+).*", "\\1", attribute)]
gff[, attribute := NULL]

# Create the final data.table and GRanges
gff.genes = gff[, list(seqname=unique(seqname), start=min(start), end=max(end), strand=unique(strand)), by="gene_id"]
gff.genes.gr = sort.GenomicRanges(with(gff.genes, GRanges(seqname, IRanges(start, end), strand, gene_id)), ignore.strand=TRUE)

rm(gff)

# ### Test ###
# a = reduce(gff.genes.gr)
# b = precede(a, a)
# vec = data.table(Width=width(a[!is.na(b)]), DistanceToNearest=distance(a[!is.na(b)], a[na.omit(b)]))
# 
# gg1 = ggplot(vec, aes(x=Width, y=DistanceToNearest)) +
#    geom_point(size=0.5, alpha=0.8, fill="dodgerblue3", colour="dodgerblue4", pch=21) +
#    scale_x_log10() + scale_y_log10() + geom_vline(xintercept=200, lty=2) + theme_light()
# gg1 = ggarrange(ggMarginal(gg1, type = "density", size = 4))
# save_and_plot(gg1, bname="Figures_and_Tables/StringTie_merged_fragments", width=8, height=8)
# 
# a = setdiff(reduce(gff.genes.gr), geneset)
# b = precede(a, a)
# vec = data.table(Width=width(a[!is.na(b)]), DistanceToNearest=distance(a[!is.na(b)], a[na.omit(b)]))
# 
# gg2 = ggplot(vec, aes(x=Width, y=DistanceToNearest)) +
#    geom_point(size=0.5, alpha=0.8, fill="dodgerblue3", colour="dodgerblue4", pch=21) +
#    scale_x_log10() + scale_y_log10() + geom_vline(xintercept=200, lty=2) + theme_light()
# save_and_plot(gg2, bname="Figures_and_Tables/StringTie_merged_fragments_noOverlapsAnno", width=8, height=8)
# ### End ###

# Remove the Mitochondrial chromosome
gff.genes.gr = reduce(gff.genes.gr[!seqnames(gff.genes.gr)%in%"chrM",], min.gapwidth=max.gapwidth)

annotated.gr = geneset

mcols(annotated.gr) = data.frame(Set = factor("Annotated", levels=c("Annotated", "New")))
mcols(gff.genes.gr) = data.frame(Set = factor("New", levels=c("Annotated", "New")))
```

![](./Figures_and_Tables/StringTie_merged_fragments.png)

![](./Figures_and_Tables/StringTie_merged_fragments_noOverlapsAnno.png)

### Expand annotation

```{r newCheck}
if( REBUILD ){
   anno.new = setdiff(gff.genes.gr, annotated.gr)
   mcols(anno.new) = data.frame(Set="Novel")
   anno.old = annotated.gr
   
   new.anno.genes = c(anno.old, anno.new)
   new.anno.genes.dt = data.table(as.data.table(copy(new.anno.genes)))
   
   # anno.gap.gr = new.anno.genes.dt[, list(start=min(start), end=max(end)), by=c("seqnames", "strand")]
   # anno.gap.gr = setdiff(with(anno.gap.gr, GRanges(seqnames, IRanges(start, end), strand)), new.anno.genes)
   anno.gap.gr = with(new.anno.genes.dt, GRanges(seqnames, IRanges(start, end), strand))
   anno.gap.gr = reduce(c(flank(reduce(anno.gap.gr), 1e4, start=TRUE, both=FALSE),
                          flank(reduce(anno.gap.gr), 1e4, start=FALSE, both=FALSE)))
   
   export.bed(anno.gap.gr, "annotatedRegions/anno.gap.gr.bed")
   
   # anno.ann.pos = new.anno.genes[new.anno.genes$Set%in%"Annotated" & strand(new.anno.genes)=="+"]
   # anno.ann.neg = new.anno.genes[new.anno.genes$Set%in%"Annotated" & strand(new.anno.genes)=="-"]
   anno.new.pos = new.anno.genes[!new.anno.genes$Set%in%"Annotated" & strand(new.anno.genes)=="+"]
   anno.new.neg = new.anno.genes[!new.anno.genes$Set%in%"Annotated" & strand(new.anno.genes)=="-"]
   anno.gap.pos = anno.gap.gr[strand(anno.gap.gr)=="+"]
   anno.gap.neg = anno.gap.gr[strand(anno.gap.gr)=="-"]
   
   qnt_th = 0.1
   
   new.anno.frags = list()
   new.anno.gaps  = list()
   for( x in gffTable[, name] ){
      # for( x in c("RNAseq_Chromatin_siLuc_HeLa_1_PE", "RNAseq_Chromatin_Untreated_K562_2_PE", "RNAseq_Cytoplasm-pAm_Untreated_HepG2_2_PE", "RNAseq_Cytoplasm_WildType_HeLa_3_SE", "RNAseq_Nuclear-pAm_Untreated_K562_1_PE", "RNAseq_Nuclear-pAp_Untreated_K562_1_PE", "RNAseq_Nucleoplasm_Untreated_HeLa_1_PE") ){
      anno.new.bwp = readBigWig(paste0("RNAseq_bw/", x, "_plus_CPM.bigwig"), anno.new.pos)
      anno.new.bwp = anno.new.bwp[seqlevels(anno.new.pos)]
      anno.new.bwp = anno.new.bwp[anno.new.pos]
      
      anno.new.bwm = readBigWig(paste0("RNAseq_bw/", x, "_minus_CPM.bigwig"), anno.new.neg)
      anno.new.bwm = anno.new.bwm[seqlevels(anno.new.neg)]
      anno.new.bwm = anno.new.bwm[anno.new.neg]
      
      anno.new.bw = c(mean(anno.new.bwp), mean(anno.new.bwm))
      
      qnt1 = quantile(anno.new.bw[anno.new.bw>0], qnt_th)
      
      new.anno.frags[[as.character(x)]] = c(anno.new.pos, anno.new.neg)[anno.new.bw>qnt1]
      
      anno.gap.bwp = readBigWig(paste0("RNAseq_bw/", x, "_plus_CPM.bigwig"), anno.gap.pos)
      anno.gap.bwp = anno.gap.bwp[seqlevels(anno.gap.pos)]
      anno.gap.bwp = anno.gap.bwp[unlist(tile(anno.gap.pos, width=min.frag.width))]
      
      anno.gap.bwm = readBigWig(paste0("RNAseq_bw/", x, "_minus_CPM.bigwig"), anno.gap.neg)
      anno.gap.bwm = anno.gap.bwm[seqlevels(anno.gap.neg)]
      anno.gap.bwm = anno.gap.bwm[unlist(tile(anno.gap.neg, width=min.frag.width))]
      
      anno.gap.bw = c(mean(anno.gap.bwp), mean(anno.gap.bwm))
      
      qnt2 = quantile(anno.gap.bw[anno.gap.bw>0], 1-qnt_th)
      
      selected.gaps = c(unlist(tile(anno.gap.pos, width=min.frag.width)), unlist(tile(anno.gap.neg, width=min.frag.width)))[anno.gap.bw>qnt2]
      
      selected.gaps = reduce(selected.gaps, min.gapwidth=max.gapwidth)
      
      # new.anno.gaps[[as.character(x)]] = selected.gaps[overlapsAny(selected.gaps, new.anno.genes, maxgap=max.gapwidth),]
      new.anno.gaps[[as.character(x)]] = selected.gaps[overlapsAny(selected.gaps, new.anno.frags[[as.character(x)]], maxgap=max.gapwidth),]
      
      print(paste(x,
                  length(new.anno.frags[[as.character(x)]]),
                  round(length(new.anno.frags[[as.character(x)]])/length(c(anno.new.pos, anno.new.neg))*100, 1),
                  length(new.anno.gaps[[as.character(x)]]),
                  round(length(new.anno.gaps[[as.character(x)]])/length(c(anno.gap.pos, anno.gap.neg))*100, 1)))
      
   }
   
   names(new.anno.frags) = NULL
   new.anno.frags = reduce(do.call(c, lapply(new.anno.frags, function(x) x)))
   export.bed(new.anno.frags, "annotatedRegions/new.anno.frags.bed")
   
   mcols(new.anno.frags) = data.frame(Set="Novel")
   saveRDS(new.anno.frags, file=paste0(working.folder, paste0("RData/newAnnotation_StringTie_frags_CPM_", qnt_th, ".rds")))
   
   names(new.anno.gaps) = NULL
   new.anno.gaps = reduce(do.call(c, lapply(new.anno.gaps, function(x) x)))
   export.bed(new.anno.gaps, "annotatedRegions/new.anno.gaps.bed")
   
   mcols(new.anno.frags) = data.frame(Set="Gaps")
   saveRDS(new.anno.gaps, file=paste0(working.folder, paste0("RData/newAnnotation_StringTie_gaps_CPM_", 1-qnt_th, ".rds")))
}
```

```{r recreateAnnotation}
if( !exists("new.anno.frags") )
   new.anno.frags = readRDS(paste0(working.folder, paste0("RData/newAnnotation_StringTie_frags_CPM_", qnt_th, ".rds")))

if( !exists("new.anno.gaps") )
   new.anno.gaps = readRDS(paste0(working.folder, paste0("RData/newAnnotation_StringTie_gaps_CPM_", 1-qnt_th, ".rds")))

gff.genes.gr = reduce(c(new.anno.frags, new.anno.gaps), min.gapwidth=max.gapwidth)
rm(new.anno.frags, new.anno.gaps)
gff.genes.gr = setdiff(gff.genes.gr, annotated.gr)
mcols(gff.genes.gr) = data.frame(Set="Novel")

# Create a GRanges object with all the 'gaps', excluding annotated+stringtie
new.anno.genes = c(annotated.gr, gff.genes.gr)
new.anno.genes.dt = data.table(as.data.table(copy(new.anno.genes)))
```

### Refine annotation

```{r processStringTie}
# Remove the intragenic fragments and separate proximal from distal
intragenic = gff.genes.gr[overlapsAny(gff.genes.gr, annotated.gr, maxgap=max.gapwidth)]
proximal = setdiff(intragenic, annotated.gr)
intergenic = reduce(gff.genes.gr[!overlapsAny(gff.genes.gr, intragenic, maxgap=max.gapwidth)])

# Create data.table objects for each set
mcols(annotated.gr) = NULL
annotated.dt = data.table(as.data.frame(annotated.gr), index=seq_along(annotated.gr), gene_id=names(annotated.gr))
annotated.dt[, `:=`(Set = "Annotated", Subset = tab.biotype[match(gene_id, tab.biotype[, Gene_id]), Biotype])]
proximal.dt = data.table(as.data.frame(proximal), index=seq_along(proximal))
intergenic.dt = data.table(as.data.frame(intergenic), index=seq_along(intergenic))

# Find the genes preceding and/or following the region
# NB: precede() returns the index of the GRange that is preceded by (i.e., comes after) the query;
# follow() returns the index of the GRange that is followed by (comes before) the query
proximal.dt[, `:=`(PreviousGene=as.numeric(NA), NextGene=as.numeric(NA),
                   PreviousDistance=as.integer(NA), NextDistance=as.integer(NA))]

# Add the indexes and distances of neighbouring genes
proximal.dt[, `:=`(PreviousGene = follow(proximal, geneset),
                   NextGene = precede(proximal, geneset))]
proximal.dt[!is.na(PreviousGene), PreviousDistance := distance(proximal[index,], geneset[PreviousGene,])]
proximal.dt[!is.na(NextGene), NextDistance := distance(proximal[index,], geneset[NextGene,])]

# Assign additional columns to the proximal set
proximal.dt[, `:=`(gene_id = paste("Proximal", seq_along(seqnames), sep="_"), Set = "Proximal", Subset = as.character(NA))]
proximal.dt[!is.na(PreviousDistance) & PreviousDistance <= max.gapwidth | is.na(NextDistance), Subset := "DownstreamOfGene"]
proximal.dt[!is.na(NextDistance) & NextDistance <= max.gapwidth | is.na(PreviousDistance), Subset := "UpstreamOfGene"]
proximal.dt[PreviousDistance <= max.gapwidth & NextDistance <= max.gapwidth, Subset := "LinkerOfGene"]

# Close the gaps with the nearest gene(s)
proximal.dt[Subset%in%"DownstreamOfGene" & strand=="+", start := as.integer(end(geneset[PreviousGene,])+1)]
proximal.dt[Subset%in%"DownstreamOfGene" & strand=="-", end := as.integer(start(geneset[PreviousGene,])-1)]
proximal.dt[Subset%in%"UpstreamOfGene" & strand=="+", end := as.integer(start(geneset[NextGene,])-1)]
proximal.dt[Subset%in%"UpstreamOfGene" & strand=="-", start := as.integer(end(geneset[NextGene,])+1)]
proximal.dt[Subset%in%"LinkerOfGene" & strand=="+", `:=`(start = as.integer(end(geneset[PreviousGene,])+1), end = as.integer(start(geneset[NextGene,])-1))]
proximal.dt[Subset%in%"LinkerOfGene" & strand=="-", `:=`(start = as.integer(end(geneset[NextGene,])+1), end = as.integer(start(geneset[PreviousGene,])-1))]

# # Remove features that are below the minimum threshold
# proximal.dt = proximal.dt[width>=min.feature.width,]

# Check that the linker of genes are actually connecting pairs of expressed genes
linkers.dt = proximal.dt[Subset%in%"LinkerOfGene",]
linkers.dt[, `:=`(NextID=names(geneset[NextGene]), PreviousID=names(geneset[PreviousGene]))]
for( i in 1:nrow(linkers.dt) ){
   temp = rbind(countTPM[rownames(countTPM)%in%linkers.dt[i, NextID],], countTPM[rownames(countTPM)%in%linkers.dt[i, PreviousID],])>min.expr
   if( any(colSums(temp)>1) ){
      next   
   }else if( rowSums(temp)[1] > rowSums(temp)[2] ){
      linkers.dt[i, Subset := "UpstreamOfGene"]
   }else{
      linkers.dt[i, Subset := "DownstreamOfGene"]
   }
}
proximal.dt[Subset%in%"LinkerOfGene", Subset := linkers.dt[, Subset]]

# Remove unneeded columns
proximal.dt[, `:=`(PreviousGene=NULL, NextGene=NULL, PreviousDistance=NULL, NextDistance=NULL)]

# Assign additional columns to the intergenic set
intergenic.dt[, `:=`(gene_id = paste("Distal", seq_along(seqnames), sep="_"), Set = "Intergenic",
                     Subset = ifelse(mcols(distanceToNearest(intergenic, geneset))$distance<min.gapwidth, "< 10Kb", "> 10Kb"))]

# # Remove features that are below the minimum threshold
# intergenic.dt = intergenic.dt[width>=min.feature.width,]

# Merge all annotation sets
new.anno.genes = rbindlist(list(annotated.dt, proximal.dt[width>=min.feature.width,], intergenic.dt[width>=min.feature.width,]))

# Create and sort the GRanges
new.anno.genes = with(new.anno.genes, GRanges(seqnames, IRanges(start, end), strand, gene_id, Set, Subset))
new.anno.genes = sort.GenomicRanges(new.anno.genes, ignore.strand=TRUE)

# new.anno.genes.dt = data.table(as.data.frame(new.anno.genes))
# 
# new.anno.genes.dt[, `:=`(index = .I,
#                          previousFeature = follow(new.anno.genes, new.anno.genes),
#                          nextFeature = precede(new.anno.genes, new.anno.genes))]
# new.anno.genes.dt[!is.na(previousFeature), `:=`(previousType = new.anno.genes[previousFeature,]$Subset,
#                                                 previousDistance = distance(new.anno.genes[index,], new.anno.genes[previousFeature,]))]
# new.anno.genes.dt[!is.na(nextFeature), `:=`(nextType = new.anno.genes[nextFeature,]$Subset,
#                                             nextDistance = distance(new.anno.genes[index,], new.anno.genes[nextFeature,]))]
# 
# select_1 = new.anno.genes.dt[, Subset%in%c("< 10Kb", "> 10Kb")] # Select only intergenic features
# select_2 = new.anno.genes.dt[, !is.na(previousDistance) & previousDistance<min.gapwidth] # Select features closer than 1000nt
# select_3 = new.anno.genes.dt[, !previousType%in%c("protein_coding", "long_ncRNA", "ncRNA", "other annotated")] # Select only features following another un-annotated feature
# select_4 = new.anno.genes.dt[, !is.na(nextDistance) & nextDistance<min.gapwidth] # Select features closer than 1000nt
# select_5 = new.anno.genes.dt[, !nextType%in%c("protein_coding", "long_ncRNA", "ncRNA", "other annotated")] # Select only features preceeding another un-annotated feature
# 
# new.anno.genes.dt = new.anno.genes.dt[!(select_1 & ((select_2 & select_3) | (select_4 & select_5)))]

export.bed(new.anno.genes, "annotatedRegions/new.anno.genes.bed")

# Save the object
saveRDS(new.anno.genes, file=paste0(working.folder, "RData/newAnnotationExtended_StringTie.rds"))
```

### Final annotation distribution

```{r plotAnnotation_1, fig.align="center", fig.width=10, fig.height=8}
# dt = as.data.table(copy(mcols(new.anno.genes[width(new.anno.genes)>=min.feature.width | new.anno.genes$Set%in%"Annotated"])))
# dt = dt[!Subset%in%"GencodeGene", .N, by="Subset"][, Freq := N/sum(N)]
# dt[, Subset := factor(Subset, levels=factBiotype2)]
# setkey(dt, Subset)
# setnames(dt, "Subset", "Type")
# 
# dt[, `:=`(Perc = round(dt[, N]/sum(dt[, N]), 3), Pos = cumsum(dt[, Freq]) - 0.5*dt[, Freq])]
# dt[Type%in%c("protein_coding", "long_ncRNA", "ncRNA", "other annotated"), Feature := "Annotated"]
# dt[is.na(Feature), Feature := "Discovered"]
# 
# # gg2 = ggplot(dt, aes(x=as.factor(1), y=Freq, fill=Type)) +
# #    ggtitle("") +
# #    geom_bar(width=1, stat="identity", position="stack", col="black") +
# #    geom_text(aes(x = 1.9, y = 1-Pos, label = paste0(Type, "\n", N, " (", Perc*100,"%)")), col="black") +
# #    coord_polar(theta="y", direction=-1, start=-0.5) +
# #    # scale_fill_manual(values=myColors, guide=FALSE) +
# #    scale_fill_tableau(palette = "Classic 10 Medium", guide=FALSE) +
# #    theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(), axis.text.y = element_blank(), axis.text.x = element_blank(),
# #          axis.title=element_blank(), panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14) )
# # save_and_plot(x=gg2, bname="Figures_and_Tables/Figure_1C_pie", width=6, height=6)
# 
# gg2 = ggplot(dt, aes(x=Type, y=N, fill=Feature)) +
#    ggtitle("Annotated and unannotated transcribed regions") +
#    geom_bar(stat="identity", position="dodge", col="black") +
#    geom_text(aes(x = Type, y = N+2e3, label = paste0(N, "\n(", Perc*100,"%)")), col="black") +
#    scale_x_discrete("") + scale_y_continuous("Count", limits=c(0, 24e3)) +
#    scale_fill_manual(values=setNames(c("#849db1", "#fbb04e"), c("Annotated", "Discovered"))) +
#    # scale_fill_tableau(palette = "Classic 10 Medium", guide=FALSE) +
#    theme_bw() + theme(legend.position=c(0.1, 0.9), axis.text=element_text(size=12), axis.title=element_text(size=14), axis.text.x=element_text(angle=45, hjust=1, vjust=1))
# save_and_plot(x=gg2, bname="Figures_and_Tables/Figure_1C", width=10, height=8)

dt = rbindlist(lapply(reduce(split(new.anno.genes, new.anno.genes$Subset), ignore.strand=TRUE),
                 function(x)
                    data.table(N=sum(width(x)))), idcol="Feature")
genome.length = data.table(as.data.frame(seqinfo(genes)), keep.rownames="seqnames")[seqnames%in%chromosomes, sum(seqlengths)]

dt[, Percentage := N/genome.length]
dt[, Feature := factor(Feature, levels=factBiotype2)]
dt[Feature%in%c("protein_coding", "long_ncRNA", "ncRNA", "other annotated"), FeatureType := "Annotated"]
dt[is.na(FeatureType), FeatureType := "Discovered"]

figure1B = ggplot(dt, aes(x=Feature, y=Percentage, fill=FeatureType)) +
   ggtitle("Genome coverage of annotated and unannotated transcribed regions") +
   geom_bar(stat="identity", position="dodge", colour="black") +
   scale_x_discrete("") +
   scale_y_continuous("Percentage of genome covered", labels=percent) +
   scale_fill_manual(values=setNames(c("#849db1", "#fbb04e"), c("Annotated", "Discovered"))) +
   # scale_fill_tableau(palette = "Classic 10 Medium", guide=FALSE) +
   facet_zoom(y = Percentage <= 0.1) +
   theme_bw() + theme(legend.position=c(0.8, 0.85), axis.text=element_text(size=12), axis.title=element_text(size=14), axis.text.x=element_text(angle=45, hjust=1, vjust=1))
save_and_plot(x=figure1B, bname="Figures_and_Tables/Figure_1ABC", width=10, height=6)
```

![](./Figures_and_Tables/Figure_1C.png)

![](./Figures_and_Tables/Figure_1ABC.png)

```{r intergenicDistance}
dist = distanceToNearest(new.anno.genes[new.anno.genes$Set%in%"Intergenic"], new.anno.genes[new.anno.genes$Set%in%"Annotated"])

dt = data.table(Feature=new.anno.genes[new.anno.genes$Set%in%"Intergenic"]$Subset, Distance=mcols(dist)$distance)

dens <- density(log10(dt[, Distance]))
df <- data.frame(x=dens$x, y=dens$y)
probs <- c(0, 0.25, 0.5, 0.75, 1)
quantiles <- quantile(log10(dt$Distance), prob=probs)
df$quant <- factor(findInterval(df$x,quantiles))
figureS2c = ggplot(df, aes(x,y)) +
   ggtitle("Independent TU distance to nearest annotated feature") +
   geom_line() +
   geom_ribbon(aes(ymin=0, ymax=y, fill=quant)) +
   scale_x_continuous("Distance (bp)", breaks=quantiles, labels=round(10^quantiles, 1)) +
   scale_y_continuous("Density") +
   scale_fill_brewer(guide="none") +
   geom_text(data=as.data.table(quantiles, keep.rownames="label"), aes(x=round(quantiles, 1), y=-0.02, label=label), size=3) +
   theme_bw() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), strip.text=element_text(size=14), strip.background=element_rect(fill = NA, colour = NA))
save_and_plot(x=figureS2c, bname="Figures_and_Tables/FigureS2C", width=8, height=6)

# gg = ggplot(dt, aes(x=Distance, fill=Feature)) +
#    ggtitle("Intergenic features distance to other annotated") +
#    geom_histogram(bins=50, alpha=0.5, col="gray", position="identity") +
#    scale_x_log10("Distance (bp)") + scale_y_continuous("Count") +
#    scale_fill_tableau("Intergenic feature", palette="Nuriel Stone") +
#    theme_light() + theme(legend.position=c(0.9, 0.9))
# save_and_plot(x=gg, bname="Figures_and_Tables/Intergenic_features_distance_to_annotated", width=8, height=6)
# 
# dist = distanceToNearest(new.anno.genes[new.anno.genes$Set%in%"Intergenic"], new.anno.genes[!new.anno.genes$Set%in%"Intergenic"])
# 
# dt = data.table(Feature=new.anno.genes[new.anno.genes$Set%in%"Intergenic"]$Subset, Distance=mcols(dist)$distance)
# 
# gg = ggplot(dt, aes(x=Distance, fill=Feature)) +
#    ggtitle("Intergenic features distance to other annotated") +
#    geom_histogram(bins=50, alpha=0.5, col="gray", position="identity") +
#    scale_x_log10("Distance (bp)") + scale_y_continuous("Count") +
#    scale_fill_tableau("Intergenic feature", palette="Nuriel Stone") +
#    theme_light() + theme(legend.position=c(0.9, 0.9))
# save_and_plot(x=gg, bname="Figures_and_Tables/Intergenic_features_distance_to_discovered", width=8, height=6)

sel1 = data.table(Feature=new.anno.genes[new.anno.genes$Set%in%"Intergenic"]$gene_id, Distance=mcols(dist)$distance)

sel1 = sel1[Distance >= quantile(dt$Distance, prob=0.25),]
sel2 = new.anno.genes$Subset%in%c("< 10Kb", "> 10Kb")
new.anno.genes[new.anno.genes$gene_id%in%sel1[, Feature] & sel2,]$Subset = "newTranscriptionalUnit"
new.anno.genes = new.anno.genes[!grepl("Distal", new.anno.genes$gene_id) | new.anno.genes$gene_id%in%sel1[, Feature],]
```

![](./Figures_and_Tables/Intergenic_features_distance_to_annotated_quartiles.png)

```{bash extractSplicedReads, eval=FALSE}
for bam in `ls RNAseq_BAM/ | grep ^"RNAseq" | grep -v "_PE" | grep ".bam"$`; do
   if [[ ! -s spliced_BAM/"$bam" ]]; then
   echo $bam;
   samtools view -h RNAseq_BAM/"$bam" | awk '{if($0 ~ /^@/ || $6 ~ /N/) {print $0}}' | samtools view -Sb - > spliced_BAM/"$bam";
   fi;
done
   
for bam in `ls RNAseq_BAM/ | grep ^"RNAseq" | grep -v "_SE" | grep ".bam"$`; do
   if [[ ! -s spliced_BAM/"$bam" ]]; then
   echo $bam;
   samtools view -h RNAseq_BAM/"$bam" | awk '{if($0 ~ /^@/ || $6 ~ /N.*N/) {print $0}}' | samtools view -Sb - > spliced_BAM/"$bam";
   fi;
done
```

```{r splicingAnalysis}
if( REBUILD ){
   
   bamTable = gffTable[grepl("_(Untreated|WildType|Bcells|Unmodified-XRN2)_", name) & !grepl("_(UV)_", name)]
   files.bam = list.files("spliced_BAM", pattern="*.bam$", full.names=TRUE)
   files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
   names(files.bam) = bamTable[, group]
   files.bam = files.bam[sapply(files.bam, length)>0]
   
   lnc = new.anno.genes[new.anno.genes$Subset%in%c("long_ncRNA", "newTranscriptionalUnit")]
   
   counts = lapply(files.bam,
                   function(x){
                      print(x); flush(stdout())
                      if(as.character(x)%like%"_SE"){
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignments(file=x, param=param)
                      }else{
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                      }
                      reads = invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse"))
                      overlaps = as.data.table(findOverlaps(lnc, reads))
                      overlaps = overlaps[order(queryHits),][!duplicated(subjectHits),]
                      overlaps = overlaps[, .N, by="queryHits"]
                      dt = with(lnc[overlaps[, queryHits],], data.table(gene_id, Subset, Count=overlaps[, N], Total=length(reads)))
                      return(dt)
                   })
   names(counts) = names(files.bam)
   
   saveRDS(counts, file="Count_of_spliced_reads_on_intergenic_and_lncRNA.rds")
   
   pc = new.anno.genes[new.anno.genes$Subset%in%c("protein_coding", "UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene")]
   
   counts = lapply(files.bam,
                   function(x){
                      print(x); flush(stdout())
                      if(as.character(x)%like%"_SE"){
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignments(file=x, param=param)
                      }else{
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                      }
                      reads = invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse"))
                      overlaps = as.data.table(findOverlaps(pc, reads))
                      overlaps = overlaps[order(queryHits),][!duplicated(subjectHits),]
                      overlaps = overlaps[, .N, by="queryHits"]
                      dt = with(pc[overlaps[, queryHits],], data.table(gene_id, Subset, Count=overlaps[, N], Total=length(reads)))
                      return(dt)
                   })
   names(counts) = names(files.bam)
   
   saveRDS(counts, file="Count_of_spliced_reads_on_intergenic_and_protein-coding.rds")

}
```

```{r plotSplicing}
counts = c(readRDS("Count_of_spliced_reads_on_intergenic_and_protein-coding.rds"), readRDS("Count_of_spliced_reads_on_intergenic_and_lncRNA.rds"))

tab = rbindlist(lapply(counts, function(x) x[, .(sum(Count)/mean(Total)), by="Subset"]))
tab[Subset=="protein_coding", Subset := "Protein coding"][Subset=="UpstreamOfGene", Subset := "Upstream of gene"][Subset=="LinkerOfGene", Subset := "Linker of genes"]
tab[Subset=="long_ncRNA", Subset := "Long ncRNA"][Subset=="DownstreamOfGene", Subset := "Downstream of gene"][Subset=="newTranscriptionalUnit", Subset := "Independent TU"]
tab[, Subset := factor(Subset, levels=c("Protein coding", "Long ncRNA", "Upstream of gene", "Linker of genes", "Downstream of gene", "Independent TU"))]
figureS2d = ggplot(tab, aes(x=Subset, y=V1)) +
   ggtitle("Distribution of spliced reads per RNA-seq dataset") +
   geom_boxplot() + geom_jitter(pch=21, colour="black", fill="grey", width = 0.2) +
   scale_x_discrete("") +
   scale_y_continuous("Fraction of spliced reads", labels=percent) +
   theme_bw() + theme(legend.position=c(0.9, 0.1), axis.text = element_text(angle=30, hjust=1, size=12), axis.title=element_text(size=14), strip.text=element_text(size=14), strip.background=element_rect(fill = NA, colour = NA))
save_and_plot(x=figureS2d, bname="Figures_and_Tables/FigureS2D", width=8, height=6)
```

![](./Figures_and_Tables/figureS2d.png)

```{r biodallianceTrack, eval=FALSE}
if( REBUILD ){
   require("colorspace")
   
   col.dt = data.table(as.data.frame(new.anno.genes))
   
   col.ref = setNames(apply(t(col2rgb(tableau_color_pal(palette = "Classic 10 Medium")(length(factBiotype2)))), 1,
                            function(x)
                               paste(x, collapse=",")), factBiotype2)
   col.dt[, Colour := col.ref[Subset]]
   
   fwrite(col.dt, file="Annotation_track_colours.bed", sep="\t", row.names=FALSE, col.names=FALSE)
   
}

# FROM BASH
# LC_COLLATE=C; sort -k1,1 -k2,2n Annotation_track_colours.bed | awk 'BEGIN{IFS=OFS="\t"}{print $1, $2, $3, $6, 1000, $5, $2, $2, $NF}' > bed_for_BB.bed
# bedToBigBed -as=bed_hg38.as -type=bed9 bed_for_BB.bed /Users/agostif/Documents/GitHub/R_Gencode_Reference/hg38.chrom.sizes myBigBed.bb
```

```{r detailed_unannotated}
if( REBUILD ){
   bamTable = allTable[group%like%"HeLa" & (name%like%"Untreated" | name%like%"WildType") & !name%like%"pA"]
   files.bam = list.files("RNAseq_BAM", pattern="*.bam$", full.names=TRUE)
   files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
   names(files.bam) = bamTable[, group]
   
   nag = copy(new.anno.genes)
   nag$Subset = factor(nag$Subset, levels=c("protein_coding", "long_ncRNA", "ncRNA", "other annotated", "UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "newTranscriptionalUnit" ))
   nag = nag[order(nag$Subset),]
   
   counts = lapply(files.bam,
                   function(x){
                      print(x); flush(stdout())
                      if(as.character(x)%like%"_SE"){
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignments(file=x, param=param)
                      }else{
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                      }
                      reads = invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse"))
                      overlaps = as.data.table(findOverlaps(new.anno.genes, reads))
                      overlaps = overlaps[order(queryHits),][!duplicated(subjectHits),]
                      dt = data.table(table(new.anno.genes[overlaps[, queryHits],]$Subset), Total=length(reads))
                      setnames(dt, "V1", "Feature")
                      return(dt)
                   })
   names(counts) = names(files.bam)
   saveRDS(counts, file="Extended_annotation_counts_HeLa_Untreated_WildType_Fractions.rds")
}
```

```{r detailed_coverage, fig.align="center", fig.width=8, fig.height=4}
counts = readRDS("Extended_annotation_counts_HeLa_Untreated_WildType_Fractions.rds")

tab =  lapply(counts,
              function(x)
                 rbindlist(list(x, data.table(Feature = "Unassigned", N = x[,unique(Total)-sum(N)], Total = unique(x[, Total])))))

tab = rbindlist(tab, idcol="Sample")[, list(N = sum(N), Total = sum(Total)), by=c("Sample", "Feature")]

tab[, Sample := gsub("RNAseq_|_HeLa", "", Sample)]
tab[Feature%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene"), Feature := "Gene-associated TU"]
tab[Feature%in%c("newTranscriptionalUnit"), Feature := "Independent TU"]
tab[Feature=="protein_coding", Feature := "Protein coding"][Feature=="long_ncRNA", Feature := "Long ncRNA"][Feature=="ncRNA", Feature := "Small ncRNA"][Feature=="other annotated", Feature := "Other annotated"]
# tab = tab[Sample%in%c("Chromatin", "Nuclear"), list(N=sum(N), Total=unique(Total)), by=c("Sample", "Feature")]
tab = tab[, list(N=sum(N), Total=unique(Total)), by=c("Sample", "Feature")]
tab[, `:=`(Percentage = N/sum(N)), by="Sample"]

tab[, `:=`(Feature=factor(Feature, levels=rev(c(names(colSet)[6:5], rev(names(colSet)[1:4])))),
           # Sample=factor(Sample, levels=c("Chromatin", "Nuclear")))]
           Sample=factor(Sample, levels=c("Chromatin", "Nucleoplasm", "Nuclear", "Cytoplasm")))]

tab[Sample%in%c("Chromatin", "Nucleoplasm"), Fraction := factor("Nucleus", levels=c("Cell", "Nucleus"))]
tab[Sample%in%c("Nuclear", "Cytoplasm"), Fraction := "Cell"]

tab = tab[!is.na(Feature),]

# ggplot_grad_rects <- function(n, ymin, ymax, dt_facets=NULL) {
#    y_steps <- seq(from = ymin, to = ymax, length.out = n + 1)
#    alpha_steps <- seq(from = 1, to = 0, length.out = n)
#    rect_grad <- data.frame(ymin = y_steps[-(n + 1)], 
#                            ymax = y_steps[-1], 
#                            alpha = alpha_steps)
#    if( !is.null(dt_facets) )
#       rect_grad = merge(dt_facets, rect_grad)
#    return(rect_grad)
# }
# add_dt = ggplot_grad_rects(100, 0.12, 0.127, dt_facets=data.frame(unique(tab[, .(Fraction)])))

figure2A = ggplot() +
   # ggtitle("RNA-seq reads distribution in HeLa cells") +
   geom_bar(data=tab, aes(x=Sample, y=Percentage, fill=Feature),
            stat="identity", position="stack", col="black", width=0.75) +
   # geom_rect(data=add_dt, xmin=-1, xmax=3,
   #            aes(ymin=ymin, ymax=ymax, 
   #                alpha=1-alpha), fill="white") +
    # guides(alpha = FALSE) +
   scale_x_discrete("") +
   scale_y_continuous("Mapped reads", labels=percent) +
   scale_fill_manual("", values = colSet) +
   # scale_fill_manual("", values = c("slategrey", rev(tableau_color_pal(palette = "Classic 10 Medium")(9)[6:9]), rev(tableau_color_pal(palette = "Classic 10 Medium")(9)[1:5]))) +
   facet_wrap(~Fraction, scales="free_x") +
   theme_bw() +
   theme(legend.position="right", 
         axis.text=element_text(size=12), axis.title=element_text(size=14),
         axis.text.x=element_text(angle=45, hjust=1, vjust=1),
         strip.background=element_rect(fill = NA, colour = NA), strip.text=element_text(size=14)) +
   coord_cartesian(ylim=c(0, 0.12)) + guides(fill = FALSE)
# + guides(fill = guide_legend(ncol = 4, byrow = TRUE))
save_and_plot(x=figure2A, bname="Figures_and_Tables/Figure_2A", width=6, height=5)

# figure1A = figure1A + guides(fill = guide_legend(reverse=T))
# figure1A = figure1A + guides(fill = FALSE)

# gg1 = ggarrange(gg1.1, figure1A, ncol=2)
# 
# save_and_plot(x=gg1, bname="Figures_and_Tables/Figure_1AB", width=8, height=4)

# figure1A = figure1A + coord_flip() + theme(axis.text.x=element_text(angle=0, hjust=0.25, vjust=0))
# save_and_plot(x=figure2A, bname="Figures_and_Tables/Figure_2A", width=6, height=4)
```

![](./Figures_and_Tables/Figure_2A.png)

```{r detailed_unassigned, eval=FALSE}
bamTable = allTable[group%like%"HeLa" & (name%like%"Untreated" | name%like%"WildType") & !name%like%"pA"]
files.bam = list.files("RNAseq_BAM", pattern="*.bam$", full.names=TRUE)
files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
names(files.bam) = bamTable[, group]

nag = copy(new.anno.genes)
nag$Subset = factor(nag$Subset, levels=c("protein_coding", "long_ncRNA", "ncRNA", "other annotated", "UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "newTranscriptionalUnit" ))
nag = nag[order(nag$Subset),]

reads = lapply(setNames(files.bam, names(files.bam)),
                function(x){
                   print(x); flush(stdout())
                   if(as.character(x)%like%"_SE"){
                      param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                            isDuplicate=FALSE,
                                                            isSecondaryAlignment=FALSE),
                                           mapqFilter=255)             
                      reads = readGAlignments(file=x, param=param)
                   }else{
                      param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                            isDuplicate=FALSE,
                                                            isSecondaryAlignment=FALSE),
                                           mapqFilter=255)             
                      reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                   }
                   reads = as(invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse")), "GRanges")
                   reads = reads[!overlapsAny(reads, nag)]
                   return(reads)
                })

ggl = lapply(reads,
             function(x){
                
                a = as.data.table(distanceToNearest(as(x, "GRanges")))
                b = as.data.table(distanceToNearest(as(x, "GRanges"), genes))
                
                tab.dist = data.table(distanceToUnassigned = a[, distance], distanceToAnnotated = b[, distance])
                tab.dist[, `:=`(dTU = findInterval(distanceToUnassigned,
                                                   seq(0, max(distanceToUnassigned)+1e3, 1e3), all.inside=TRUE),
                                dTA = findInterval(distanceToAnnotated,
                                                   seq(0, max(distanceToAnnotated)+1e3, 1e3), all.inside=TRUE))]
                ggplot(tab.dist[, .N, by=c("dTU", "dTA")], aes(x=dTA*1e3, dTU*1e3, col=log10(N))) +
                   geom_point(size=0.5) +
                   scale_x_log10("Distance to nearest annotated feature") +
                   scale_y_log10("Distance to nearest unassigned read") +
                   scale_colour_viridis(option="magma", begin=0.1, end=0.9) +
                   theme_bw()
             })
gg = ggarrange(plotlist=ggl, ncol=4, nrow=2, legend="bottom", common.legend=TRUE)
```

```{r novel_nascent_expressed}
if( REBUILD ){
   bamTable = allTable[group%like%"HeLa" & (name%like%"Untreated" | name%like%"WildType") & !name%like%"pA"]
   files.bam = list.files("RNAseq_BAM", pattern="*.bam$", full.names=TRUE)
   files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
   names(files.bam) = bamTable[, group]
   
   nag = copy(new.anno.genes)
   nag$Subset = factor(nag$Subset, levels=c("protein_coding", "long_ncRNA", "ncRNA", "other annotated", "UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "newTranscriptionalUnit" ))
   nag = nag[order(nag$Subset),]
   
   counts = lapply(files.bam,
                   function(x){
                      print(x); flush(stdout())
                      if(as.character(x)%like%"_SE"){
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignments(file=x, param=param)
                      }else{
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                      }
                      reads = keepSeqlevels(reads, chromosomes, pruning.mode="coarse")
                      countsFS = countOverlapsUnion(new.anno.genes, reads)
                      reads = invertStrand(reads)
                      countsRS = countOverlapsUnion(new.anno.genes, reads)
                      dt = data.table(gene_id = new.anno.genes$gene_id, FS = countsFS, RS = countsRS)
                      return(dt)
                   })
   names(counts) = names(files.bam)
   ncounts = lapply(counts, function(x) with(x, data.table(gene_id, tpm = countToTPM(RS, new.anno.genes))))
   saveRDS(ncounts, file="Extended_annotation_TPM_HeLa_Untreated_WildType_Fractions.rds")
}
```

```{r plot_novel_nascent_expressed}
ncounts = readRDS("Extended_annotation_TPM_HeLa_Untreated_WildType_Fractions.rds")

ncounts = rbindlist(ncounts, idcol="Set")[, list(tpm=mean(tpm)), by=c("Set", "gene_id")]
ncounts[, Subset := new.anno.genes[match(gene_id, new.anno.genes$gene_id),]$Subset]

tmp = new.anno.genes[new.anno.genes$gene_id%in%ncounts[Subset=="LinkerOfGene", gene_id]]

upLoG = as.data.table(findOverlaps(flank(tmp, 1, start=TRUE, both=FALSE),
                                   new.anno.genes[new.anno.genes$gene_id%in%ncounts[Subset!="LinkerOfGene", gene_id]], select="first"))
upLoG = data.table(as.data.frame(with(upLoG, new.anno.genes[new.anno.genes$gene_id%in%ncounts[Subset!="LinkerOfGene", gene_id]][V1,])))
doLoG = as.data.table(findOverlaps(flank(tmp, 1, start=FALSE, both=FALSE),
                                   new.anno.genes[new.anno.genes$gene_id%in%ncounts[Subset!="LinkerOfGene", gene_id]], select="first"))
doLoG = data.table(as.data.frame(with(doLoG, new.anno.genes[new.anno.genes$gene_id%in%ncounts[Subset!="LinkerOfGene", gene_id]][V1,])))

dt = data.table(upstream=ncounts[match(upLoG[, gene_id], ncounts[, gene_id]), tpm],
                LoG=ncounts[match(tmp$gene_id, ncounts[, gene_id]), tpm],
                downstream=ncounts[match(doLoG[, gene_id], ncounts[, gene_id]), tpm],
                upType = upLoG[, Subset], doType = doLoG[, Subset],
                upWidth = upLoG[, width], LoG=width(tmp), doWidth=doLoG[, width])
dt = dt[LoG>=1 & upType=="protein_coding" & doType=="protein_coding",]
gg1 = ggplot(melt.data.table(dt[, 1:3, with=FALSE]), aes(x=variable, y=log2(value+1))) + geom_boxplot() + scale_y_continuous("log2(TPM expression)") + scale_x_discrete("", breaks=colnames(dt[, 1:3, with=FALSE]), labels=c("Upstream\nprotein-coding", "LinkerOfGene", "Downstream\nprotein-coding"))
gg2 = ggplot(melt.data.table(dt[, 6:8, with=FALSE]), aes(x=variable, y=value)) + geom_boxplot() + scale_y_log10("Feature width") + scale_x_discrete("", breaks=colnames(dt[, 6:8, with=FALSE]), labels=c("Upstream\nprotein-coding", "LinkerOfGene", "Downstream\nprotein-coding"))
gg = ggarrange(gg1, gg2, ncol=2)
save_and_plot(x=gg, bname="Figures_and_Tables/LoG_upstream_downstream", width=8, height=6)

ncounts[, Set := tstrsplit(Set, "_")[[2]]]
ncounts_nuc = copy(ncounts)

# ncounts[Subset%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene"), Subset := "Nascent TU"]
ncounts[Subset%in%c("newTranscriptionalUnit"), Subset := "Independent TU"]
ncounts[Subset=="protein_coding", Subset := "Protein coding"][Subset=="long_ncRNA", Subset := "Long ncRNA"][Subset=="ncRNA", Subset := "Small ncRNA"][Subset=="other annotated", Subset := "Other annotated"]
ncounts[, `:=`(Subset=factor(Subset, levels=names(colSet2)),
               Set=factor(Set, levels=c("Chromatin", "Nucleoplasm")))]

tab = ncounts[Set%in%c("Chromatin", "Nucleoplasm"),][tpm>0, list(N=length(unique(gene_id))), by=c("Set", "Subset")][order(Subset)]
# tab = test[tpm>0, list(N=length(unique(gene_id))), by="Subset"][order(Subset)]

figure2E = ggplot(ncounts[Set%in%c("Chromatin", "Nucleoplasm"),][tpm>0,], aes(x=Subset, y=log2(tpm), fill=Subset, colour=Subset)) +
   geom_hline(yintercept=0, lty=2, lwd=0.5, colour="grey") +
   # ggtitle("Expression in HeLa cells") +
   geom_violin(position="dodge", col="black", alpha=0.5) +
   geom_boxplot(notch=TRUE, width=0.1, position=position_dodge(width=0.9), outlier.colour=NA, fill="white") +
   geom_text(data=tab, aes(label=N, x=Subset, y=-12),
             position=position_dodge(width=0.9), inherit.aes=FALSE, colour="black", size=3) +
   scale_x_discrete("") + scale_y_continuous("Normalised expression [ log2(TPM) ]") +
   scale_colour_manual("", values=colSet2) +
   scale_fill_manual("", values=colSet2) +
   facet_grid(~Set) +
   theme_bw() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14),
         axis.text.x=element_text(angle=45, hjust=1, vjust=1),
         strip.text=element_text(size=14), strip.background=element_rect(fill = NA, colour = NA)) +
   guides(fill=FALSE, colour=FALSE)
save_and_plot(x=figure2E, bname="Figures_and_Tables/Figure_2E", width=7, height=5)

expressed_gene_ids = ncounts[Set=="Chromatin" & tpm>=min.expr, unique(gene_id)]

tab = ncounts[Set%in%c("Chromatin", "Nucleoplasm"),][tpm>min.expr, list(N=length(unique(gene_id))), by=c("Set", "Subset")][order(Subset)]

figure2E_bis = ggplot(ncounts[Set%in%c("Chromatin", "Nucleoplasm"),][tpm>min.expr,], aes(x=Subset, y=log2(tpm), fill=Subset, colour=Subset)) +
   # ggtitle("Expression in HeLa cells") +
   geom_violin(position="dodge", col="black", alpha=0.5, trim=TRUE) +
   geom_boxplot(notch=TRUE, width=0.1, position=position_dodge(width=0.9), outlier.shape=20, fill="white") +
   geom_text(data=tab, aes(label=N, x=Subset, y=15),
             position=position_dodge(width=0.9), inherit.aes=FALSE, colour="black", size=3) +
   scale_x_discrete("") + scale_y_continuous("Normalised expression [ log2(TPM) ]") +
   scale_colour_manual("", values=rep("grey30", 8)) +
   scale_fill_manual("", values=colSet2) +
   facet_grid(~Set) +
   theme_bw() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14),
         axis.text.x=element_text(angle=45, hjust=1, vjust=1),
         strip.text=element_text(size=14), strip.background=element_rect(fill = NA, colour = NA)) +
   guides(fill=FALSE, colour=FALSE)
save_and_plot(x=figure2E_bis, bname="Figures_and_Tables/Figure_2E_bis", width=7, height=5)
```

![](./Figures_and_Tables/Figure_2E.png)

```{r plot_novel_nascent_expressed_width}
dt = data.table(as.data.frame(new.anno.genes))

dt[, tpm := ncounts[Set%in%"Nuclear",][match(gene_id, ncounts[Set%in%"Nuclear", gene_id]), tpm]]

# dt[Subset%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene"), Subset := "Nascent TU"]
dt[Subset%in%c("newTranscriptionalUnit"), Subset := "Independent TU"]
dt[Subset=="protein_coding", Subset := "Protein coding"][Subset=="long_ncRNA", Subset := "Long ncRNA"][Subset=="ncRNA", Subset := "Small ncRNA"][Subset=="other annotated", Subset := "Other annotated"]
dt[, `:=`(Subset=factor(Subset, levels=names(colSet2)))]

tab = dt[gene_id%in%expressed_gene_ids, .N, by="Subset"][order(Subset)]

dt[, Subset_fill := Subset]
dt[Subset%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene"), Subset_fill := "Gene-associated TU"]
dt[, Subset_fill := factor(Subset_fill, levels=names(colSet))]

figure2D = ggplot(dt, aes(x=Subset, y=width, fill=Subset_fill)) +
   # ggtitle("Features width") +
   geom_boxplot(notch=TRUE, outlier.colour=NA) +
   # geom_text(data=tab, aes(label=N, x=Subset, y=1), inherit.aes=FALSE, colour="black", size=3) +
   scale_x_discrete("") + scale_y_log10("Expressed feature width (bp)", breaks=10^c(0:6), labels=comma) +
   scale_fill_manual("", values=colSet) +
   facet_wrap(~as.factor(1)) +
   theme_light() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14),
                         axis.text.x=element_text(angle=45, hjust=1, vjust=1),
                         strip.text=element_text(colour="white", size=NA), strip.background=element_rect(fill = NA, colour = NA))
# + guides(fill=FALSE)
save_and_plot(x=figure2D, bname="Figures_and_Tables/Figure_2D", width=6, height=5)
```

![](./Figures_and_Tables/Figure_2D.png)

```{r genomicCoverage}
dt = copy(new.anno.genes)
# dt$Subset[dt$Subset%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene")] = "Nascent TU"
dt$Subset[dt$Subset%in%c("newTranscriptionalUnit")] = "Independent TU"
dt$Subset[dt$Subset=="protein_coding"]= "Protein coding"
dt$Subset[dt$Subset=="long_ncRNA"] = "Long ncRNA"
dt$Subset[dt$Subset=="ncRNA"] = "Small ncRNA"
dt$Subset[dt$Subset=="other annotated"] = "Other annotated"
dt$Subset = factor(dt$Subset, levels=names(colSet2))

tab = data.table(as.data.frame(dt))[gene_id%in%expressed_gene_ids, .N, by="Subset"][order(Subset)]

figure2B = ggplot(tab, aes(x=Subset, y=N, fill=Subset)) +
   # ggtitle("Annotated and unannotated features") +
   geom_bar(stat="identity", position="identity", colour="black") +
   geom_text(data=tab, aes(label=N, x=Subset, y=N+250),
             position=position_dodge(width=0.9), inherit.aes=FALSE, colour="black", size=3) +
   scale_x_discrete("") +
   scale_y_continuous("Expressed feature count", labels=comma) +
   scale_fill_manual(values=colSet2, guide=FALSE) +
   facet_wrap(~as.factor(1)) +
   theme_bw() +
   theme(legend.position=c(0.8, 0.85),
         axis.text=element_text(size=12), axis.title=element_text(size=14),
         axis.text.x=element_text(angle=45, hjust=1, vjust=1),
         strip.text=element_text(colour="white", size=NA), strip.background=element_rect(fill = NA, colour = NA))
save_and_plot(x=figure2B, bname="Figures_and_Tables/Figure_2B", width=6, height=5)

dt = rbindlist(lapply(reduce(split(dt, dt$Subset), ignore.strand=TRUE),
                 function(x)
                    data.table(N=sum(width(x)))), idcol="Feature")
genome.length = data.table(as.data.frame(seqinfo(genes)), keep.rownames="seqnames")[seqnames%in%chromosomes, sum(seqlengths)]

dt[, Percentage := N/genome.length]
dt[, Feature := factor(Feature, levels=names(colSet2))]

figure2C = ggplot(dt, aes(x=Feature, y=Percentage, fill=Feature)) +
   # ggtitle("Genomic coverage") +
   geom_bar(stat="identity", position="stack", colour="black") +
   scale_x_discrete("") +
   scale_y_continuous("Genomic coverage", labels=percent) +
   scale_fill_manual(values=colSet2, guide=FALSE) +
   geom_text(aes(label=paste0(round(Percentage*100, 2), "%")), y=0.005, size=3) +
   facet_wrap(~as.factor(1)) +
   theme_bw() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14),
                      axis.text.x=element_text(angle=45, hjust=1, vjust=1),
                      strip.text=element_text(colour="white", size=NA), strip.background=element_rect(fill = NA, colour = NA)) + coord_cartesian(ylim=c(0,0.15))
save_and_plot(x=figure2C, bname="Figures_and_Tables/Figure_2C", width=6, height=5)
```

![](./Figures_and_Tables/Figure_2B.png)

![](./Figures_and_Tables/Figure_2C.png)

```{r Figure2}
figure2 = ggarrange(ggarrange(figure2A, figure2C, align = 'h', ncol=2, nrow=1, widths=c(4/7, 3/7), labels=c("A", "B")),
                    ggarrange(figure2E, labels=c("C")),
                    ggarrange(figure2B, figure2D, align = 'h', ncol=2, nrow=1, labels=c("D", "E")),
                    ncol=1, nrow=3)

save_and_plot(x=figure2, bname="Figures_and_Tables/Figure2", width=12, height=16, plot=FALSE)
```

![](./Figures_and_Tables/Figure2.png)

```{r plotTrackFunctions}
theme_blank <- function(...) {
  ret <- theme_bw(...)
  ret$line <- element_blank()
  ret$rect <- element_blank()
  ret$strip.text <- element_blank()
  ret$axis.text <- element_blank()
  ret$plot.title <- element_blank()
  ret$axis.title <- element_blank()
  # ret$plot.margin <- structure(c(0, 0, -1, -1), unit = "lines", valid.unit = 3L, class = "unit")
  ret
}

createAnnoTrack <- function(TxDb=NULL, region, min.unit=100, species, genome_assembly, anno_version, rosetta_stone){

   message("Check the annotation...")
   if( is.null(TxDb) ){
      message("Annotation not found. Download GFF...")
      TxDb = getTxDb(species, genome_assembly, anno_version)
   }
   
   message("Extract features...")
   genes = genes(TxDb)
   genes = genes[!grepl("PAR", names(genes)),]
   
   if( base::exists("rosetta_stone") )
      genes$gene_name = rosetta_stone[match(genes$gene_id, rosetta_stone$gene_id),]$gene_name

   exons = exonsBy(TxDb, by="gene")
   exons = exons[!grepl("PAR", names(exons)),]
   
   genes = genes[overlapsAny(genes, region),]
   
   genes = genes[order(width(genes), decreasing=TRUE),]
   
   exons = reduce(exons[names(exons)%in%names(genes)], min.gapwidth=min.unit/10)
   
   message("Apply layering...")
   genes$Y = as.numeric(NA)
   exons = lapply(exons,
                  function(x){
                     x$Y = as.numeric(NA);
                     x
                  })
   
   if( !identical(names(genes), names(exons)) )
      exons = exons[match(names(genes), names(exons))]

   level = 0
   for( strand in c("+", "-") ){
      
      strand_sel = genes[strand(genes)==strand,]$gene_id
      overlaps = as.data.table(findOverlaps(genes[strand_sel]))[!queryHits==subjectHits,]
      
      if( length(which(!seq_along(exons[strand_sel])%in%overlaps[, queryHits])) > 0 )
         genes[strand_sel][which(!seq_along(exons[strand_sel])%in%overlaps[, queryHits]),]$Y = level
      for( i in which(!seq_along(exons[strand_sel])%in%overlaps[, queryHits]) )
         exons[strand_sel][[i]]$Y = level 
      
      while( nrow(overlaps)>0 ){
         temp = copy(overlaps)
         temp = split(temp, temp[,queryHits])
         
         for( i in names(temp) ){
            if( i%in%names(temp) ){
               temp = temp[!names(temp)%in%as.character(temp[[i]][, subjectHits])]
            }
         }
         genes[strand_sel][as.numeric(names(temp)),]$Y = level
         for( i in as.numeric(names(temp)) )
            exons[strand_sel][[i]]$Y = level
         
         overlaps = overlaps[!queryHits%in%as.numeric(names(temp)),]
         level = level-0.25
      }
      level = level-0.25
   }
   
   message("Create track data...")
   myList = list()
   myList[["Genes"]] = data.table(as.data.frame(genes), key="gene_id")
   myList[["Exons"]] = rbindlist(lapply(exons, as.data.table), idcol="gene_id")
   
   arrows.pl = genes[width(genes)>min.unit,]
   arrows.pl = split(arrows.pl, names(arrows.pl))
   arrows.pl = lapply(arrows.pl, function(x) data.table(as.data.frame(tile(x, width=min.unit))))
   arrows.pl = rbindlist(arrows.pl, idcol="gene_id")
   setkey(arrows.pl, gene_id)
   myList[["Arrows"]] = arrows.pl[myList[["Genes"]], Y := i.Y]
   
   myList[["Genes"]] = myList[["Genes"]][!is.na(Y)]
   myList[["Exons"]] = myList[["Exons"]][!is.na(Y)]
   myList[["Arrows"]] = myList[["Arrows"]][!is.na(Y)]
   
   message("Finished")
   return(myList)
}

plotanno <- function(data=list(), region=NULL, gene.height = 1, colorVar="grey30", fillVar=tableau_color_pal(palette="Tableau 10")(1), textSize=2, clean=FALSE, outPlot=FALSE){
   plot <- ggplot() +
      geom_segment(data=data[["Genes"]][strand=="+"], mapping=aes(x=start, xend=end, y=Y, yend=Y), colour=colorVar, lwd=1) +
      geom_segment(data=data[["Genes"]][strand=="-"], mapping=aes(x=end, xend=start, y=Y, yend=Y), colour=colorVar, lwd=1) +
      geom_segment(data=data[["Exons"]][strand=="+", .(start=min(start)), by=c("gene_id", "Y")], mapping=aes(x=start, xend=start, y=Y, yend=Y+(0.2*gene.height)/2), colour=colorVar, lwd=0.25) +
      geom_segment(data=data[["Exons"]][strand=="+", .(start=min(start)), by=c("gene_id", "Y")], mapping=aes(x=start, xend=start+(width(region)/50), y=Y+(0.2*gene.height)/2, yend=Y+(0.2*gene.height)/2), colour=colorVar, lwd=0.25, arrow = arrow(length = unit(0.1, "cm"), type="closed")) +
      geom_segment(data=data[["Exons"]][strand=="-", .(end=max(end)), by=c("gene_id", "Y")], mapping=aes(x=end, xend=end, y=Y, yend=Y-(0.2*gene.height)/2), colour=colorVar, lwd=0.25) +
      geom_segment(data=data[["Exons"]][strand=="-", .(end=max(end)), by=c("gene_id", "Y")], mapping=aes(x=end, xend=end-(width(region)/50), y=Y-(0.2*gene.height)/2, yend=Y-(0.2*gene.height)/2), colour=colorVar, lwd=0.25, arrow = arrow(length = unit(0.1, "cm"), type="closed")) +
      geom_segment(data=data[["Genes"]], mapping=aes(x=start, xend=start, y=Y-(0.05*gene.height)/2, yend=Y+(0.05*gene.height)/2), colour=colorVar, lineend="square") +
      geom_segment(data=data[["Genes"]], mapping=aes(x=end, xend=end, y=Y-(0.05*gene.height)/2, yend=Y+(0.05*gene.height)/2), colour=colorVar, lineend="square") +
      geom_rect(data=data[["Exons"]], mapping=aes(xmin=start, xmax=end, ymin=Y-(0.05*gene.height), ymax=Y+(0.05*gene.height)), fill=fillVar, colour=colorVar, lwd=0.25) +
      geom_text(data=data[["Genes"]][strand=="+"], mapping=aes(x=(start+end)/2, y=Y-(0.1*gene.height), label=gene_name), colour=colorVar, size=textSize) +
      geom_text(data=data[["Genes"]][strand=="-"], mapping=aes(x=(start+end)/2, y=Y+(0.1*gene.height), label=gene_name), colour=colorVar, size=textSize) +
      scale_x_continuous(position = "top") +
      coord_cartesian(xlim=range(as.data.table(region)[, list(start, end)]), ylim=data[["Genes"]][, c(min(Y)-0.2, max(Y)+0.2)], expand=FALSE) +
      theme_few() + theme(axis.title.x=element_blank(), axis.text.y=element_blank(), axis.title.y=element_blank(), axis.ticks.y=element_blank()) + guides(fill=FALSE, colour=FALSE) 
   if( clean )
      plot = plot + theme_blank()
   if( outPlot )
      print(plot)
   return(plot)
}

require("zoo")

createTracks <- function(cov_pos, cov_neg=NULL, track_order=NULL, region, fillVar=c("#d62728", "#1f77b4"), useSmooth=TRUE, smoothWin=101){
   
   if( !is.null(cov_neg) ){
      if( !identical(names(cov_pos), names(cov_neg)) ){
         warning("The positive and negative coverage names do not match!")
         return(1)
      }
   }
   
   if( !is.null(track_order) ){
      if( !all(track_order%in%names(cov_pos)) ){
         warning("The track order do not match the positive coverage names!")
         return(1)
      }
      cov_pos = cov_pos[track_order]
      if( !is.null(cov_neg) ){
         if( !all(track_order%in%names(cov_pos)) ){
            warning("The track order do not match the negative coverage names!")
            return(1)
         }
         cov_neg = cov_neg[track_order]
      }
   }

   message("Importing files...")
   cov_pos = lapply(cov_pos,
                    function(x)
                       data.table(as.data.frame(import(x[1],
                                                       which=resize(region,
                                                                    width=width(region)+width(region)/20, fix="center")))))
   cov_pos = lapply(cov_pos,
                    function(x){
                       tmp = x[, .(pos = seq(start, end, 1), score), by="start"][, .(pos, score)]
                       tmp = tmp[pos>=(start(region)-width(region)/10) & pos<=(end(region)+width(region)/10),]
                       return(tmp)
                    })
   
   if( !is.null(cov_neg) ){
      cov_neg = lapply(cov_neg,
                       function(x)
                          data.table(as.data.frame(import(x[1],
                                                          which=resize(region,
                                                                       width=width(region)+width(region)/20, fix="center")))))
      cov_neg = lapply(cov_neg,
                       function(x){
                          tmp = x[, .(pos = seq(start, end, 1), score=-score), by="start"][, .(pos, score)]
                          tmp = tmp[pos>=(start(region)-width(region)/10) & pos<=(end(region)+width(region)/10),]
                          return(tmp)
                       })
   }
   
   message("Create plot tracks...")
   ggl = lapply(seq_along(cov_pos),
                function(i){
                   gg = ggplot()
                   if( useSmooth ){
                      if( !is.null(cov_neg) ){
                         gg = gg +
                            geom_hline(yintercept=0, linetype=2, col="grey30", lwd=0.25) +
                            geom_area(data=cov_neg[[i]], aes(x=pos, y=rollmean(score, smoothWin, fill=0)),
                                             col=fillVar[2], fill=fillVar[2], position="identity", stat="identity", alpha=0.5)
                      }
                      gg = gg + geom_area(data=cov_pos[[i]], aes(x=pos, y=rollmean(score, smoothWin, fill=0)),
                                          col=fillVar[1], fill=fillVar[1], position="identity", stat="identity", alpha=0.5)
                   }else{
                      if( !is.null(cov_neg) ){
                         gg = gg +
                            geom_hline(yintercept=0, linetype=2, col="grey30", lwd=0.25) +
                            stat_summary_bin(data=cov_neg[[i]],
                                                    aes(x=pos, y=score), bins=1000, fun.y="mean", geom="area",
                                                    col=fillVar[2], fill=fillVar[2], alpha=0.5, lwd=0.5)
                      }
                      gg = gg + stat_summary_bin(data=cov_pos[[i]],
                                                 aes(x=pos, y=score), bins=1000, fun.y="mean", geom="area",
                                                 col=fillVar[1], fill=fillVar[1], alpha=0.5, lwd=0.5)
                   }
                   y_min = layer_scales(gg)$y$range$range[1]
                   y_max = layer_scales(gg)$y$range$range[2]
                   y_gap = (y_max - y_min)/10
                   gg = gg + 
                      scale_y_continuous("Coverage") +
                      coord_cartesian(xlim=range(as.data.table(region)[, list(start, end)]),
                                               ylim=c(y_min-y_gap, y_max+y_gap),
                                               expand=FALSE) +
                      annotate("text", x=Inf, y=Inf,
                               label=paste0(as.character(seqnames(region)), ":", start(region),"-", end(region)),
                                            hjust=1.1, vjust=1.2, parse=TRUE) +
                      # annotate("text", x = Inf, y = Inf, label = names(cov_pos)[[i]], hjust = 1.1, vjust = 1.2, parse = TRUE) + 
                      theme_few() + theme(axis.text.x=element_blank(), axis.title.x=element_blank(), axis.ticks.x=element_blank()) +
                      guides(fill=FALSE, colour=FALSE)
                   return(gg)
                })
   
   # print(plot_grid(plotlist=ggl, nrow=length(ggl), axis="lr", align="v", rel_heights=rep(1,length(ggl))))
   
   message("Finished!")
   return(ggl)
}
```

```{r plotPreTracks}
cov_pos = lapply(setNames(c("Chromatin", "Nucleoplasm", "Nuclear", "Cytoplasm"), c("Chromatin", "Nucleoplasm", "Nuclear", "Cytoplasm")),
                 function(x)
                    grep("FWD", 
                         grep("Untreated|WildType",
                              grep(x, list.files("RPGC_bw", pattern="*HeLa*", full.names=TRUE),
                                   value=TRUE),
                              value=TRUE),
                         value=TRUE))
cov_neg = lapply(setNames(c("Chromatin", "Nucleoplasm", "Nuclear", "Cytoplasm"), c("Chromatin", "Nucleoplasm", "Nuclear", "Cytoplasm")),
                 function(x)
                    grep("REV",
                         grep("Untreated|WildType",
                              grep(x, list.files("RPGC_bw", pattern="*HeLa*", full.names=TRUE),
                                   value=TRUE),
                              value=TRUE),
                         value=TRUE))
```

```{r loadTxDb}
anno_version = "27"
genome_assembly = "hg38"

if( file.exists(paste(genome_assembly, anno_version, "TxDb.db", sep="_")) ){
   TxDb = loadDb(paste(genome_assembly, anno_version, "TxDb.db", sep="_"))
}else{
   TxDb = makeTxDbFromGFF(file = file.path(working.folder, paste0("gencode.v", anno_version, ".annotation.gff3.gz")),
                          format = "gff3",
                          dataSource = paste("Gencode version", anno_version),
                          organism = gsub("_", " ", species))
   saveDb(TxDb, paste(genome_assembly, anno_version, "TxDb.db", sep="_"))
}
```

```{r plotUoG}
region_select = GRanges("chr19", IRanges(start=46855000, end=47010000))
region_select_dt = as.data.table(region_select)
anno.pl = createAnnoTrack(TxDb, region_select, min.unit=100, species, genome_assembly, anno_version, gene.metadata)

pl = plotanno(anno.pl, region_select, textSize=3)
pl = pl + theme(panel.border = element_blank(),
                axis.line.x.top=element_line(size=0.25, arrow=arrow(length = unit(0.05, "inches"), ends = "both", type = "open")))

new_select = data.table(as.data.frame(subsetByOverlaps(new.anno.genes[!new.anno.genes$Set=="Annotated",], region_select)))
pl = pl + geom_rect(data=new_select, aes(xmin=start, xmax=end-1, ymin=0-(0.05*1), ymax=0+(0.05*1)), fill=colSet[["Gene-associated TU"]], colour="grey30", lwd=0.25)

ggl = createTracks(cov_pos=cov_pos, cov_neg=cov_neg, track_order=c("Cytoplasm", "Nucleoplasm", "Nuclear", "Chromatin"),
                   region=region_select, useSmooth=FALSE)

ggUoG = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))

ggl = createTracks(cov_pos=cov_pos["Nuclear"], cov_neg=cov_neg["Nuclear"], track_order="Nuclear",
                   region=region_select, useSmooth=FALSE)

ggUoG_nuc = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))
```

```{r plotDoG}
region_select = GRanges("chr4", IRanges(start=173369000, end=173480000))
# region_select = GRanges("chr15", IRanges(start=43825311, end=43905584))
# region_select = GRanges("chrX", IRanges(start=13732266, end=13785709))
region_select_dt = as.data.table(region_select)
anno.pl = createAnnoTrack(TxDb, region_select, min.unit=100, species, genome_assembly, anno_version, gene.metadata)

pl = plotanno(anno.pl, region_select, textSize=3)
pl = pl + theme(panel.border = element_blank(),
                axis.line.x.top=element_line(size=0.25, arrow=arrow(length = unit(0.05, "inches"), ends = "both", type = "open")))

new_select = data.table(as.data.frame(subsetByOverlaps(new.anno.genes[!new.anno.genes$Set=="Annotated",], region_select)))
pl = pl + geom_rect(data=new_select, aes(xmin=start, xmax=end-1, ymin=0-(0.05*1), ymax=0+(0.05*1)), fill=colSet[["Gene-associated TU"]], colour="grey30", lwd=0.25)

ggl = createTracks(cov_pos=cov_pos, cov_neg=cov_neg, track_order=c("Cytoplasm", "Nucleoplasm", "Nuclear", "Chromatin"),
                   region=region_select, useSmooth=FALSE)

ggDoG = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))

ggl = createTracks(cov_pos=cov_pos["Nuclear"], cov_neg=cov_neg["Nuclear"], track_order="Nuclear",
                   region=region_select, useSmooth=FALSE)

ggDoG_nuc = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))
```

```{r plotLoG}
# region_select = GRanges("chr3", IRanges(start=38345000, end=38495000))
region_select = GRanges("chr12", IRanges(start=20346000, end=20778000))
region_select_dt = as.data.table(region_select)
anno.pl = createAnnoTrack(TxDb, region_select, min.unit=100, species, genome_assembly, anno_version, gene.metadata)

pl = plotanno(anno.pl, region_select, textSize=3)
pl = pl + theme(panel.border = element_blank(),
                axis.line.x.top=element_line(size=0.25, arrow=arrow(length = unit(0.05, "inches"), ends = "both", type = "open")))

new_select = data.table(as.data.frame(subsetByOverlaps(new.anno.genes[!new.anno.genes$Set=="Annotated",], region_select)))
new_select = new_select[Subset=="LinkerOfGene"] # discard the Downstream of gene
pl = pl + geom_rect(data=new_select, aes(xmin=start, xmax=end-1, ymin=0-(0.05*1), ymax=0+(0.05*1)), fill=colSet[["Gene-associated TU"]], colour="grey30", lwd=0.25)

ggl = createTracks(cov_pos=cov_pos, cov_neg=cov_neg, track_order=c("Cytoplasm", "Nucleoplasm", "Nuclear", "Chromatin"), region=region_select, useSmooth=FALSE)

ggLoG = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))

ggl = createTracks(cov_pos=cov_pos["Nuclear"], cov_neg=cov_neg["Nuclear"], track_order="Nuclear",
                   region=region_select, useSmooth=FALSE)

ggLoG_nuc = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))
```

```{r plotNew}
region_select = GRanges("chr2", IRanges(start=81458000, end=81710000))
# region_select = GRanges("chr4", IRanges(start=24308839, end=24407554))
# region_select = GRanges("chr6", IRanges(start=134328826, end=134420492))
region_select_dt = as.data.table(region_select)
anno.pl = createAnnoTrack(TxDb, region_select, min.unit=100, species, genome_assembly, anno_version, gene.metadata)

pl = plotanno(anno.pl, region_select, textSize=3)
pl = pl + theme(panel.border = element_blank(),
                axis.line.x.top=element_line(size=0.25, arrow=arrow(length = unit(0.05, "inches"), ends = "both", type = "open")))

new_select = data.table(as.data.frame(reduce(subsetByOverlaps(new.anno.genes[!new.anno.genes$Set=="Annotated",], region_select), min.gapwidth=5e3)))
pl = pl + geom_rect(data=new_select, aes(xmin=start, xmax=end-1, ymin=0-(0.05*1), ymax=0+(0.05*1)), fill=colSet[["Independent TU"]], colour="grey30", lwd=0.25)

# Fix the Y axis limits
pl = pl + coord_cartesian(xlim=range(region_select_dt[, list(start, end)]), ylim=c(-0.5, 0+0.25), expand=FALSE)

ggl = createTracks(cov_pos=cov_pos, cov_neg=cov_neg, track_order=c("Cytoplasm", "Nucleoplasm", "Nuclear", "Chromatin"),
                   region=region_select, useSmooth=FALSE)

ggNew = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))

ggl = createTracks(cov_pos=cov_pos["Nuclear"], cov_neg=cov_neg["Nuclear"], track_order="Nuclear",
                   region=region_select, useSmooth=FALSE)

ggNew_nuc = plot_grid(plotlist=c(ggl, list(pl)), nrow=length(ggl)+1, axis="lr", align="v", rel_heights=c(rep(1, length(ggl)), 1.5))
```

```{r plotAll}
figure2.2 = ggarrange(ggUoG, ggDoG, ggLoG, ggNew, ncol=2, nrow=2, labels=c("G", "H", "I", "L"))
save_and_plot(x=figure2.2, bname="Figures_and_Tables/Figure_2.2", width=16, height=12)
```

![](./Figures_and_Tables/Figure_2.2.png)

```{r figure2_alt}
figure2 = ggarrange(ggarrange(ggUoG_nuc, ggLoG_nuc, ggDoG_nuc, ggNew_nuc, ncol=4, nrow=1, align = 'h', labels="B"),
                    ggarrange(figure2B, figure2D, align = 'h', ncol=2, nrow=1, labels=c("C", "D")),
                    ggarrange(figure2A, figure2E, align = 'h', ncol=2, nrow=1, widths=c(3/7, 4/7), labels=c("E", "F")),
                    ncol=1, nrow=3, heights=c(0.5, 1, 1))
save_and_plot(x=figure2, bname="Figures_and_Tables/Figure_2_alternative", width=16, height=16)

figure2 = ggarrange(ggarrange(ggUoG_nuc, ggLoG_nuc, ggDoG_nuc, ggNew_nuc, ncol=4, nrow=1, align = 'h', labels="B"),
                    ggarrange(figure2A, figure2E_bis, align = 'h', ncol=2, nrow=1, widths=c(3/7, 4/7), labels=c("C", "D")),
                    ggarrange(figure2B, figure2D, align = 'h', ncol=2, nrow=1, labels=c("E", "F")),
                    ncol=1, nrow=3, heights=c(0.5, 1, 1))
save_and_plot(x=figure2, bname="Figures_and_Tables/Figure_2_alternative2", width=16, height=16)
ggsave(figure2, filename="Figures_and_Tables/Figure_2_alternative2.eps", width=16, height=16, device=cairo_ps)
```

![](./Figures_and_Tables/Figure_2_alternative.png)

![](./Figures_and_Tables/Figure_2_alternative2.png)

```{r distanceToAnnotated}
dt1 = new.anno.genes[new.anno.genes$Subset%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene")]
dt2 = new.anno.genes[new.anno.genes$Set%in%c("Annotated")]

overlaps = as.data.table(distanceToNearest(dt1,dt2))

tab = with(overlaps, data.table(New=dt1[queryHits,]$Subset, Old=dt2[subjectHits,]$Subset))[, .N, by="Old"]
tab[Old=="protein_coding", Old := "Protein coding"]
tab[Old=="long_ncRNA", Old := "Long ncRNA"]
tab[Old=="ncRNA", Old := "Small ncRNA"]
tab[Old=="other annotated", Old := "Other annotated"]

tab[, Old := factor(Old, levels=c("Protein coding", "Long ncRNA", "Small ncRNA", "Other annotated"))]

figureS2b = ggplot(tab, aes(x=Old, y=N)) +
   ggtitle("Gene-associated TU nearest annotated feature") +
   geom_bar(stat="identity", position="stack", fill="grey", colour="black") +
   scale_x_discrete("") +
   scale_y_continuous("Count", labels=comma) +
   facet_wrap(~as.factor(1)) +
   theme_bw() +
   theme(legend.position=c(0.8, 0.85),
         axis.text=element_text(size=12), axis.title=element_text(size=14),
         axis.text.x=element_text(angle=45, hjust=1, vjust=1),
                      strip.text=element_text(colour="white", size=14), strip.background=element_rect(fill = NA, colour = NA))
save_and_plot(x=figureS2b, bname="Figures_and_Tables/Figure_S2B", width=6, height=5)


dt1 = new.anno.genes[new.anno.genes$Subset%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene", "newTranscriptionalUnit")]
dt2 = new.anno.genes[new.anno.genes$Set%in%c("Annotated")]

overlaps = as.data.table(distanceToNearest(dt1,dt2))

tab = with(overlaps, data.table(New=dt1[queryHits,]$Subset, New_id=dt1[queryHits,]$gene_id,
                                Old=dt2[subjectHits,]$Subset, Old_id=dt2[subjectHits,]$gene_id))

ncounts.chr = ncounts[Set=="Chromatin",]

tab[, `:=`(New_ex = ncounts.chr[match(New_id, ncounts.chr[, gene_id]), tpm],
           Old_ex = ncounts.chr[match(Old_id, ncounts.chr[, gene_id]), tpm])]

tab[New%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene"), Lab := "Gene-associated TU"]
tab[New%in%c("newTranscriptionalUnit"), Lab := "Independent TU"]

corr = tab[, cor(New_ex, Old_ex, method="spearman"), by=c("Lab")]

ranges = range(tab[New_ex>0 & Old_ex>0, list(range(log2(New_ex)), range(log2(Old_ex)))])

figureS2a = ggplot(tab[New_ex>0 & Old_ex>0], aes(x=log2(Old_ex), y=log2(New_ex), fill=Lab)) +
   # ggtitle("TU expression vs nearest gene expression") +
   geom_smooth(method="glm", se=FALSE, colour="firebrick3") +
   geom_point(pch=21, colour="black", size=0.75, alpha=0.5) +
   scale_fill_manual(values=colSet) +
   scale_x_continuous("Nearest annotated gene expression [ log2(TPM) ]", limits=ranges) +
   scale_y_continuous("Transcriptional unit expression [ log2(TPM) ]", limits=ranges) +
   geom_text_repel(data=corr, aes(label=paste(expression(~rho[(Spearman)]), "==", round(V1, 3))),
                   x=-Inf, y=Inf, hjust=-0.1, vjust=2, parse=TRUE, segment.colour=NA, direction="y", show.legend = FALSE) +
   facet_wrap(~Lab) +
   theme_bw() + theme(legend.position=c(0.9, 0.1), axis.text=element_text(size=12), axis.title=element_text(size=14),
                      strip.text=element_text(size=14), strip.background=element_rect(fill = NA, colour = NA)) +
   guides(fill=FALSE)
save_plot(figureS2a + coord_fixed(), filename="Figures_and_Tables/Figure_S2A.png", base_height = 5, base_aspect_ratio=2)

figureS1 = ggarrange(ggarrange(figureS2c, figureS2d, ncol=2, nrow=1, widths=c(3/5, 2/5), align="h", labels=c("A", "B")),
                     ggarrange(figureS2b, figureS2a, ncol=2, nrow=1, widths=c(2/5, 3/5), align="h", labels=c("C", "D")),
                     ncol=1, nrow=2)
save_and_plot(x=figureS1, bname="Figures_and_Tables/Figure_S1", width=14, height=12)
```

![](./Figures_and_Tables/Figure_S2B.png)

![](./Figures_and_Tables/Figure_S2A.png)

![](./Figures_and_Tables/Figure_S1.png)

```{r createExpressedSetHeLa}
sel.expressed = ncounts_nuc[Set%in%"Nuclear" & grepl("ENSG", gene_id),][tpm>=min.expr, gene_id]
sel.not.expressed = ncounts_nuc[Set%in%"Nuclear" & grepl("ENSG", gene_id),][tpm<(min.expr/10), gene_id]

new.anno.genes.expressed = new.anno.genes[new.anno.genes$gene_id%in%sel.expressed,]
new.anno.genes.expressed$Subset = as.character(NA)
new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%protein.coding,]$Subset = "Protein_coding"
new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%long.ncRNA,]$Subset = "Long_non-coding"
new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%ncRNA,]$Subset = "Non-coding"
new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%other,]$Subset = "Other annotated"

new.anno.genes.expressed = sort.GenomicRanges(c(new.anno.genes.expressed,
                                                new.anno.genes[!new.anno.genes$Set%in%"Annotated",]),
                                              ignore.strand=TRUE)

saveRDS(new.anno.genes.expressed, file="RData/newAnnotation_StringTie_expressed.rds")
```

```{r selectFeaturesPC}
Cod = new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%protein.coding,]

UoG = new.anno.genes.expressed[new.anno.genes.expressed$Subset%in%"UpstreamOfGene", ]
UoG = UoG[width(UoG)>5e3]
DoG = new.anno.genes.expressed[new.anno.genes.expressed$Subset%in%"DownstreamOfGene", ]
DoG = DoG[width(DoG)>5e3]

UoG = UoG[!is.na(precede(UoG, Cod)),]
UoG.dt = as.data.table(UoG)
UoG.dt[, FollowingDistance := distance(UoG, Cod[precede(UoG, Cod),])]
UoG = UoG[UoG$gene_id%in%UoG.dt[FollowingDistance==0, gene_id],]

UoG.Cod = Cod[precede(UoG, Cod),]
UoG.Cod$Subset = "UoG_Protein_coding"

DoG = DoG[!is.na(follow(DoG, Cod)),]
DoG.dt = as.data.table(DoG)
DoG.dt[, PrecedingDistance := distance(DoG, Cod[follow(DoG, Cod),])]
DoG = DoG[DoG$gene_id%in%DoG.dt[PrecedingDistance==0, gene_id],]

DoG$Set = "Proximal"
DoG$Subset = "DownstreamOfGene"

DoG.Cod = Cod[follow(DoG, Cod),]
DoG.Cod$Subset = "DoG_Protein_coding"

new.anno.genes.pc.proximal = c(UoG, UoG.Cod, DoG.Cod, DoG)
saveRDS(new.anno.genes.pc.proximal, file="RData/newAnnotation_StringTie_Proximal_PC.rds")

LoG = new.anno.genes.expressed[new.anno.genes.expressed$Subset%in%"LinkerOfGene", ]
LoG = LoG[width(LoG)>5e3]

LoG = LoG[!is.na(follow(LoG, Cod)) & !is.na(precede(LoG, Cod)),]
LoG.dt = as.data.table(LoG)
LoG.dt[, `:=`(PrecedingDistance = distance(LoG, Cod[follow(LoG, Cod),]),
              FollowingDistance = distance(LoG, Cod[precede(LoG, Cod),]))]
LoG = LoG[LoG$gene_id%in%LoG.dt[PrecedingDistance==0 & FollowingDistance==0, gene_id],]

LoG$Set = "Proximal"
LoG$Subset = "LinkerOfGene"

LoG.Upc = Cod[follow(LoG, Cod),]
LoG.Upc$Subset = "Upstream_protein_coding"

LoG.Dpc = Cod[precede(LoG, Cod),]
LoG.Dpc$Subset = "Downstream_protein_coding"

new.anno.genes.pc.linker = c(LoG.Upc, LoG, LoG.Dpc)
saveRDS(new.anno.genes.pc.linker, file="RData/newAnnotation_StringTie_Linker_PC.rds")
```

```{r selectFeaturesNC}
# Snc = new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%ncRNA,]
Lnc = new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%long.ncRNA,]
Lnc = Lnc[-as.data.table(findOverlaps(Lnc, maxgap=100))[queryHits!=subjectHits, queryHits],]

overlaps = as.data.table(distanceToNearest(new.anno.genes.expressed[new.anno.genes.expressed$gene_id%in%long.ncRNA,],
                                           new.anno.genes.expressed[new.anno.genes.expressed$Set%in%"Proximal",]))
overlaps = overlaps[distance<100,]
overlaps = overlaps[, N := .N, by="subjectHits"][N==1,]

red = reduce(c(Lnc, new.anno.genes.expressed[new.anno.genes.expressed$Set%in%"Proximal",][overlaps[, subjectHits],]),
             min.gapwidth=100)

Lnc = subsetByOverlaps(Lnc, red, type="equal")


DtG = new.anno.genes.expressed[new.anno.genes.expressed$Subset%in%"newTranscriptionalUnit", ]
DtG = DtG[width(DtG)>5e3]

# new.anno.genes.nc.distal = c(Snc, Lnc, DtG)
new.anno.genes.nc.distal = c(Lnc, DtG)
saveRDS(new.anno.genes.nc.distal, file="RData/newAnnotation_StringTie_Distal_NC.rds")
```

*****

### Metadata profiles

```{r setMetaprofileParam}
upstream = 5e3
downstream = 5e3
step = 50

if(!exists("new.anno.genes.pc.proximal"))
   new.anno.genes.pc.proximal = readRDS("RData/newAnnotation_StringTie_Proximal_PC.rds")

names.px = c("UpstreamOfGene", "UoG_Protein_coding", "DoG_Protein_coding", "DownstreamOfGene")
bedfiles.px = split(new.anno.genes.pc.proximal,
                    new.anno.genes.pc.proximal$Subset)[names.px]

tab.px = data.table(as.data.frame(unlist(lapply(bedfiles.px, length))), keep.rownames=TRUE)
setnames(tab.px, c("Unit", "N"))
tab.px[, Unit := factor(Unit, levels=names.px)]
setkey(tab.px, Unit)

if(!exists("new.anno.genes.pc.linker"))
   new.anno.genes.pc.linker = readRDS("RData/newAnnotation_StringTie_Linker_PC.rds")

names.lk = c("Upstream_protein_coding", "LinkerOfGene", "Downstream_protein_coding")
bedfiles.lk = split(new.anno.genes.pc.linker,
                    new.anno.genes.pc.linker$Subset)[names.lk]

tab.lk = data.table(as.data.frame(unlist(lapply(bedfiles.lk, length))), keep.rownames=TRUE)
setnames(tab.lk, c("Unit", "N"))
tab.lk[, Unit := factor(Unit, levels=names.lk)]
setkey(tab.lk, Unit)

if(!exists("new.anno.genes.nc.distal"))
   new.anno.genes.nc.distal = readRDS("RData/newAnnotation_StringTie_Distal_NC.rds")

names.nc = c("Long_non-coding", "newTranscriptionalUnit")
bedfiles.nc = split(new.anno.genes.nc.distal,
                    new.anno.genes.nc.distal$Subset)[names.nc]

tab.nc = data.table(as.data.frame(unlist(lapply(bedfiles.nc, length))), keep.rownames=TRUE)
setnames(tab.nc, c("Unit", "N"))
tab.nc[, Unit := factor(Unit, levels=names.nc)]
setkey(tab.nc, Unit)

smooth.spline2 <- function(formula, data, ...) { 
   mat <- model.frame(formula, data) 
   smooth.spline(mat[, 2], mat[, 1]) 
} 

predictdf.smooth.spline <- function(model, xseq, se, level) { 
   pred <- predict(model, xseq) 
   data.frame(x = xseq, y = pred$y) 
}
```

```{r noncodingExpression}
if( REBUILD ){
   bamTable = allTable[group%like%"HeLa" & (name%like%"Untreated" | name%like%"WildType") & !name%like%"pA"]
   files.bam = list.files("RNAseq_BAM", pattern="*.bam$", full.names=TRUE)
   files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
   names(files.bam) = bamTable[, group]
   
   counts = sapply(files.bam,
                   function(x){
                      print(x); flush(stdout())
                      if(as.character(x)%like%"_SE"){
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignments(file=x, param=param)
                      }else{
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                      }
                      reads = invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse"))
                      countOverlapsUnion(new.anno.genes, reads)
                   }, USE.NAMES=TRUE)
   rownames(counts) = new.anno.genes$gene_id
   
   saveRDS(counts, file="Annotation_counts_HeLa_Untreated_WildType_Fractions.rds")
}
```

```{r evaluateCodingExpression}
counts =  readRDS("Annotation_counts_HeLa_Untreated_WildType_Fractions.rds")

ncounts = apply(counts, 2, function(x) countToTPM(x, new.anno.genes))

ncounts.px = ncounts[new.anno.genes$Subset%in%c("UpstreamOfGene", "protein_coding", "DownstreamOfGene"),]

ncounts.px = data.table(gene_id = rownames(ncounts.px), Chromatin = ncounts.px[,1], Nucleoplasm = ncounts.px[,8], Nuclear = rowMeans(ncounts.px[,5:7]), Cytosol = rowMeans(ncounts.px[,2:4]))

bedfiles.px = new.anno.genes.pc.proximal[new.anno.genes.pc.proximal$gene_id%in%ncounts.px[rowMeans(ncounts.px[, 2:ncol(ncounts.px)])>1, gene_id],]

uog = bedfiles.px[bedfiles.px$Subset%in%"UpstreamOfGene",]
uogpc = bedfiles.px[bedfiles.px$Subset%in%"UoG_Protein_coding",]
overlaps = as.data.table(findOverlaps(uogpc, flank(uog, 1, start=F, both=F)))
uog = uog[overlaps[, subjectHits],]
uogpc = uogpc[overlaps[, queryHits],]

dog = bedfiles.px[bedfiles.px$Subset%in%"DownstreamOfGene",]
dogpc = bedfiles.px[bedfiles.px$Subset%in%"DoG_Protein_coding",]
overlaps = as.data.table(findOverlaps(dogpc, flank(dog, 1, start=T, both=F)))
dog = dog[overlaps[, subjectHits],]
dogpc = dogpc[overlaps[, queryHits],]

bedfiles.px = c(uog, uogpc, dogpc, dog)

bedfiles.px = split(bedfiles.px, bedfiles.px$Subset)[names.px]

tab.px = data.table(as.data.frame(unlist(lapply(bedfiles.px, length))), keep.rownames=TRUE)
setnames(tab.px, c("Unit", "N"))
tab.px[, Unit := factor(Unit, levels=names.px)]
setkey(tab.px, Unit)
```

```{r evaluateNoncodingExpression}
counts =  readRDS("Annotation_counts_HeLa_Untreated_WildType_Fractions.rds")

ncounts = apply(counts, 2, function(x) countToTPM(x, new.anno.genes))

ncounts.nc = ncounts[new.anno.genes$Subset%in%c("long_ncRNA", "newTranscriptionalUnit"),]

ncounts.nc = data.table(gene_id = rownames(ncounts.nc), Chromatin = ncounts.nc[,1], Nucleoplasm = ncounts.nc[,8], Nuclear = rowMeans(ncounts.nc[,5:7]), Cytosol = rowMeans(ncounts.nc[,2:4]))

names.nc = c("Long_non-coding", "newTranscriptionalUnit")
bedfiles.nc = split(new.anno.genes.nc.distal[new.anno.genes.nc.distal$gene_id%in%ncounts.nc[rowMeans(ncounts.nc[, 2:ncol(ncounts.nc)])>1, gene_id],],
                    new.anno.genes.nc.distal[new.anno.genes.nc.distal$gene_id%in%ncounts.nc[rowMeans(ncounts.nc[, 2:ncol(ncounts.nc)])>1, gene_id],]$Subset)[names.nc]

tab.nc = data.table(as.data.frame(unlist(lapply(bedfiles.nc, length))), keep.rownames=TRUE)
setnames(tab.nc, c("Unit", "N"))
tab.nc[, Unit := factor(Unit, levels=names.nc)]
setkey(tab.nc, Unit)
```

```{r exportRegionsBed, eval=FALSE}
for( i in seq_along(bedfiles.px) ){
   export.bed(bedfiles.px[[i]][strand(bedfiles.px[[i]])=="+",],
              file.path(working.folder, "Regions_bedfiles", paste0(names(bedfiles.px)[i], "_plus.bed")))
   export.bed(bedfiles.px[[i]][strand(bedfiles.px[[i]])=="-",],
              file.path(working.folder, "Regions_bedfiles", paste0(names(bedfiles.px)[i], "_minus.bed")))
}
for( i in seq_along(bedfiles.lk) ){
   export.bed(bedfiles.lk[[i]][strand(bedfiles.lk[[i]])=="+",],
              file.path(working.folder, "Regions_bedfiles", paste0(names(bedfiles.lk)[i], "_plus.bed")))
   export.bed(bedfiles.lk[[i]][strand(bedfiles.lk[[i]])=="-",],
              file.path(working.folder, "Regions_bedfiles", paste0(names(bedfiles.lk)[i], "_minus.bed")))
}
for( i in seq_along(bedfiles.nc) ){
   export.bed(bedfiles.nc[[i]][strand(bedfiles.nc[[i]])=="+",],
              file.path(working.folder, "Regions_bedfiles", paste0(names(bedfiles.nc)[i], "_plus.bed")))
   export.bed(bedfiles.nc[[i]][strand(bedfiles.nc[[i]])=="-",],
              file.path(working.folder, "Regions_bedfiles", paste0(names(bedfiles.nc)[i], "_minus.bed")))
}

export.gff3(new.anno.genes, file.path(working.folder, "New_regions.gff3"))
```

```{r generateTables}
files = list.files("NETseq_bw", pattern="*_plus_CPM.bigwig")

# NETseq
netTable = data.table(name=gsub("_plus_CPM.bigwig", "", files))
netTable[name%in%c("NETseq_mNETseq_Untreated-HeLa_1_PE", "NETseq_mNETseq_Untreated-HeLa_2_PE"), name := gsub("-", "_", name)]
netTable[, c("technique", "group", "treatment", "cell", "replicate", "library") := tstrsplit(name, "_")]

# Merge with RNAseq table
newTable = data.table(name=c(allTable[, name], netTable[, name]))[, c("technique", "group", "treatment", "cell", "replicate", "library") := tstrsplit(name, "_")]
newTable[group%in%"PROseq", technique := "GROseq"]

# Histone marks
histoneTable = fread("HistoneMarks_bw/metadata.tsv", select=c(1, 3, 7, 13, 26, 39))
histoneTable = histoneTable[`Output type`%in%"fold change over control" &
                               `Biosample term name`%in%c("K562", "HeLa-S3", "HepG2") &
                               `Biological replicate(s)`%in%"1, 2" &
                               Assembly%in%"GRCh38"]
setnames(histoneTable, c("Accession", "Type", "Cell", "Modification", "Replicate", "Assembly"))
histoneTable[, `:=`(Type=NULL, Modification=gsub("-human", "", Modification), Replicate=NULL, Assembly=NULL, File=paste0(Accession, ".bigWig"))]

# Add CHD7 and EP300 (P300) enhancer marks
histoneTable = rbindlist(list(histoneTable,
                         data.table(Accession = c("ENCFF575OWE", "ENCFF491ZOF", "ENCFF907QZG", "ENCFF949BRD", "ENCFF432IMS"), 
                                    Cell = c(rep("H1-hESC", 2), "K562", "HeLa-S3", "HepG2"),
                                    Modification = c("CHD7", rep("EP300", 4)),
                                    File = c("ENCFF575OWE.bigWig", "ENCFF491ZOF.bigWig", "ENCFF907QZG.bigWig", "ENCFF949BRD.bigWig", "ENCFF432IMS.bigWig"))))

# CAGEseq
cageTable = fread("FANTOM5_hg38/metadata.tsv", select=c(1, 3, 7, 26, 33, 39))
cageTable = cageTable[`Output type`%like%"strand signal of unique reads" &
                               `Biosample term name`%in%c("K562", "HeLa-S3", "HepG2") &
                               `Biological replicate(s)`%in%c(1, 2) &
                               Assembly%in%"GRCh38"]
setnames(cageTable, c("Accession", "Type", "Cell", "Replicate", "Set", "Assembly"))
cageTable = cageTable[order(Cell, Set, Replicate)]
cageTable[, `:=`(Replicate=NULL, Assembly=NULL, File=paste0(Accession, ".bigWig"))]
```

```{r generateProfiles_PX}
# Regions Start
for( grp in unique(newTable[, group]) ){

   print(grp)
  
   technique = ifelse(unique(newTable[group%in%grp, technique])%in%"RNAseq", "RNAseq", "NETseq")
   
   poswigfiles = as.list(paste0(technique, "_bw/", newTable[group%in%grp, name], "_plus_CPM.bigwig"))
   names(poswigfiles) = newTable[group%in%grp, name]
   negwigfiles = as.list(paste0(technique, "_bw/", newTable[group%in%grp, name], "_minus_CPM.bigwig"))
   names(negwigfiles) = newTable[group%in%grp, name]

      for( pos in c("start", "end") ){
      for( value.normalise in c("none", "max") ){
         value.start = ifelse(pos == "start", TRUE, FALSE)
         nor = ifelse(value.normalise == "none", "", "max_")
         
         if( !file.exists(paste0("RDS/profiles_proximal_", pos, "_", nor, technique, "_", grp, ".rds")) | REBUILD ){
            profiles = suppressMessages(genomationProfileWig(bedfiles.px, poswigfiles, negwigfiles,
                                                             upstream=upstream, downstream=downstream, step=step,
                                                             normalise=value.normalise, start=value.start,
                                                             normGroup=list(c("UpstreamOfGene", "UoG_Protein_coding"), c("DoG_Protein_coding", "DownstreamOfGene"))))
            
            profiles[, Unit := factor(Unit, levels=names.px)]
            
            profiles[, c("technique", "group", "treatment", "cell", "replicate", "library") := tstrsplit(Sample, "_")]
            
            saveRDS(profiles, file=paste0("RDS/profiles_proximal_", pos, "_", nor, technique, "_", grp, ".rds"))  
         }        
      }
   }
}

# Histone Marks
for( ce in unique(histoneTable[, Cell]) ){
   
   print(ce)

   poswigfiles = as.list(paste0("HistoneMarks_bw/", histoneTable[Cell%in%ce, File]))
   names(poswigfiles) = histoneTable[Cell%in%ce, Modification]

   for( pos in c("start", "end") ){
      for( value.normalise in c("none", "max") ){
         value.start = ifelse(pos == "start", TRUE, FALSE)
         nor = ifelse(value.normalise == "none", "", "max_")
         
         if( !file.exists(paste0("RDS/profiles_proximal_", pos, "_", nor, "HistoneMarks_", ce, ".rds"))  | REBUILD ){
            profiles = suppressMessages(genomationProfileWig(bedfiles.px, posReadsList=poswigfiles, negReadsList=NULL,
                                                             upstream=upstream, downstream=downstream, step=step,
                                                             normalise=value.normalise, start=value.start, stranded=FALSE,
                                                             normGroup=list(c("UpstreamOfGene", "UoG_Protein_coding"), c("DoG_Protein_coding", "DownstreamOfGene"))))
            
            profiles[, Unit := factor(Unit, levels=names.px)]
            
            saveRDS(profiles, file=paste0("RDS/profiles_proximal_", pos, "_", nor, "HistoneMarks_", ce, ".rds"))
         }
      }
   }
}

# CAGEseq
for( ce in unique(cageTable[, Cell]) ){
   
   print(ce)
   
   poswigfiles = as.list(paste0("FANTOM5_hg38/", cageTable[Cell%in%ce & grepl("plus", Type), File]))
   names(poswigfiles) = cageTable[Cell%in%ce & grepl("plus", Type), Set]
   negwigfiles = as.list(paste0("FANTOM5_hg38/", cageTable[Cell%in%ce & grepl("minus", Type), File]))
   names(negwigfiles) = cageTable[Cell%in%ce & grepl("minus", Type), Set]

   for( pos in c("start", "end") ){
      value.start = ifelse(pos == "start", TRUE, FALSE)
      if( !file.exists(paste0("RDS/profiles_proximal_", pos, "_max_CAGE_", ce, ".rds"))  | REBUILD ){
         profiles = suppressMessages(genomationProfileWig(bedfiles.px, posReadsList=poswigfiles, negReadsList=negwigfiles,
                                                          upstream=upstream, downstream=downstream, step=step,
                                                          normalise="max", start=value.start))
         
         profiles[, Unit := factor(Unit, levels=names.px)]
         
         saveRDS(profiles, file=paste0("RDS/profiles_proximal_", pos, "_max_CAGE_", ce, ".rds"))
      }
   }
}
```

```{r generateProfiles_LK}
# Regions Start
for( grp in unique(newTable[, group]) ){
   
   print(grp)

   technique = ifelse(unique(newTable[group%in%grp, technique])%in%"RNAseq", "RNAseq", "NETseq")
   
   poswigfiles = as.list(paste0(technique, "_bw/", newTable[group%in%grp, name], "_plus_CPM.bigwig"))
   names(poswigfiles) = newTable[group%in%grp, name]
   negwigfiles = as.list(paste0(technique, "_bw/", newTable[group%in%grp, name], "_minus_CPM.bigwig"))
   names(negwigfiles) = newTable[group%in%grp, name]
   
   for( pos in c("start", "end") ){
      for( value.normalise in c("none", "max") ){
         value.start = ifelse(pos == "start", TRUE, FALSE)
         nor = ifelse(value.normalise == "none", "", "max_")
         
         if( !file.exists(paste0("RDS/profiles_linker_", pos, "_", nor, technique, "_", grp, ".rds"))  | REBUILD ){
            profiles = suppressMessages(genomationProfileWig(bedfiles.lk, poswigfiles, negwigfiles,
                                                             upstream=upstream, downstream=downstream, step=step,
                                                             normalise=value.normalise, start=value.start,
                                                             normGroup=list(c("Upstream_protein_coding", "LinkerOfGene", "Downstream_protein_coding"))))
            
            profiles[, Unit := factor(Unit, levels=names.lk)]
            
            profiles[, c("technique", "group", "treatment", "cell", "replicate", "library") := tstrsplit(Sample, "_")]
            
            saveRDS(profiles, file=paste0("RDS/profiles_linker_", pos, "_", nor, technique, "_", grp, ".rds"))  
         }     
      }
   }
}

# Histone Marks
for( ce in unique(histoneTable[, Cell]) ){
   
   print(ce)

   poswigfiles = as.list(paste0("HistoneMarks_bw/", histoneTable[Cell%in%ce, File]))
   names(poswigfiles) = histoneTable[Cell%in%ce, Modification]
   
   for( pos in c("start", "end") ){
      for( value.normalise in c("none", "max") ){
         value.start = ifelse(pos == "start", TRUE, FALSE)
         nor = ifelse(value.normalise == "none", "", "max_")
         
         if( !file.exists(paste0("RDS/profiles_linker_", pos, "_", nor, "HistoneMarks_", ce, ".rds"))  | REBUILD ){
            profiles = suppressMessages(genomationProfileWig(bedfiles.lk, posReadsList=poswigfiles, negReadsList=NULL,
                                                             upstream=upstream, downstream=downstream, step=step,
                                                             normalise=value.normalise, start=value.start, stranded=FALSE,
                                                             normGroup=list(c("Upstream_protein_coding", "LinkerOfGene", "Downstream_protein_coding"))))
            
            profiles[, Unit := factor(Unit, levels=names.lk)]
            
            saveRDS(profiles, file=paste0("RDS/profiles_linker_", pos, "_", nor, "HistoneMarks_", ce, ".rds"))
         }
      }
   }
}

# CAGEseq
for( ce in unique(cageTable[, Cell]) ){

   print(ce)
   
   poswigfiles = as.list(paste0("FANTOM5_hg38/", cageTable[Cell%in%ce & grepl("plus", Type), File]))
   names(poswigfiles) = cageTable[Cell%in%ce & grepl("plus", Type), Set]
   negwigfiles = as.list(paste0("FANTOM5_hg38/", cageTable[Cell%in%ce & grepl("minus", Type), File]))
   names(negwigfiles) = cageTable[Cell%in%ce & grepl("minus", Type), Set]
   
   for( pos in c("start", "end") ){
      value.start = ifelse(pos == "start", TRUE, FALSE)
      if( !file.exists(paste0("RDS/profiles_linker_", pos, "_max_CAGE_", ce, ".rds"))  | REBUILD ){
         profiles = suppressMessages(genomationProfileWig(bedfiles.lk, posReadsList=poswigfiles, negReadsList=negwigfiles,
                                                          upstream=upstream, downstream=downstream, step=step,
                                                          normalise="max", start=value.start))
         
         profiles[, Unit := factor(Unit, levels=names.lk)]
         
         saveRDS(profiles, file=paste0("RDS/profiles_linker_", pos, "_max_CAGE_", ce, ".rds"))
      }
   }
}
```

```{r generateProfiles_NC}
# Regions Start
for( grp in unique(newTable[, group]) ){

   print(grp)

   technique = ifelse(unique(newTable[group%in%grp, technique])%in%"RNAseq", "RNAseq", "NETseq")
   
   poswigfiles = as.list(paste0(technique, "_bw/", newTable[group%in%grp, name], "_plus_CPM.bigwig"))
   names(poswigfiles) = newTable[group%in%grp, name]
   negwigfiles = as.list(paste0(technique, "_bw/", newTable[group%in%grp, name], "_minus_CPM.bigwig"))
   names(negwigfiles) = newTable[group%in%grp, name]

   for( pos in c("start", "end") ){
      for( value.normalise in c("none", "max") ){
         value.start = ifelse(pos == "start", TRUE, FALSE)
         nor = ifelse(value.normalise == "none", "", "max_")
         
         if( !file.exists(paste0("RDS/profiles_distal_", pos, "_", nor, technique, "_", grp, ".rds"))  | REBUILD ){
            profiles = suppressMessages(genomationProfileWig(bedfiles.nc, poswigfiles, negwigfiles,
                                                             upstream=upstream, downstream=downstream, step=step,
                                                             normalise=value.normalise, start=value.start))
            
            profiles[, Unit := factor(Unit, levels=names.nc)]
            
            profiles[, c("technique", "group", "treatment", "cell", "replicate", "library") := tstrsplit(Sample, "_")]
            
            saveRDS(profiles, file=paste0("RDS/profiles_distal_", pos, "_", nor, technique, "_", grp, ".rds"))  
         }        
      }
   }
}

# Histone Marks
for( ce in unique(histoneTable[, Cell]) ){
   
   print(ce)

   poswigfiles = as.list(paste0("HistoneMarks_bw/", histoneTable[Cell%in%ce, File]))
   names(poswigfiles) = histoneTable[Cell%in%ce, Modification]
   
   for( pos in c("start", "end") ){
      for( value.normalise in c("none", "max") ){
         value.start = ifelse(pos == "start", TRUE, FALSE)
         nor = ifelse(value.normalise == "none", "", "max_")
         
         if( !file.exists(paste0("RDS/profiles_distal_", pos, "_", nor, "HistoneMarks_", ce, ".rds"))  | REBUILD ){
            profiles = suppressMessages(genomationProfileWig(bedfiles.nc, posReadsList=poswigfiles, negReadsList=NULL,
                                                             upstream=upstream, downstream=downstream, step=step,
                                                             normalise=value.normalise, start=value.start, stranded=FALSE))
            
            profiles[, Unit := factor(Unit, levels=names.nc)]
            
            saveRDS(profiles, file=paste0("RDS/profiles_distal_", pos, "_", nor, "HistoneMarks_", ce, ".rds"))
         }
      }
   }
}

# CAGEseq
for( ce in unique(cageTable[, Cell]) ){
   
   print(ce)

   poswigfiles = as.list(paste0("FANTOM5_hg38/", cageTable[Cell%in%ce & grepl("plus", Type), File]))
   names(poswigfiles) = cageTable[Cell%in%ce & grepl("plus", Type), Set]
   negwigfiles = as.list(paste0("FANTOM5_hg38/", cageTable[Cell%in%ce & grepl("minus", Type), File]))
   names(negwigfiles) = cageTable[Cell%in%ce & grepl("minus", Type), Set]
   
   for( pos in c("start", "end") ){
      value.start = ifelse(pos == "start", TRUE, FALSE)
      if( !file.exists(paste0("RDS/profiles_distal_", pos, "_max_CAGE_", ce, ".rds"))  | REBUILD ){
         profiles = suppressMessages(genomationProfileWig(bedfiles.nc, posReadsList=poswigfiles, negReadsList=negwigfiles,
                                                          upstream=upstream, downstream=downstream, step=step,
                                                          normalise="max", start=value.start))
         
         profiles[, Unit := factor(Unit, levels=names.nc)]
         
         saveRDS(profiles, file=paste0("RDS/profiles_distal_", pos, "_max_CAGE_", ce, ".rds"))
      }
   }
}
```

```{r profilesFiguresCoding}
remove_suffix <- function(str) return(gsub("UoG_|DoG_", "", str))

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               rbindlist(list(Chromatin=readRDS(paste("RDS/profiles_proximal", reg, "max_RNAseq_Chromatin.rds", sep="_"))[treatment%in%"Untreated"],
                                              Nuclear=readRDS(paste("RDS/profiles_proximal", reg, "max_RNAseq_Nuclear.rds", sep="_"))[treatment%in%"WildType"],
                                              Cytoplasm=readRDS(paste("RDS/profiles_proximal", reg, "max_RNAseq_Cytoplasm.rds", sep="_"))[treatment%in%"WildType"]))), idcol="Region")

profiles = profiles[(Region%in%"start" & Unit%in%c("UpstreamOfGene", "UoG_Protein_coding"))|(Region%in%"end" & Unit%in%c("DownstreamOfGene", "DoG_Protein_coding"))]

profiles[group%in%"Cytoplasm", group := "Cytoplasmic"][, group := factor(group, levels=c("Chromatin", "Nuclear", "Cytoplasmic"))]
profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "group")]

profiles[, Unit := factor(Unit, levels=names.px)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=group)) +
                # ggtitle("RNA-seq across cellular compartments") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_tableau(name="", palette = "Miller Stone") +
                geom_text(data=tab.px[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figure3A = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure3A = annotate_figure(figure3A, top = text_grob("RNA-seq across cellular compartments", face = "bold", size = 12))


profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               rbindlist(list(Chromatin=readRDS(paste("RDS/profiles_proximal", reg, "max_RNAseq_Chromatin.rds", sep="_"))[treatment%in%"Untreated"],
                                              PROseq=readRDS(paste("RDS/profiles_proximal", reg, "NETseq_PROseq.rds", sep="_"))[, list(Sample, Unit, Position, Value, technique, group, treatment, cell, replicate, library)] ))), idcol="Region")

profiles = profiles[(Region%in%"start" & Unit%in%c("UpstreamOfGene", "UoG_Protein_coding"))|(Region%in%"end" & Unit%in%c("DownstreamOfGene", "DoG_Protein_coding"))]

profiles[group%in%"Chromatin", treatment := "RNAseq (HeLa)"][group%in%"PROseq" & treatment%in%"Chromatin", treatment := "RNAseq (B cells)"]
profiles[group%in%"PROseq" & treatment%in%"GROseq", treatment := "GROseq (B cells)"][group%in%"PROseq" & treatment%in%"PROseq", treatment := "PROseq (B cells)"]

profiles = profiles[, treatment := factor(treatment, levels=c("RNAseq (HeLa)", "RNAseq (B cells)", "GROseq (B cells)", "PROseq (B cells)"))]
   
profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "group", "treatment")]

profiles[, Unit := factor(Unit, levels=names.px)]

cols = setNames(c("#4f6980", "#638b66", "#ffae34", "#ef6f6a"), levels(profiles[, treatment]))

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("RNA-seq, GRO-seq and PRO-seq (B cells)") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values = cols) +
                geom_text(data=tab.px[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figure3B = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure3B = annotate_figure(figure3B, top = text_grob("RNA-seq, GRO-seq and PRO-seq (B cells)", face = "bold", size = 12))


profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste("RDS/profiles_proximal", reg, "NETseq_mNETseq.rds", sep="_"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("UpstreamOfGene", "UoG_Protein_coding"))|(Region%in%"end" & Unit%in%c("DownstreamOfGene", "DoG_Protein_coding"))]

profiles = profiles[treatment%like%"Untreated",][treatment%in%"Untreated", treatment := "Total"][, treatment := gsub("Untreated-", "", treatment)]
   
profiles[, treatment := factor(treatment, levels=c("Total", "Y1P", "S2P", "T4P", "S5P", "S7P"))]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "treatment")]

profiles[, Unit := factor(Unit, levels=names.px)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,][treatment%in%c("Total", "Y1P", "T4P", "S5P"),], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("NET-seq and Pol II CTD modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_viridis("", discrete=TRUE) +
                geom_text(data=tab.px[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure3C = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure3C = annotate_figure(figure3C, top = text_grob("NET-seq and Pol II CTD modifications", face = "bold", size = 12))

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("NET-seq and Pol II CTD modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_viridis("", discrete=TRUE) +
                geom_text(data=tab.px[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure3CS = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure3CS = annotate_figure(figure3CS, top = text_grob("NET-seq and Pol II CTD modifications", face = "bold", size = 12))


HM.active = c("H3K9ac","H3K4me3","H3K4me2","H2AFZ","H3K79me2")
HM.enhancer = c("H3K9me3","H3K4me1","H3K27ac", "CHD7", "EP300")
HM.inactive = c("H3K27me3","H4K20me1","H3K36me3")
HM.suppl = c("H3K4me1", "H3K36me3", "EP300", "H3K9me3")

# HM.selection = c("H3K9ac", "H3K4me3", "CHD7", "EP300", "H3K27me3", "H3K36me3")
HM.selection = c("H3K4me3", "H3K27ac", "EP300", "H3K27me3", "H3K9me3")
# HM.selection = c("H3K4me3", "H3K27ac", "H3K4me1", "H3K27me3", "H3K9me3")

myColours = c("#99000d", "#cb181d", "#ef3b2c", "#fb6a4a", "#fc9272",
              "#006d2c", "#2ca25f", "#66c2a4", "#ff7f00", "#810f7c",
              "#08519c", "#3182bd", "#6baed6")
names(myColours) = c(HM.active, HM.enhancer, HM.inactive)

myColours.selection = setNames(c("#99000d", "#ef3b2c", "#ffd308", "#08519c", "#6baed6"), HM.selection)
myColours.suppl = setNames(c("#006d2c", "#66c2a4", "#ffd308", "#810f7c"), HM.suppl)

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste0("RDS/profiles_proximal_", reg, "_max_HistoneMarks_HeLa-S3.rds"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("UpstreamOfGene", "UoG_Protein_coding"))|(Region%in%"end" & Unit%in%c("DownstreamOfGene", "DoG_Protein_coding"))]

profiles = profiles[Sample%in%HM.selection,][, Sample := factor(Sample, levels=HM.selection)]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "Sample")]

profiles[, Sample := factor(Sample, levels=names(myColours.selection))]

profiles[, Value := Value-min(Value), by=c("Unit", "Sample")]
   
profiles[, Unit := factor(Unit, levels=names.px)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Sample)) +
                # ggtitle("Histone modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values=myColours.selection) +
                geom_text(data=tab.px[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure3D = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure3D = annotate_figure(figure3D, top = text_grob("Histone modifications", face = "bold", size = 12))

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste0("RDS/profiles_proximal_", reg, "_max_HistoneMarks_HeLa-S3.rds"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("UpstreamOfGene", "UoG_Protein_coding"))|(Region%in%"end" & Unit%in%c("DownstreamOfGene", "DoG_Protein_coding"))]

profiles = profiles[Sample%in%HM.suppl,][, Sample := factor(Sample, levels=HM.suppl)]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "Sample")]

profiles[, Sample := factor(Sample, levels=names(myColours.suppl))]

profiles[, Value := Value-min(Value), by=c("Unit", "Sample")]
   
profiles[, Unit := factor(Unit, levels=names.px)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Sample)) +
                # ggtitle("Histone modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values=myColours.suppl) +
                geom_text(data=tab.px[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure3DS = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure3DS = annotate_figure(figure3DS, top = text_grob("Histone modifications", face = "bold", size = 12))


if( !exists("length.seq") )
   length.seq = seqlengths(Hsapiens)[names(seqlengths(Hsapiens))%in%seqlevels(unlist(bedfiles.px))]

gr = with(fread("gunzip -c FANTOM5_hg38/hg38_fair+new_CAGE_peaks_phase1and2.bed.gz", sep="\t"),
          GRanges(V1, IRanges(V2, V3), V6))
gr = keepSeqlevels(gr, names(length.seq), pruning.mode="coarse")
seqlengths(gr) = length.seq

poswigfiles = list(CAGEcluster = gr[strand(gr)=="+"])
negwigfiles = list(CAGEcluster = gr[strand(gr)=="-"])

profiles  = rbindlist(lapply(setNames(c(TRUE, FALSE), c("start", "end")),
       function(x)
          suppressMessages(genomationProfileWig(bedfiles.px, poswigfiles, negwigfiles, upstream=upstream, downstream=downstream, step=step, normalise="none", start=x))), idcol="Region")

# Add the antisense
antisense = rbindlist(lapply(setNames(c(TRUE, FALSE), c("start", "end")),
       function(x)
          suppressMessages(genomationProfileWig(bedfiles.px, negwigfiles, poswigfiles, upstream=upstream, downstream=downstream, step=step, normalise="none", start=x))), idcol="Region")

profiles = rbindlist(list(Sense=profiles, Antisense=antisense), idcol="Orientation")

profiles[Orientation%in%"Antisense", Value := -Value]

profiles = profiles[(Region%in%"start" & Unit%in%c("UpstreamOfGene", "UoG_Protein_coding"))|(Region%in%"end" & Unit%in%c("DownstreamOfGene", "DoG_Protein_coding"))]

profiles[, `:=`(Unit = factor(Unit, levels=names.px), Orientation=factor(Orientation, levels=c("Sense", "Antisense")))]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Orientation)) +
                # ggtitle("CAGE-seq identified clusters") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage", limits=c(-0.1, 0.3)) + # Using manual limits!!!
                scale_colour_tableau(name="", palette = "Traffic") +
                geom_text(data=tab.px[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figure3E = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure3E = annotate_figure(figure3E, top = text_grob("CAGE-seq identified clusters", face = "bold", size = 12))

# figure3 = ggarrange(figure3A, figure3B, figure3C, figure3E, figure3D, nrow=5, ncol=1, labels=c("A", "B", "C", "D", "E"))
figure3 = ggarrange(figure3A, figure3E, figure3C, figure3D, nrow=4, ncol=1, labels=c("A", "B", "C", "D"))
save_and_plot(x=figure3, bname="Figures_and_Tables/Figure3", width=14, height=12, plot=FALSE)
```

![](./Figures_and_Tables/Figure3.png)

```{r profilesFiguresNonCoding}
remove_suffix <- function(str) return(gsub("Long_non-coding", "Long ncRNA", gsub("newTranscriptionalUnit", "Independent TU", str)))

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               rbindlist(list(Chromatin=readRDS(paste("RDS/profiles_distal", reg, "max_RNAseq_Chromatin.rds", sep="_"))[treatment%in%"Untreated"],
                                              Nuclear=readRDS(paste("RDS/profiles_distal", reg, "max_RNAseq_Nuclear.rds", sep="_"))[treatment%in%"WildType"],
                                              Cytoplasm=readRDS(paste("RDS/profiles_distal", reg, "max_RNAseq_Cytoplasm.rds", sep="_"))[treatment%in%"WildType"]))), idcol="Region")

profiles = profiles[(Region%in%"start" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))|(Region%in%"end" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))]

profiles[group%in%"Cytoplasm", group := "Cytoplasmic"][, group := factor(group, levels=c("Chromatin", "Nuclear", "Cytoplasmic"))]
profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "group")]

profiles[, Unit := factor(Unit, levels=names.nc)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=group)) +
                # ggtitle("RNA-seq across cellular compartments") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_tableau(name="", palette = "Miller Stone") +
                geom_text(data=tab.nc[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figure4A = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure4A = annotate_figure(figure4A, top = text_grob("RNA-seq across cellular compartments", face = "bold", size = 12))


profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               rbindlist(list(Chromatin=readRDS(paste("RDS/profiles_distal", reg, "max_RNAseq_Chromatin.rds", sep="_"))[treatment%in%"Untreated"],
                                              PROseq=readRDS(paste("RDS/profiles_distal", reg, "NETseq_PROseq.rds", sep="_"))[, list(Sample, Unit, Position, Value, technique, group, treatment, cell, replicate, library)] ))), idcol="Region")

profiles = profiles[(Region%in%"start" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))|(Region%in%"end" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))]

profiles[group%in%"Chromatin", treatment := "RNAseq (HeLa)"][group%in%"PROseq" & treatment%in%"Chromatin", treatment := "RNAseq (B cells)"]
profiles[group%in%"PROseq" & treatment%in%"GROseq", treatment := "GROseq (B cells)"][group%in%"PROseq" & treatment%in%"PROseq", treatment := "PROseq (B cells)"]

profiles = profiles[, treatment := factor(treatment, levels=c("RNAseq (HeLa)", "RNAseq (B cells)", "GROseq (B cells)", "PROseq (B cells)"))]
   
profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "group", "treatment")]

profiles[, Unit := factor(Unit, levels=names.nc)]

cols = setNames(c("#4f6980", "#638b66", "#ffae34", "#ef6f6a"), levels(profiles[, treatment]))

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("RNA-seq, GRO-seq and PRO-seq (B cells)") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values = cols) +
                geom_text(data=tab.nc[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figure4B = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure4B = annotate_figure(figure4B, top = text_grob("RNA-seq, GRO-seq and PRO-seq (B cells)", face = "bold", size = 12))

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste("RDS/profiles_distal", reg, "NETseq_mNETseq.rds", sep="_"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))|(Region%in%"end" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))]

profiles = profiles[treatment%like%"Untreated",][treatment%in%"Untreated", treatment := "Total"][, treatment := gsub("Untreated-", "", treatment)]
   
profiles[, treatment := factor(treatment, levels=c("Total", "Y1P", "S2P", "T4P", "S5P", "S7P"))]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "treatment")]

profiles[, Unit := factor(Unit, levels=names.nc)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,][treatment%in%c("Total", "Y1P", "T4P", "S5P"),], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("NET-seq and Pol II CTD modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_viridis("", discrete=TRUE) +
                geom_text(data=tab.nc[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure4C = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure4C = annotate_figure(figure4C, top = text_grob("NET-seq and Pol II CTD modifications", face = "bold", size = 12))

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("NET-seq and Pol II CTD modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_viridis("", discrete=TRUE) +
                geom_text(data=tab.nc[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure4CS = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure4CS = annotate_figure(figure4CS, top = text_grob("NET-seq and Pol II CTD modifications", face = "bold", size = 12))


HM.active = c("H3K9ac","H3K4me3","H3K4me2","H2AFZ","H3K79me2")
HM.enhancer = c("H3K9me3","H3K4me1","H3K27ac", "CHD7", "EP300")
HM.inactive = c("H3K27me3","H4K20me1","H3K36me3")
HM.suppl = c("H3K4me1", "H3K36me3", "EP300", "H3K9me3")

# HM.selection = c("H3K9ac", "H3K4me3", "CHD7", "EP300", "H3K27me3", "H3K36me3")
HM.selection = c("H3K4me3", "H3K27ac", "EP300", "H3K27me3", "H3K9me3")
# HM.selection = c("H3K4me3", "H3K27ac", "H3K4me1", "H3K27me3", "H3K9me3")

myColours = c("#99000d", "#cb181d", "#ef3b2c", "#fb6a4a", "#fc9272",
              "#006d2c", "#2ca25f", "#66c2a4", "#ff7f00", "#810f7c",
              "#08519c", "#3182bd", "#6baed6")
names(myColours) = c(HM.active, HM.enhancer, HM.inactive)

myColours.selection = setNames(c("#99000d", "#ef3b2c", "#ffd308", "#08519c", "#6baed6"), HM.selection)
myColours.suppl = setNames(c("#006d2c", "#66c2a4", "#ffd308", "#810f7c"), HM.suppl)

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste0("RDS/profiles_distal_", reg, "_max_HistoneMarks_HeLa-S3.rds"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))|(Region%in%"end" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))]

profiles = profiles[Sample%in%HM.selection,][, Sample := factor(Sample, levels=HM.selection)]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "Sample")]

profiles[, Sample := factor(Sample, levels=names(myColours.selection))]

profiles[, Value := Value-min(Value), by=c("Unit", "Sample")]
   
profiles[, Unit := factor(Unit, levels=names.nc)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Sample)) +
                # ggtitle("Histone modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values=myColours.selection) +
                geom_text(data=tab.nc[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure4D = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure4D = annotate_figure(figure4D, top = text_grob("Histone modifications", face = "bold", size = 12))

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste0("RDS/profiles_distal_", reg, "_max_HistoneMarks_HeLa-S3.rds"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))|(Region%in%"end" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))]

profiles = profiles[Sample%in%HM.suppl,][, Sample := factor(Sample, levels=HM.suppl)]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "Sample")]

profiles[, Sample := factor(Sample, levels=names(myColours.suppl))]

profiles[, Value := Value-min(Value), by=c("Unit", "Sample")]
   
profiles[, Unit := factor(Unit, levels=names.nc)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Sample)) +
                # ggtitle("Histone modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values=myColours.suppl) +
                geom_text(data=tab.nc[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figure4DS = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure4DS = annotate_figure(figure4DS, top = text_grob("Histone modifications", face = "bold", size = 12))


if( !exists("length.seq") )
   length.seq = seqlengths(Hsapiens)[names(seqlengths(Hsapiens))%in%seqlevels(unlist(bedfiles.px))]

gr = with(fread("gunzip -c FANTOM5_hg38/hg38_fair+new_CAGE_peaks_phase1and2.bed.gz", sep="\t"),
          GRanges(V1, IRanges(V2, V3), V6))
gr = keepSeqlevels(gr, names(length.seq), pruning.mode="coarse")
seqlengths(gr) = length.seq

poswigfiles = list(CAGEcluster = gr[strand(gr)=="+"])
negwigfiles = list(CAGEcluster = gr[strand(gr)=="-"])

profiles  = rbindlist(lapply(setNames(c(TRUE, FALSE), c("start", "end")),
       function(x)
          suppressMessages(genomationProfileWig(bedfiles.nc, poswigfiles, negwigfiles, upstream=upstream, downstream=downstream, step=step, normalise="none", start=x))), idcol="Region")

# Add the antisense
antisense = rbindlist(lapply(setNames(c(TRUE, FALSE), c("start", "end")),
       function(x)
          suppressMessages(genomationProfileWig(bedfiles.nc, negwigfiles, poswigfiles, upstream=upstream, downstream=downstream, step=step, normalise="none", start=x))), idcol="Region")

profiles = rbindlist(list(Sense=profiles, Antisense=antisense), idcol="Orientation")

profiles[Orientation%in%"Antisense", Value := -Value]

profiles = profiles[(Region%in%"start" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))|(Region%in%"end" & Unit%in%c("Long_non-coding", "newTranscriptionalUnit"))]

profiles[, `:=`(Unit = factor(Unit, levels=names.nc), Orientation=factor(Orientation, levels=c("Sense", "Antisense")))]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Orientation)) +
                # ggtitle("CAGE-seq identified clusters") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage", limits=c(-0.1, 0.2)) + # Using manual limits!!!
                scale_colour_tableau(name="", palette = "Traffic") +
                geom_text(data=tab.nc[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figure4E = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figure4E = annotate_figure(figure4E, top = text_grob("CAGE-seq identified clusters", face = "bold", size = 12))

# figure4 = ggarrange(figure4A, figure4B, figure4C, figure4E, figure4D, nrow=5, ncol=1, labels=c("A", "B", "C", "D", "E"))
figure4 = ggarrange(figure4A, figure4E, figure4C, figure4D, nrow=4, ncol=1, labels=c("A", "B", "C", "D"))
save_and_plot(x=figure4, bname="Figures_and_Tables/Figure4", width=14, height=12, plot=FALSE)
```

![](./Figures_and_Tables/Figure4.png)

```{r profilesFiguresLinkers}
remove_suffix <- function(str) return(gsub("LinkerOfGene_start", "Linker start", gsub("LinkerOfGene_end", "Linker end", gsub("Upstream_protein_coding_start", "Upstream protein coding", gsub("Downstream_protein_coding_end", "Downstream protein coding", str)))))

tab.lk = data.table(Unit = c("Upstream_protein_coding_start", "LinkerOfGene_start", "LinkerOfGene_end", "Downstream_protein_coding_end"),
                    N = unique(tab.lk[, N]))
tab.lk[, Unit := factor(Unit, levels=Unit)]
names.lk = tab.lk[, Unit] 

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               rbindlist(list(Chromatin=readRDS(paste("RDS/profiles_linker", reg, "max_RNAseq_Chromatin.rds", sep="_"))[treatment%in%"Untreated"],
                                              Nuclear=readRDS(paste("RDS/profiles_linker", reg, "max_RNAseq_Nuclear.rds", sep="_"))[treatment%in%"WildType"],
                                              Cytoplasm=readRDS(paste("RDS/profiles_linker", reg, "max_RNAseq_Cytoplasm.rds", sep="_"))[treatment%in%"WildType"]))), idcol="Region")

profiles = profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene"))|(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding"))]
profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene")), Unit := paste(Unit, "start", sep="_")]
profiles[(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding")), Unit := paste(Unit, "end", sep="_")]

profiles[group%in%"Cytoplasm", group := "Cytoplasmic"][, group := factor(group, levels=c("Chromatin", "Nuclear", "Cytoplasmic"))]
profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "group")]

profiles[, Unit := factor(Unit, levels=names.lk)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=group)) +
                # ggtitle("RNA-seq across cellular compartments") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_tableau(name="", palette = "Miller Stone") +
                geom_text(data=tab.lk[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figureS3A = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figureS3A = annotate_figure(figureS3A, top = text_grob("RNA-seq across cellular compartments", face = "bold", size = 12))


profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               rbindlist(list(Chromatin=readRDS(paste("RDS/profiles_linker", reg, "max_RNAseq_Chromatin.rds", sep="_"))[treatment%in%"Untreated"],
                                              PROseq=readRDS(paste("RDS/profiles_linker", reg, "NETseq_PROseq.rds", sep="_"))[, list(Sample, Unit, Position, Value, technique, group, treatment, cell, replicate, library)] ))), idcol="Region")

profiles = profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene"))|(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding"))]
profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene")), Unit := paste(Unit, "start", sep="_")]
profiles[(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding")), Unit := paste(Unit, "end", sep="_")]

profiles[group%in%"Chromatin", treatment := "RNAseq (HeLa)"][group%in%"PROseq" & treatment%in%"Chromatin", treatment := "RNAseq (B cells)"]
profiles[group%in%"PROseq" & treatment%in%"GROseq", treatment := "GROseq (B cells)"][group%in%"PROseq" & treatment%in%"PROseq", treatment := "PROseq (B cells)"]

profiles = profiles[, treatment := factor(treatment, levels=c("RNAseq (HeLa)", "RNAseq (B cells)", "GROseq (B cells)", "PROseq (B cells)"))]
   
profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "group", "treatment")]

profiles[, Unit := factor(Unit, levels=names.lk)]

cols = setNames(c("#4f6980", "#638b66", "#ffae34", "#ef6f6a"), levels(profiles[, treatment]))

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("RNA-seq, GRO-seq and PRO-seq (B cells)") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values = cols) +
                geom_text(data=tab.lk[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figureS3B = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figureS3B = annotate_figure(figureS3B, top = text_grob("RNA-seq, GRO-seq and PRO-seq (B cells)", face = "bold", size = 12))


profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste("RDS/profiles_linker", reg, "NETseq_mNETseq.rds", sep="_"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene"))|(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding"))]
profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene")), Unit := paste(Unit, "start", sep="_")]
profiles[(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding")), Unit := paste(Unit, "end", sep="_")]

profiles = profiles[treatment%like%"Untreated",][treatment%in%"Untreated", treatment := "Total"][, treatment := gsub("Untreated-", "", treatment)]
   
profiles[, treatment := factor(treatment, levels=c("Total", "Y1P", "S2P", "T4P", "S5P", "S7P"))]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "treatment")]

profiles[, Unit := factor(Unit, levels=names.lk)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,][treatment%in%c("Total", "Y1P", "T4P", "S5P"),], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("NET-seq and Pol II CTD modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_viridis("", discrete=TRUE) +
                geom_text(data=tab.lk[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figureS3C = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figureS3C = annotate_figure(figureS3C, top = text_grob("NET-seq and Pol II CTD modifications", face = "bold", size = 12))

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=treatment)) +
                # ggtitle("NET-seq and Pol II CTD modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_viridis("", discrete=TRUE) +
                geom_text(data=tab.lk[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figureS3CS = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figureS3CS = annotate_figure(figureS3CS, top = text_grob("NET-seq and Pol II CTD modifications", face = "bold", size = 12))


HM.active = c("H3K9ac","H3K4me3","H3K4me2","H2AFZ","H3K79me2")
HM.enhancer = c("H3K9me3","H3K4me1","H3K27ac", "CHD7", "EP300")
HM.inactive = c("H3K27me3","H4K20me1","H3K36me3")
HM.suppl = c("H3K4me1", "H3K36me3", "EP300", "H3K9me3")

# HM.selection = c("H3K9ac", "H3K4me3", "CHD7", "EP300", "H3K27me3", "H3K36me3")
HM.selection = c("H3K4me3", "H3K27ac", "EP300", "H3K27me3", "H3K9me3")
# HM.selection = c("H3K4me3", "H3K27ac", "H3K4me1", "H3K27me3", "H3K9me3")

myColours = c("#99000d", "#cb181d", "#ef3b2c", "#fb6a4a", "#fc9272",
              "#006d2c", "#2ca25f", "#66c2a4", "#ff7f00", "#810f7c",
              "#08519c", "#3182bd", "#6baed6")
names(myColours) = c(HM.active, HM.enhancer, HM.inactive)

myColours.selection = setNames(c("#99000d", "#ef3b2c", "#ffd308", "#08519c", "#6baed6"), HM.selection)
myColours.suppl = setNames(c("#006d2c", "#66c2a4", "#ffd308", "#810f7c"), HM.suppl)

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste0("RDS/profiles_linker_", reg, "_max_HistoneMarks_HeLa-S3.rds"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene"))|(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding"))]
profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene")), Unit := paste(Unit, "start", sep="_")]
profiles[(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding")), Unit := paste(Unit, "end", sep="_")]

profiles = profiles[Sample%in%HM.selection,][, Sample := factor(Sample, levels=HM.selection)]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "Sample")]

profiles[, Sample := factor(Sample, levels=names(myColours.selection))]

profiles[, Value := Value-min(Value), by=c("Unit", "Sample")]
   
profiles[, Unit := factor(Unit, levels=names.lk)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Sample)) +
                # ggtitle("Histone modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values=myColours.selection) +
                geom_text(data=tab.lk[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figureS3D = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figureS3D = annotate_figure(figureS3D, top = text_grob("Histone modifications", face = "bold", size = 12))

profiles = rbindlist(lapply(setNames(c("start", "end"), c("start", "end")),
                            function(reg)
                               readRDS(paste0("RDS/profiles_linker_", reg, "_max_HistoneMarks_HeLa-S3.rds"))), idcol="Region")
   
profiles = profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene"))|(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding"))]
profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene")), Unit := paste(Unit, "start", sep="_")]
profiles[(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding")), Unit := paste(Unit, "end", sep="_")]

profiles = profiles[Sample%in%HM.suppl,][, Sample := factor(Sample, levels=HM.suppl)]

profiles = profiles[, list(Value=mean(Value)), by=c("Region", "Unit", "Position", "Sample")]

profiles[, Sample := factor(Sample, levels=names(myColours.suppl))]

profiles[, Value := Value-min(Value), by=c("Unit", "Sample")]
   
profiles[, Unit := factor(Unit, levels=names.lk)]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Sample)) +
                # ggtitle("Histone modifications") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage") +
                scale_colour_manual("", values=myColours.suppl) +
                geom_text(data=tab.lk[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1), ncol = 6, byrow = TRUE)))
figureS3DS = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figureS3DS = annotate_figure(figureS3DS, top = text_grob("Histone modifications", face = "bold", size = 12))

if( !exists("length.seq") )
   length.seq = seqlengths(Hsapiens)[names(seqlengths(Hsapiens))%in%seqlevels(unlist(bedfiles.lk))]

gr = with(fread("gunzip -c FANTOM5_hg38/hg38_fair+new_CAGE_peaks_phase1and2.bed.gz", sep="\t"),
          GRanges(V1, IRanges(V2, V3), V6))
gr = keepSeqlevels(gr, names(length.seq), pruning.mode="coarse")
seqlengths(gr) = length.seq

poswigfiles = list(CAGEcluster = gr[strand(gr)=="+"])
negwigfiles = list(CAGEcluster = gr[strand(gr)=="-"])

profiles  = rbindlist(lapply(setNames(c(TRUE, FALSE), c("start", "end")),
       function(x)
          suppressMessages(genomationProfileWig(bedfiles.lk, poswigfiles, negwigfiles, upstream=upstream, downstream=downstream, step=step, normalise="none", start=x))), idcol="Region")

# Add the antisense
antisense = rbindlist(lapply(setNames(c(TRUE, FALSE), c("start", "end")),
       function(x)
          suppressMessages(genomationProfileWig(bedfiles.lk, negwigfiles, poswigfiles, upstream=upstream, downstream=downstream, step=step, normalise="none", start=x))), idcol="Region")

profiles = rbindlist(list(Sense=profiles, Antisense=antisense), idcol="Orientation")

profiles[Orientation%in%"Antisense", Value := -Value]

profiles = profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene"))|(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding"))]
profiles[(Region%in%"start" & Unit%in%c("Upstream_protein_coding", "LinkerOfGene")), Unit := paste(Unit, "start", sep="_")]
profiles[(Region%in%"end" & Unit%in%c("LinkerOfGene", "Downstream_protein_coding")), Unit := paste(Unit, "end", sep="_")]

profiles[, `:=`(Unit = factor(Unit, levels=names.lk), Orientation=factor(Orientation, levels=c("Sense", "Antisense")))]

ggl = lapply(profiles[, unique(Region)],
             function(reg)
                ggplot(profiles[Region%in%reg,], aes(x=Position, y=Value, col=Orientation)) +
                # ggtitle("CAGE-seq identified clusters") +
                geom_line(lwd=0.75) + geom_vline(xintercept=0, lwd=0.5, lty=2) +
                scale_x_continuous(paste("Relative position (Kb)"), breaks=c(-5:5),
                                   labels=c("-5", "-4", "-3", "-2", "-1", paste("Region", as.character(reg), sep="\n"), "1", "2", "3", "4", "5")) +
                scale_y_continuous("Normalised coverage", limits=c(-0.1, 0.5)) + # Using manual limits!!!
                scale_colour_tableau(name="", palette = "Traffic") +
                geom_text(data=tab.lk[Unit%in%profiles[Region%in%reg, Unit],], aes(label=N), x=-Inf, y=Inf, hjust=-0.5, vjust=1.5, colour="black", inherit.aes=FALSE, parse=FALSE, size=3) +
                facet_wrap(~Unit, ncol=4, labeller=labeller(Unit = remove_suffix)) +
                theme_bw() + theme(legend.text=element_text(size=10)) +
                guides(colour=guide_legend(override.aes = list(size=1))))
figureS3E = ggarrange(plotlist=ggl, ncol=2, legend="bottom", common.legend=TRUE)
figureS3E = annotate_figure(figureS3E, top = text_grob("CAGE-seq identified clusters", face = "bold", size = 12))

# figureS3 = ggarrange(figureS3A, figureS3B, figureS3C, figureS3E, figureS3D, nrow=5, ncol=1, labels=c("A", "B", "C", "D", "E"))
figureS3 = ggarrange(figureS3A, figureS3E, figureS3C, figureS3D, nrow=4, ncol=1, labels=c("A", "B", "C", "D"))
save_and_plot(x=figureS3, bname="Figures_and_Tables/FigureS3", width=14, height=12, plot=FALSE)
```

![](./Figures_and_Tables/FigureS3.png)

```{r profilesSupplementary}
figureNETseq = ggarrange(figure3CS, figureS3CS, figure4CS, nrow=3, ncol=1, labels=c("A", "B", "C"))
save_and_plot(x=figureNETseq, bname="Figures_and_Tables/FigureSuppNET", width=14, height=9, plot=FALSE)

figureHistones = ggarrange(figure3DS, figureS3DS, figure4DS, nrow=3, ncol=1, labels=c("A", "B", "C"))
save_and_plot(x=figureHistones, bname="Figures_and_Tables/FigureSuppHist", width=14, height=9, plot=FALSE)
```

### EXOSC3

```{r count_EXOSC3}
if( REBUILD ){
   bamTable = newTable[group%in%c("Chromatin", "Nucleoplasm") & treatment%in%c("Untreated", "siLuc", "siEXOC3") & cell%in%"HeLa",]
   files.bam = list.files("RNAseq_BAM", pattern="*.bam$", full.names=TRUE)
   files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
   names(files.bam) = bamTable[, name]
   
   files.bam = files.bam[grepl("Untreated", files.bam)|grepl("_3_", files.bam)|(grepl("_4_", files.bam))]
   
   counts = lapply(files.bam,
                   function(x){
                      print(x); flush(stdout())
                      if(as.character(x)%like%"_SE"){
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignments(file=x, param=param)
                      }else{
                         param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                               isDuplicate=FALSE,
                                                               isSecondaryAlignment=FALSE),
                                              mapqFilter=255)             
                         reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                      }
                      reads = invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse"))
                      countOverlapsUnion(new.anno.genes, reads)
                   })
   counts = lapply(counts, function(x) data.table(Index = seq_along(x), gene_id = new.anno.genes$gene_id, count = x))
   
   counts = rbindlist(counts, idcol="Sample")
   
   saveRDS(counts, file="RDS/RNAseq_all_EXOC3_ChromatinNucleoplasm.rds")
}
```

```{r plot_EXOSC3}
counts = readRDS("RDS/RNAseq_all_EXOC3_ChromatinNucleoplasm.rds")

counts[, Subset := new.anno.genes[match(gene_id, new.anno.genes$gene_id),]$Subset]

# pl = counts[, list(Count = sum(count)), by=c("Sample", "Subset")][, list(Subset, Percentage = Count/sum(Count)), by="Sample"]
# pl[, c("Fraction", "Treatment", "Replicate") := tstrsplit(Sample, "_")[c(2,3,5)]]
# 
# # pl[Subset%in%c("UpstreamOfGene", "LinkerOfGene", "DownstreamOfGene"), Subset := "Nascent TU"]
# pl[Subset%in%c("newTranscriptionalUnit"), Subset := "New TU"]
# 
# # pl[, Subset := factor(Subset, levels=rev(c(factBiotype2[1:5], "Unassigned", factBiotype2[9:6])))]
# myCol = setNames(c(tableau_color_pal(palette="Tableau 10")(10)[c(1,2,5,3,7)], tableau_color_pal(palette="Classic Gray 5")(5)), c(factBiotype2, "Unassigned"))
# 
# gg = ggplot(pl, aes(x=paste(Fraction, Treatment, Replicate, sep="\n"), y=Percentage, fill=Subset)) +
#    geom_bar(position="stack", stat="identity") +
#    scale_x_discrete("") + scale_y_continuous(labels=percent) +
#    # scale_fill_manual(values=myCol) +
#    facet_zoom(y=Subset!="protein_coding") +
#    theme_light()
# save_and_plot(x=gg, bname="Figures_and_Tables/Check_EXOC3_reads_distribution", width=14, height=5, plot=FALSE)

df = data.frame(dcast.data.table(counts, gene_id~Sample, value.var="count"), row.names="gene_id")

sizeFactors = DESeq2::estimateSizeFactorsForMatrix(as.matrix(df[rowSums(df>=1)>1,]))
df.norm = t(t(df)/sizeFactors)

# df.norm = t(t(df)/colSums(df)*1e6)

# df_tpm = apply(df, 2, function(x) countToTPM(x, new.anno.genes[match(rownames(df), new.anno.genes$gene_id),]))

bedfiles.all = list(UpstreamOfGene=bedfiles.px[[1]],
                    Protein_coding=unique(unlist(bedfiles.px[2:3])),
                    DownstreamOfGene=bedfiles.px[[4]],
                    Long_non_coding=bedfiles.nc[[1]],
                    `Independent TU`=bedfiles.nc[[2]])

profiles = lapply(bedfiles.all,
                  function(x)
                     melt.data.table(data.table(Index = 1:length(x),
                                                data.table(df.norm[match(x$gene_id, rownames(df.norm)),])),
                                     id.vars="Index", variable.name="Sample", value.name="Score"))
profiles = rbindlist(profiles, idcol="Unit")
profiles[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
profiles = dcast.data.table(profiles[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
                            Unit+Index+Treatment~Fraction, value.var="Score")

bedfiles.log = list(UpstreamGene=bedfiles.lk[[1]],
                    LinkerOfGenes=bedfiles.lk[[2]],
                    DownstreamGene=bedfiles.lk[[3]])

profiles.log = lapply(bedfiles.log,
                  function(x)
                     melt.data.table(data.table(Index = 1:length(x),
                                                data.table(df.norm[match(x$gene_id, rownames(df.norm)),])),
                                     id.vars="Index", variable.name="Sample", value.name="Score"))
profiles.log = rbindlist(profiles.log, idcol="Unit")
profiles.log[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
profiles.log = dcast.data.table(profiles.log[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
                            Unit+Index+Treatment~Fraction, value.var="Score")

# profiles_tpm = lapply(bedfiles.all,
#                   function(x)
#                      melt.data.table(data.table(Index = 1:length(x),
#                                                 data.table(df_tpm[match(x$gene_id, rownames(df_tpm)),])),
#                                      id.vars="Index", variable.name="Sample", value.name="Score"))
# profiles_tpm = rbindlist(profiles_tpm, idcol="Unit")
# profiles_tpm[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
# profiles_tpm = dcast.data.table(profiles_tpm[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
#                             Unit+Index+Treatment~Fraction, value.var="Score")

data = profiles[Chromatin>0 | Nucleoplasm>0,]
data[, Ratio := (Nucleoplasm)/(Chromatin)]

data = data[Treatment%in%"Untreated",]
setkeyv(data, c("Unit", "Index"))

check = data[is.finite(Ratio), .N, by=key(data)][N==1]
data = data[check, nomatch=0]

data[, Unit := factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU"))]
data[, log2Ratio := log2(Ratio)]

data.log = profiles.log[Chromatin>0 | Nucleoplasm>0,]
data.log[, Ratio := (Nucleoplasm)/(Chromatin)]

data.log = data.log[Treatment%in%"Untreated",]
setkeyv(data.log, c("Unit", "Index"))

check = data.log[is.finite(Ratio), .N, by=key(data.log)][N==1]
data.log = data.log[check, nomatch=0]

data.log[, Unit := factor(Unit, levels=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene"))]
data.log[, log2Ratio := log2(Ratio)]

# figure5A = ggplot(data, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
#    ggtitle("Nucleoplasm over chromatin expression") +
#    geom_hline(yintercept=0, lty=2, lwd=0.75) +
#    geom_violin(alpha=0.5) +
#    geom_boxplot(notch=TRUE, width=0.25, outlier.colour="grey30", fill="white") +
#    scale_x_discrete("", limits = rev(levels(data[,Unit]))) +
#    scale_y_continuous("Expression ratio (log2)", breaks=seq(-8, 8, 2)) +
#    scale_fill_tableau(name="Treatment", palette = "Tableau 10", guide=FALSE) +
#    theme_bw() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14)) + coord_flip()
figure5A = ggplot(data, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
   ggtitle("Nucleoplasm over chromatin expression") +
   # geom_hline(yintercept=0, lty=2, lwd=0.75) +
   geom_violin(position="dodge", alpha=0.5, trim=TRUE) +
   geom_boxplot(notch=TRUE, width=0.1, position=position_dodge(width=0.9), aes(col=Treatment), fill="white", outlier.shape=20) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)") +
   scale_colour_manual(values=c("grey30")) +
   scale_fill_tableau(name="Treatment", palette = "Tableau 10", guide=FALSE) +
   theme_cleveland() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90)) +
   guides(colour=FALSE) + theme_cleveland()
save_and_plot(x=figure5A, bname="Figures_and_Tables/Figure5A", width=8, height=6.5, plot=FALSE)

figure5A.log = ggplot(data.log, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
   ggtitle("Nucleoplasm over chromatin expression") +
   # geom_hline(yintercept=0, lty=2, lwd=0.75) +
   geom_violin(position="dodge", alpha=0.5, trim=TRUE) +
   geom_boxplot(notch=TRUE, width=0.1, position=position_dodge(width=0.9), aes(col=Treatment), fill="white", outlier.shape=20) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)") +
   scale_colour_manual(values=c("grey30")) +
   scale_fill_tableau(name="Treatment", palette = "Tableau 10", guide=FALSE) +
   theme_cleveland() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90)) +
   guides(colour=FALSE) + theme_cleveland()

pl = melt.data.table(profiles[Treatment=="Untreated",], id.vars=c("Unit", "Index", "Treatment"))
pl[, Min := .SD[value>0, min(value)], by=c("Unit", "variable")]
pl[value==0, value := Min]
pl[, log2Value := log2(value)]
tab = pl[, .(log2Value = -3, N=.N/2), by=c("Unit")]

pl.log = melt.data.table(profiles.log[Treatment=="Untreated",], id.vars=c("Unit", "Index", "Treatment"))
pl.log[, Min := .SD[value>0, min(value)], by=c("Unit", "variable")]
pl.log[value==0, value := Min]
pl.log[, log2Value := log2(value)]
tab.log = pl.log[, .(log2Value = -3, N=.N/2), by=c("Unit")]

figure5A_box = ggboxplot(pl, x="Unit", y="log2Value", fill="variable",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Normalised expression (log2)") +
   scale_colour_manual(values=c("grey30")) +
   scale_fill_tableau(name="Treatment", palette = "Tableau 10") +
   geom_text(data=tab, aes(x = Unit, y = log2Value, label = N), vjust = 1) + 
   theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Nuclear fraction:")) +
   stat_compare_means(aes(group=variable), paired = TRUE, label = "p.signif", label.y=18)
save_and_plot(x=figure5A_box, bname="Figures_and_Tables/Figure5A_box", width=8, height=6.5, plot=FALSE)

figure5A_box.log = ggboxplot(pl.log, x="Unit", y="log2Value", fill="variable",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Normalised expression (log2)") +
   scale_colour_manual(values=c("grey30")) +
   scale_fill_tableau(name="Treatment", palette = "Tableau 10") +
   geom_text(data=tab.log, aes(x = Unit, y = log2Value, label = N), vjust = 1) + 
   theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Nuclear fraction:")) +
   stat_compare_means(aes(group=variable), paired = TRUE, label = "p.signif", label.y=18)

data = profiles[Chromatin>0 | Nucleoplasm>0,]
data[, Ratio := (Nucleoplasm)/(Chromatin)]

data = data[Treatment%in%c("siLuc", "siEXOC3"),]
setkeyv(data, c("Unit", "Index"))

check = data[is.finite(Ratio), .N, by=key(data)][N==2]
data = data[check, nomatch=0]

data[, `:=`(Treatment = factor(Treatment, levels=c("siLuc", "siEXOC3")),
            Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]

figure5B = ggplot(data, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
   # ggtitle("Nucleoplasm over chromatin expression") +
   geom_hline(yintercept=0, lty=2, lwd=0.5) +
   geom_boxplot(notch=TRUE, position=position_dodge(width=0.9), outlier.colour=NA, alpha=0.75, outlier.shape=20) +
   scale_x_discrete("", limits = rev(levels(data[,Unit]))) +
   scale_y_continuous("Expression ratio (log2)", breaks=seq(-8, 8, 2)) +
   scale_fill_tableau(name="Treatment", palette = "Tableau 10") +
   theme_bw() + theme(legend.position=c(0.125, 0.9), legend.box.margin=margin(0,0,0,0,"cm"),
                      axis.text=element_text(size=12), axis.title=element_text(size=14)) +
                      # legend.text=element_text(size=8), legend.title=element_text(size=8), legend.key.size=unit(0.5, "cm")) +
   coord_flip(ylim=c(-8,8))
# theme(legend.title=element_blank(), legend.position=c(0.9, 0.9), legend.box.margin=margin(0,0,0,0,"cm")) + guides(col=guide_legend(ncol=2, byrow=TRUE), fill=guide_legend(ncol=2, byrow=TRUE))
   # guides(col=FALSE, fill=FALSE)
save_and_plot(x=figure5B, bname="Figures_and_Tables/Figure5B", width=8, height=6.5, plot=FALSE)

# sapply(as.character(unique(data[, Unit])),
#        function(x)
#           data[Unit%in%as.character(x), wilcox.test(.SD[Treatment%in%"siEXOC3", Ratio], .SD[Treatment%in%"siLuc", Ratio])$p.value])
# 
# # data = dcast.data.table(profiles[Treatment%in%c("siLuc", "siEXOC3")], Unit+Index~Treatment, value.var="Ratio")
# # data[, Ratio := log2(siEXOC3/siLuc)]
# # setkeyv(data, c("Unit", "Index"))
# # 
# # check = data[is.finite(Ratio), .N, by=key(data)][N==1]
# # data = data[check, nomatch=0]
# # 
# # figure6B = ggplot(data, aes(x=Unit, y=Ratio, fill=as.factor(1), colour=as.factor(1))) +
# #    ggtitle("siEXOSC3 over siLuc expression") +
# #    geom_hline(yintercept=0, lty=2, lwd=0.5) +
# #    geom_violin(position="dodge", col="black", alpha=0.5) +
# #    geom_boxplot(notch=TRUE, width=0.1, position=position_dodge(width=0.9), outlier.colour=NA, fill="white") +
# #    scale_x_discrete("") +
# #    scale_y_continuous("Expression ratio (log2)") +
# #    scale_colour_manual("Treatment", values=tableau_color_pal(palette = "Tableau 10")(3)[3]) +
# #    scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Tableau 10")(3)[3]) +
# #    theme_bw() + guides(col=FALSE, fill=FALSE)
# # save_and_plot(x=figure6B, bname="Figures_and_Tables/Figure6B", width=8, height=5, plot=FALSE)

data = melt.data.table(profiles[Treatment%in%c("siLuc", "siEXOC3"),], id.vars=c("Unit", "Index", "Treatment"))
data[, Min := .SD[value>0, min(value)], by=c("Unit", "variable")]
data[value==0, value := Min]

data = dcast.data.table(data, Unit+Index~variable+Treatment, value.var="value")
data[, `:=`(Chromatin = log2(Chromatin_siEXOC3/Chromatin_siLuc), Nucleoplasm = log2(Nucleoplasm_siEXOC3/Nucleoplasm_siLuc))]
data = melt.data.table(data, id.vars=c("Unit", "Index"), measure.vars=c("Chromatin", "Nucleoplasm"), variable.name="Fraction", value.name="log2Ratio")
setkeyv(data, c("Unit", "Index"))

check = data[is.finite(log2Ratio), .N, by=key(data)][N==2]
data = data[check, nomatch=0]

data[, `:=`(Fraction = factor(Fraction, levels=c("Chromatin", "Nucleoplasm")),
            Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]

# figure5B_bis = ggplot(data, aes(x=Unit, y=log2Ratio, fill=Fraction)) +
#    ggtitle("siEXOSC3 over siLuc expression") +
#    geom_hline(yintercept=0, lty=2, lwd=0.5) +
#    geom_boxplot(notch=TRUE, position=position_dodge(width=0.9), outlier.colour="grey30", alpha=0.75) +
#    scale_x_discrete("", limits = rev(levels(data[,Unit]))) +
#    scale_y_continuous("Expression ratio (log2)") +
#    scale_fill_manual("Fraction", values=tableau_color_pal(palette = "Nuriel Stone")(3)) +
#    theme_bw() + theme(legend.position=c(0.125, 0.9), legend.box.margin=margin(0,0,0,0,"cm"),
#                       axis.text=element_text(size=12), axis.title=element_text(size=14)) +
#                       # legend.text=element_text(size=8), legend.title=element_text(size=8), legend.key.size=unit(0.5, "cm")) +
#    coord_flip()
figure5B_box = ggboxplot(data, x="Unit", y="log2Ratio", fill="Fraction",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)", limits=c(-6, 7)) +
   scale_colour_manual(values=c("grey30", "grey30")) +
   scale_fill_manual("Fraction", values=tableau_color_pal(palette = "Nuriel Stone")(3)) +
   theme(legend.position="top", axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90)) +
   guides(fill = guide_legend(title="siEXOSC3 over siLuc expression; nuclear fraction:")) +
   stat_compare_means(aes(group = Fraction), paired=TRUE, label = "p.signif", label.y=7)
save_and_plot(x=figure5B_box, bname="Figures_and_Tables/Figure5B_box", width=8, height=6.5, plot=FALSE)

data = melt.data.table(profiles[Treatment%in%c("siLuc", "siEXOC3"),], id.vars=c("Unit", "Index", "Treatment"))
data[, Min := .SD[value>0, min(value)], by=c("Unit", "variable")]
data[value==0, value := Min]

data = dcast.data.table(data, Unit+Index~variable+Treatment, value.var="value")
data[, `:=`(siEXOSC3 = log2(Nucleoplasm_siEXOC3/Chromatin_siEXOC3), siLuc = log2(Nucleoplasm_siLuc/Chromatin_siLuc))]
data = melt.data.table(data, id.vars=c("Unit", "Index"), measure.vars=c("siEXOSC3", "siLuc"), variable.name="Treatment", value.name="log2Ratio")
setkeyv(data, c("Unit", "Index"))

check = data[is.finite(log2Ratio), .N, by=key(data)][N==2]
data = data[check, nomatch=0]

data[, `:=`(Treatment = factor(Treatment, levels=c("siLuc", "siEXOSC3")),
            Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]

data.log = melt.data.table(profiles.log[Treatment%in%c("siLuc", "siEXOC3"),], id.vars=c("Unit", "Index", "Treatment"))
data.log[, Min := .SD[value>0, min(value)], by=c("Unit", "variable")]
data.log[value==0, value := Min]

data.log = dcast.data.table(data.log, Unit+Index~variable+Treatment, value.var="value")
data.log[, `:=`(siEXOSC3 = log2(Nucleoplasm_siEXOC3/Chromatin_siEXOC3), siLuc = log2(Nucleoplasm_siLuc/Chromatin_siLuc))]
data.log = melt.data.table(data.log, id.vars=c("Unit", "Index"), measure.vars=c("siEXOSC3", "siLuc"), variable.name="Treatment", value.name="log2Ratio")
setkeyv(data.log, c("Unit", "Index"))

check = data.log[is.finite(log2Ratio), .N, by=key(data.log)][N==2]
data.log = data.log[check, nomatch=0]

data.log[, `:=`(Treatment = factor(Treatment, levels=c("siLuc", "siEXOSC3")),
            Unit = factor(Unit, levels=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene")))]

figure5B_box_bis = ggboxplot(data, x="Unit", y="log2Ratio", fill="Treatment",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)") +
   scale_colour_manual(values=c("grey30", "grey30")) +
   scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Nuriel Stone")(3)) +
   theme(legend.position="top", axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90)) +
   guides(fill = guide_legend(title="Nucleoplasm over chromatin expression; treatment:")) +
   stat_compare_means(aes(group = Treatment), paired=TRUE, label = "p.signif", label.y=8)
save_and_plot(x=figure5B_box_bis, bname="Figures_and_Tables/Figure5B_box_bis", width=8, height=6.5, plot=FALSE)

figure5B_box_bis.log = ggboxplot(data.log, x="Unit", y="log2Ratio", fill="Treatment",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)") +
   scale_colour_manual(values=c("grey30", "grey30")) +
   scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Nuriel Stone")(3)) +
   theme(legend.position="top", axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90)) +
   guides(fill = guide_legend(title="Nucleoplasm over chromatin expression; treatment:")) +
   stat_compare_means(aes(group = Treatment), paired=TRUE, label = "p.signif", label.y=8)
```

![](./Figures_and_Tables/Figure5A.png)

![](./Figures_and_Tables/Figure5A_box.png)

![](./Figures_and_Tables/Figure5B.png)

![](./Figures_and_Tables/Figure5B_box.png)

![](./Figures_and_Tables/Figure5B_box_bis.png)

### CSTF2

```{r count_CSTF2_CPSF73}
bamTable = newTable[(group%in%c("ChrRNA") & cell%in%"HeLa") & !treatment%in%c("siLuc-T4P", "siCPSF73-T4P"),]
files.bam = list.files("NETseq_BAM", pattern="*.bam$", full.names=TRUE)
files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
names(files.bam) = bamTable[, name]

counts = lapply(files.bam,
                function(x){
                   print(x); flush(stdout())
                   if(as.character(x)%like%"_SE"){
                      param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                            isDuplicate=FALSE,
                                                            isSecondaryAlignment=FALSE),
                                           mapqFilter=255)             
                      reads = readGAlignments(file=x, param=param)
                   }else{
                      param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                            isDuplicate=FALSE,
                                                            isSecondaryAlignment=FALSE),
                                           mapqFilter=255)             
                      reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                   }
                   reads = invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse"))
                   countOverlapsUnion(new.anno.genes, reads)
                })
counts = lapply(counts, function(x) data.table(Index = seq_along(x), gene_id = new.anno.genes$gene_id, count = x))

counts = rbindlist(counts, idcol="Sample")

saveRDS(counts, file="RDS/RNAseq_all_CSTF2_CPSF73_ChromatinNucleoplasm.rds")
```

```{r plot_CSTF2_CPSF73}
counts = readRDS("RDS/RNAseq_all_CSTF2_CPSF73_ChromatinNucleoplasm.rds")

counts[, Subset := new.anno.genes[match(gene_id, new.anno.genes$gene_id),]$Subset]

# pl = counts[, list(Count = sum(count)), by=c("Sample", "Subset")][, list(Subset, Percentage = Count/sum(Count)), by="Sample"]
# pl[, c("Fraction", "Treatment", "Replicate") := tstrsplit(Sample, "_")[c(2,3,5)]]
# 
# pl[, Subset := factor(Subset, levels=rev(c(factBiotype2[1:5], "Unassigned", factBiotype2[9:6])))]
# myCol = setNames(c(tableau_color_pal(palette="Tableau 10")(10)[c(1,2,5,3,7)], tableau_color_pal(palette="Classic Gray 5")(5)), c(factBiotype2, "Unassigned"))
# 
# gg = ggplot(pl, aes(x=paste(Fraction, Treatment, Replicate, sep="\n"), y=Percentage, fill=Subset)) +
#    geom_bar(position="stack", stat="identity") +
#    scale_x_discrete("") + scale_y_continuous(labels=percent) +
#    scale_fill_manual(values=myCol) +
#    facet_zoom(y=Subset!="protein_coding") +
#    theme_light()
# save_and_plot(x=gg, bname="Figures_and_Tables/Check_CSTF2_CPSF7_reads_distribution", width=24, height=5, plot=FALSE)

df = data.frame(dcast.data.table(counts[!grepl("T4P", Sample),], gene_id~Sample, value.var="count"), row.names="gene_id")

sizeFactors = DESeq2::estimateSizeFactorsForMatrix(as.matrix(df[rowSums(df>=1)>1,]))

df.norm = t(t(df)/sizeFactors)

bedfiles.all = list(UpstreamOfGene=bedfiles.px[[1]],
                    Protein_coding=unique(unlist(bedfiles.px[2:3])),
                    DownstreamOfGene=bedfiles.px[[4]],
                    Long_non_coding=bedfiles.nc[[1]],
                    `Independent TU`=bedfiles.nc[[2]])

profiles = lapply(bedfiles.all,
                  function(x)
                     melt.data.table(data.table(Index = 1:length(x),
                                                data.table(df.norm[match(x$gene_id, rownames(df.norm)),])),
                                     id.vars="Index", variable.name="Sample", value.name="Score"))
profiles = rbindlist(profiles, idcol="Unit")
profiles[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
profiles = dcast.data.table(profiles[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
                            Unit+Index+Treatment~Fraction, value.var="Score")
setkeyv(profiles, c("Unit", "Index"))

bedfiles.log = list(UpstreamGene=bedfiles.lk[[1]],
                    LinkerOfGenes=bedfiles.lk[[2]],
                    DownstreamGene=bedfiles.lk[[3]])

profiles.log = lapply(bedfiles.log,
                  function(x)
                     melt.data.table(data.table(Index = 1:length(x),
                                                data.table(df.norm[match(x$gene_id, rownames(df.norm)),])),
                                     id.vars="Index", variable.name="Sample", value.name="Score"))
profiles.log = rbindlist(profiles.log, idcol="Unit")
profiles.log[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
profiles.log = dcast.data.table(profiles.log[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
                            Unit+Index+Treatment~Fraction, value.var="Score")
setkeyv(profiles.log, c("Unit", "Index"))

data = profiles[profiles[grepl("Untreated", Treatment),], `:=`(Untreated = i.ChrRNA, Ratio = ChrRNA/i.ChrRNA)]
# data = profiles[profiles[grepl("siLuc", Treatment),], `:=`(Untreated = i.ChrRNA, Ratio = ChrRNA/i.ChrRNA)]

# data = data[!grepl("Untreated", Treatment) & !grepl("siLuc", Treatment)][, Treatment := gsub("\\.", "-", Treatment)]
data = data[!grepl("Untreated", Treatment),][, Treatment := gsub("\\.", "-", Treatment)]
data[, Treatment := gsub("siLuc-CSTF2", "siLuc", Treatment)]
data[, `:=`(Treatment = factor(Treatment, levels=c("siLuc", "siCSTF2", "siCSTF2T", "siCSTF2-siCSTF2T", "siCPSF3", "siXRN2")),
            Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]
data = data[Treatment%in%c("siCSTF2-siCSTF2T", "siCPSF3"),]

data.log = profiles.log[profiles.log[grepl("Untreated", Treatment),], `:=`(Untreated = i.ChrRNA, Ratio = ChrRNA/i.ChrRNA)]

data.log = data.log[!grepl("Untreated", Treatment),][, Treatment := gsub("\\.", "-", Treatment)]
data.log[, Treatment := gsub("siLuc-CSTF2", "siLuc", Treatment)]
data.log[, `:=`(Treatment = factor(Treatment, levels=c("siLuc", "siCSTF2", "siCSTF2T", "siCSTF2-siCSTF2T", "siCPSF3", "siXRN2")),
            Unit = factor(Unit, levels=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene")))]
data.log = data.log[Treatment%in%c("siCSTF2-siCSTF2T", "siCPSF3"),]

tmpCol = setNames(tableau_color_pal(palette = "Superfishel Stone")(6),
                  c("siLuc", "siCSTF2-siCSTF2T", "siCPSF3", "siCSTF2", "siCSTF2T", "siXRN2"))
tmpCol = tmpCol[names(tmpCol)%in%c("siLuc", "siCSTF2-siCSTF2T", "siCPSF3")]

# max.Y = ceiling(max(abs(data[is.finite(log2(Ratio)), log2(Ratio)])))

# figure5C = ggplot(data, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
#    ggtitle("Knockdowns over siLuc expression (Chromatin level)") +
#    geom_hline(yintercept=0, lty=2, lwd=0.5) +
#    geom_boxplot(notch=TRUE, position=position_dodge(width=0.9), alpha=0.75, outlier.colour=NA, outlier.shape=20) +
#    scale_x_discrete("", limits = rev(levels(data[,Unit]))) +
#    scale_y_continuous("Expression ratio (log2)", breaks=seq(-8,8,2)) +
#    scale_fill_manual("Treatment", values=tmpCol) +
#    theme_bw() + theme(legend.position=c(0.15, 0.9), legend.box.margin=margin(0,0,0,0,"cm"),
#                       axis.text=element_text(size=12), axis.title=element_text(size=14)) +
#                       # legend.text=element_text(size=6), legend.title=element_text(size=6), legend.key.size=unit(0.4, "cm")) +
#    coord_flip(ylim=c(-4,4))
# # + theme(legend.title=element_blank(), legend.position=c(0.9, 0.9), legend.box.margin=margin(0,0,0,0,"cm")) + guides(col=guide_legend(ncol=2, byrow=TRUE), fill=guide_legend(ncol=2, byrow=TRUE))
# save_and_plot(x=figure5C, bname="Figures_and_Tables/Figure5C", width=8, height=6.5, plot=FALSE)

pl = profiles[profiles[grepl("Untreated", Treatment),], Untreated := i.ChrRNA]
pl[, Treatment := gsub("\\.", "-", Treatment)]
pl = melt.data.table(pl, id.vars=c("Unit", "Index", "Treatment"))
pl[, Min := .SD[value>0, min(value)], by=c("Unit", "Treatment", "variable")]
pl[value==0, value := Min]
pl = dcast.data.table(pl, Unit+Index+Treatment~variable, value.var="value")
pl[, Ratio := ChrRNA/Untreated]
pl[, log2Ratio := log2(Ratio)]
# tab = pl[, .(log2Value = -3, N=.N/2), by=c("Unit")]
pl[, Treatment := gsub("siLuc-CSTF2", "siLuc", Treatment)]
pl[, `:=`(Treatment = factor(Treatment, levels=c("siLuc", "siCSTF2", "siCSTF2T", "siCSTF2-siCSTF2T", "siCPSF3", "siXRN2")),
            Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]
setkeyv(pl, c("Unit", "Index"))
tab = pl[pl[grepl("siLuc", Treatment),],][, .(group1 = unique(i.Treatment),
                                              group2 = unique(Treatment),
                                              pval = wilcox.test(i.log2Ratio, log2Ratio, paired = TRUE)$p.value), by=c("Unit", "Treatment")]
tab[, p.signif := cut(pval,
                      breaks = c(-Inf, 0.0001, 0.001, 0.01, 0.05, Inf),
                      labels = c("****", "***", "**", "*", "ns"))]
figure5C_box = ggboxplot(pl[Treatment%in%c("siLuc", "siCSTF2-siCSTF2T", "siCPSF3"),], x="Unit", y="log2Ratio", fill="Treatment",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Treated over untreated expressin (log2)") +
   scale_fill_manual("Treatment", values=tmpCol) +
   stat_pvalue_manual(tab[group2%in%c("siCSTF2-siCSTF2T", "siCPSF3"),],
                      x = "Unit", y.position = 9, label = "p.signif", position = position_dodge(0.8)) +
   theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Chromatin-fraction expression; treatment:"))
save_and_plot(x=figure5C_box, bname="Figures_and_Tables/Figure5C_box", width=8, height=6.5, plot=FALSE) 

figure5C_box_bis = ggplot(data, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
   # ggtitle("Knockdowns over siLuc expression (Chromatin level)") +
   # geom_hline(yintercept=0, lty=2, lwd=0.75) +
   geom_boxplot(notch=TRUE,position=position_dodge(width=0.9), outlier.shape=20, alpha=0.75) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)") +
   scale_fill_manual("Treatment", values=tmpCol) +
   theme_cleveland() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14),
                             axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Knockdowns over siLuc expression (Chromatin level); treatment:"))
                      # legend.text=element_text(size=6), legend.title=element_text(size=6), legend.key.size=unit(0.4, "cm")) +
# + theme(legend.title=element_blank(), legend.position=c(0.9, 0.9), legend.box.margin=margin(0,0,0,0,"cm")) + guides(col=guide_legend(ncol=2, byrow=TRUE), fill=guide_legend(ncol=2, byrow=TRUE))
save_and_plot(x=figure5C_box_bis, bname="Figures_and_Tables/Figure5C_box_bis", width=8, height=6.5, plot=FALSE) 

pl.log = profiles.log[profiles.log[grepl("Untreated", Treatment),], Untreated := i.ChrRNA]
pl.log[, Treatment := gsub("\\.", "-", Treatment)]
pl.log = melt.data.table(pl.log, id.vars=c("Unit", "Index", "Treatment"))
pl.log[, Min := .SD[value>0, min(value)], by=c("Unit", "Treatment", "variable")]
pl.log[value==0, value := Min]
pl.log = dcast.data.table(pl.log, Unit+Index+Treatment~variable, value.var="value")
pl.log[, Ratio := ChrRNA/Untreated]
pl.log[, log2Ratio := log2(Ratio)]
# tab = pl.log[, .(log2Value = -3, N=.N/2), by=c("Unit")]
pl.log[, Treatment := gsub("siLuc-CSTF2", "siLuc", Treatment)]
pl.log[, `:=`(Treatment = factor(Treatment, levels=c("siLuc", "siCSTF2", "siCSTF2T", "siCSTF2-siCSTF2T", "siCPSF3", "siXRN2")),
            Unit = factor(Unit, levels=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene")))]
setkeyv(pl.log, c("Unit", "Index"))
tab = pl.log[pl.log[grepl("siLuc", Treatment),],][!is.na(Treatment), .(group1 = unique(i.Treatment),
                                              group2 = unique(Treatment),
                                              pval = wilcox.test(i.log2Ratio, log2Ratio, paired = TRUE)$p.value), by=c("Unit", "Treatment")]
tab[, p.signif := cut(pval,
                      breaks = c(-Inf, 0.0001, 0.001, 0.01, 0.05, Inf),
                      labels = c("****", "***", "**", "*", "ns"))]
figure5C_box.log = ggboxplot(pl.log[Treatment%in%c("siLuc", "siCSTF2-siCSTF2T", "siCPSF3"),], x="Unit", y="log2Ratio", fill="Treatment",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Treated over untreated expressin (log2)") +
   scale_fill_manual("Treatment", values=tmpCol) +
   stat_pvalue_manual(tab[group2%in%c("siCSTF2-siCSTF2T", "siCPSF3"),],
                      x = "Unit", y.position = 9, label = "p.signif", position = position_dodge(0.8)) +
   theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Chromatin-fraction expression; treatment:"))

figure5C_box_bis.log = ggplot(data.log, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
   # ggtitle("Knockdowns over siLuc expression (Chromatin level)") +
   # geom_hline(yintercept=0, lty=2, lwd=0.75) +
   geom_boxplot(notch=TRUE,position=position_dodge(width=0.9), outlier.shape=20, alpha=0.75) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)") +
   scale_fill_manual("Treatment", values=tmpCol) +
   theme_cleveland() + theme(axis.text=element_text(size=12), axis.title=element_text(size=14),
                             axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Knockdowns over siLuc expression (Chromatin level); treatment:"))

# df = data.frame(dcast.data.table(counts[grepl("T4P", Sample),], gene_id~Sample, value.var="count"), row.names="gene_id")
# 
# sizeFactors = DESeq2::estimateSizeFactorsForMatrix(as.matrix(df[rowSums(df>=1)>1,]))
# 
# df.norm = t(t(df)/sizeFactors)
# 
# bedfiles.all = list(UpstreamOfGene=bedfiles.px[[1]],
#                     Protein_coding=unique(unlist(bedfiles.px[2:3])),
#                     DownstreamOfGene=bedfiles.px[[4]],
#                     Long_non_coding=bedfiles.nc[[1]],
#                     `> 10Kb`=bedfiles.nc[[2]])
# 
# profiles = lapply(bedfiles.all,
#                   function(x)
#                      melt.data.table(data.table(Index = 1:length(x),
#                                                 data.table(df.norm[match(x$gene_id, rownames(df.norm)),])),
#                                      id.vars="Index", variable.name="Sample", value.name="Score"))
# profiles = rbindlist(profiles, idcol="Unit")
# profiles[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
# profiles = dcast.data.table(profiles[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
#                             Unit+Index+Treatment~Fraction, value.var="Score")
# 
# setkeyv(profiles, c("Unit", "Index"))
# 
# data = profiles[, Treatment := gsub("\\.", "-", Treatment)]
# data = dcast.data.table(data, Unit+Index~Treatment, value.var="mNETseq")
# data[, Ratio := `siCPSF73-T4P`/`siLuc-T4P`]
# 
# data[, Unit := factor(Unit, levels=c("UpstreamOfGene", "Protein_coding", "DownstreamOfGene", "Long_non_coding", "> 10Kb"))]
# 
# figure6G = ggplot(data, aes(x=Unit, y=log2(Ratio))) +
#    ggtitle("CPSF73 (CPSF3) knockdowns over siLuc expression (T4P)") +
#    geom_hline(yintercept=0, lty=2, lwd=0.5) +
#    geom_violin(position="dodge", col="black", alpha=0.5) +
#    geom_boxplot(notch=TRUE, width=0.1, position=position_dodge(width=0.9), outlier.colour=NA, fill="white") +
#    scale_x_discrete("") +
#    scale_y_continuous("Expression ratio (log2)") +
#    scale_colour_manual("Treatment", values=tableau_color_pal(palette = "Tableau 10")(7)[3:7]) +
#    scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Tableau 10")(7)[3:7]) +
#    theme_bw() + guides(fill=FALSE)
# save_and_plot(x=figure6G, bname="Figures_and_Tables/Figure6G", width=8, height=5, plot=FALSE)
```

![](./Figures_and_Tables/Figure5C_box.png)

![](./Figures_and_Tables/Figure5C_box_bis.png)

### XRN2

```{r count_XRN2}
bamTable = newTable[cell%in%"HCT116" & treatment%like%"XRN2",]
files.bam = list.files("RNAseq_BAM", pattern="*.bam$", full.names=TRUE)
files.bam = sapply(bamTable[, name], function(x) files.bam[files.bam%like%x])
names(files.bam) = bamTable[, name]

counts = lapply(files.bam,
                function(x){
                   print(x); flush(stdout())
                   if(as.character(x)%like%"_SE"){
                      param = ScanBamParam(flag=scanBamFlag(isProperPair=FALSE,
                                                            isDuplicate=FALSE,
                                                            isSecondaryAlignment=FALSE),
                                           mapqFilter=255)             
                      reads = readGAlignments(file=x, param=param)
                   }else{
                      param = ScanBamParam(flag=scanBamFlag(isProperPair=TRUE,
                                                            isDuplicate=FALSE,
                                                            isSecondaryAlignment=FALSE),
                                           mapqFilter=255)             
                      reads = readGAlignmentPairs(file=x, strandMode=1, param=param)
                   }
                   reads = invertStrand(keepSeqlevels(reads, chromosomes, pruning.mode="coarse"))
                   countOverlapsUnion(new.anno.genes, reads)
                })
counts = lapply(counts, function(x) data.table(Index = seq_along(x), gene_id = new.anno.genes$gene_id, count = x))

counts = rbindlist(counts, idcol="Sample")

saveRDS(counts, file="RDS/RNAseq_all_XRN2_Nuclear.rds")
```

```{r plot_XRN2}
counts = readRDS("RDS/RNAseq_all_XRN2_Nuclear.rds")

counts[, Subset := new.anno.genes[match(gene_id, new.anno.genes$gene_id),]$Subset]

df = data.frame(dcast.data.table(counts, gene_id~Sample, value.var="count"), row.names="gene_id")

sizeFactors = DESeq2::estimateSizeFactorsForMatrix(as.matrix(df[rowSums(df>=1)>1,]))

df.norm = t(t(df)/sizeFactors)

bedfiles.all = list(UpstreamOfGene=bedfiles.px[[1]],
                    Protein_coding=unique(unlist(bedfiles.px[2:3])),
                    DownstreamOfGene=bedfiles.px[[4]],
                    Long_non_coding=bedfiles.nc[[1]],
                    `Independent TU`=bedfiles.nc[[2]])

profiles = lapply(bedfiles.all,
                  function(x)
                     melt.data.table(data.table(Index = 1:length(x),
                                                data.table(df.norm[match(x$gene_id, rownames(df.norm)),])),
                                     id.vars="Index", variable.name="Sample", value.name="Score"))
profiles = rbindlist(profiles, idcol="Unit")
profiles[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
profiles = dcast.data.table(profiles[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
                            Unit+Index+Treatment~Fraction, value.var="Score")
setkeyv(profiles, c("Unit", "Index"))

bedfiles.log = list(UpstreamGene=bedfiles.lk[[1]],
                    LinkerOfGenes=bedfiles.lk[[2]],
                    DownstreamGene=bedfiles.lk[[3]])

profiles.log = lapply(bedfiles.log,
                  function(x)
                     melt.data.table(data.table(Index = 1:length(x),
                                                data.table(df.norm[match(x$gene_id, rownames(df.norm)),])),
                                     id.vars="Index", variable.name="Sample", value.name="Score"))
profiles.log = rbindlist(profiles.log, idcol="Unit")
profiles.log[, c("Fraction", "Treatment") := tstrsplit(Sample, "_")[c(2,3)]]
profiles.log = dcast.data.table(profiles.log[, list(Score = mean(Score)), by=c("Unit", "Index", "Treatment", "Fraction")],
                            Unit+Index+Treatment~Fraction, value.var="Score")
setkeyv(profiles.log, c("Unit", "Index"))

data = profiles[profiles[grepl("Unmodified", Treatment),], `:=`(Untreated = i.Nuclear, Ratio = Nuclear/i.Nuclear)]
data[!is.finite(Ratio) & is.finite(Nuclear) & Nuclear>0, Ratio := Nuclear]

data = data[!grepl("Unmodified", Treatment)][, Treatment := gsub("\\.", "-", Treatment)]
data[, Treatment := gsub("XRN2-AID-", "", Treatment)]
data[, `:=`(Treatment = factor(Treatment, levels=c("mAuxin", "pAuxin")),
            Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]

data = melt.data.table(dcast.data.table(data, Unit+Index~Treatment, value.var="Ratio", fill=NA), id.vars=c("Unit", "Index"), measure.vars= c("mAuxin", "pAuxin"), variable.name="Treatment", value.name="Ratio")

data.log = profiles.log[profiles.log[grepl("Unmodified", Treatment),], `:=`(Untreated = i.Nuclear, Ratio = Nuclear/i.Nuclear)]
data.log[!is.finite(Ratio) & is.finite(Nuclear) & Nuclear>0, Ratio := Nuclear]

data.log = data.log[!grepl("Unmodified", Treatment)][, Treatment := gsub("\\.", "-", Treatment)]
data.log[, Treatment := gsub("XRN2-AID-", "", Treatment)]
data.log[, `:=`(Treatment = factor(Treatment, levels=c("mAuxin", "pAuxin")),
            Unit = factor(Unit, levels=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene")))]

data.log = melt.data.table(dcast.data.table(data.log, Unit+Index~Treatment, value.var="Ratio", fill=NA), id.vars=c("Unit", "Index"), measure.vars= c("mAuxin", "pAuxin"), variable.name="Treatment", value.name="Ratio")

# figure5D = ggplot(data, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
#    ggtitle("XRN2 over Unmodified expression (Nuclear level)") +
#    geom_hline(yintercept=0, lty=2, lwd=0.5) +
#    geom_boxplot(notch=TRUE, position=position_dodge(width=0.9), alpha=0.75, outlier.colour=NA) +
#    scale_x_discrete("", limits = rev(levels(data[,Unit]))) +
#    scale_y_continuous("Expression ratio (log2)", breaks=seq(-8,8,2)) +
#    # scale_colour_manual("Treatment", values=tableau_color_pal(palette = "Tableau 10")(7)[3:7]) +
#    scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Classic Blue-Red 6")(2)) +
#    theme_bw() + theme(legend.position=c(0.1, 0.9), legend.box.margin=margin(0,0,0,0,"cm"),
#                       axis.text=element_text(size=12), axis.title=element_text(size=14)) +
#    coord_flip(ylim=c(-3,3)) # + theme(legend.title=element_blank(), legend.position=c(0.9, 0.9), legend.box.margin=margin(0,0,0,0,"cm")) + guides(col=guide_legend(ncol=2, byrow=TRUE), fill=guide_legend(ncol=2, byrow=TRUE))
figure5D = ggplot(data[order(Unit, Index)], aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
   ggtitle("XRN2 over Unmodified expression (Nuclear level)") +
   geom_hline(yintercept=0, lty=2, lwd=0.5) +
   geom_violin(position="dodge", alpha=0.5, trim=TRUE) +
   geom_boxplot(notch=TRUE, width=0.1, position=position_dodge(width=0.9), aes(col=Treatment), fill="white", outlier.shape=20) +
   scale_x_discrete("") +
   scale_y_continuous("Expression ratio (log2)") +
   scale_colour_manual(values=c("grey30", "grey30")) +
   scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Classic Blue-Red 6")(2)) +
   theme_bw() + theme(legend.position=c(0.1, 0.1), legend.box.margin=margin(0,0,0,0,"cm"),
                      axis.text=element_text(size=12), axis.title=element_text(size=14)) +
   stat_compare_means(aes(group = Treatment), label = "p.signif")
save_and_plot(x=figure5D, bname="Figures_and_Tables/Figure5D", width=8, height=6.5, plot=FALSE)

pl = profiles[profiles[grepl("Unmodified", Treatment),], Unmodified := i.Nuclear]
pl[, Treatment := gsub("\\.", "-", Treatment)]
pl = melt.data.table(pl, id.vars=c("Unit", "Index", "Treatment"))
pl[, Min := .SD[value>0, min(value)], by=c("Unit", "Treatment", "variable")]
pl[value==0, value := Min]
pl = dcast.data.table(pl, Unit+Index+Treatment~variable, value.var="value")
pl[, Ratio := Nuclear/Unmodified]
pl[, log2Ratio := log2(Ratio)]
# tab = pl[, .(log2Value = -3, N=.N/2), by=c("Unit")]
pl[, `:=`(Treatment = factor(Treatment, levels=c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin")),
          Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]
setkeyv(pl, c("Unit", "Index"))

figure5D_box = ggboxplot(pl[Treatment%in%c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin"),], x="Unit", y="log2Ratio", fill="Treatment",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Treated over untreated expressin (log2)") +
   scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Classic Blue-Red 6")(2),
                     labels = setNames(c("Uninduced", "XRN2 knockdown"), c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin"))) +
                     # labels = setNames(c("-Auxin", "+Auxin"), c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin"))) +
   stat_compare_means(aes(group = Treatment), paired=TRUE, label = "p.signif", label.y=9) +
   theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Nuclear-fraction expression; treatment:"))
save_and_plot(x=figure5D_box, bname="Figures_and_Tables/Figure5D_box", width=8, height=6.5, plot=FALSE) 


pl.log = profiles.log[profiles.log[grepl("Unmodified", Treatment),], Unmodified := i.Nuclear]
pl.log[, Treatment := gsub("\\.", "-", Treatment)]
pl.log = melt.data.table(pl.log, id.vars=c("Unit", "Index", "Treatment"))
pl.log[, Min := .SD[value>0, min(value)], by=c("Unit", "Treatment", "variable")]
pl.log[value==0, value := Min]
pl.log = dcast.data.table(pl.log, Unit+Index+Treatment~variable, value.var="value")
pl.log[, Ratio := Nuclear/Unmodified]
pl.log[, log2Ratio := log2(Ratio)]

pl.log[, `:=`(Treatment = factor(Treatment, levels=c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin")),
          Unit = factor(Unit, levels=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene")))]
setkeyv(pl.log, c("Unit", "Index"))

figure5D_box.log = ggboxplot(pl.log[Treatment%in%c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin"),], x="Unit", y="log2Ratio", fill="Treatment",
                         notch=TRUE, alpha=0.75, outlier.shape=20,
                         order=c("UpstreamGene", "LinkerOfGenes", "DownstreamGene"),
                         ggtheme = theme_cleveland()) +
   scale_x_discrete("") +
   scale_y_continuous("Treated over untreated expressin (log2)") +
   scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Classic Blue-Red 6")(2),
                     labels = setNames(c("Uninduced", "XRN2 knockdown"), c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin"))) +
                     # labels = setNames(c("-Auxin", "+Auxin"), c("XRN2-AID-mAuxin", "XRN2-AID-pAuxin"))) +
   stat_compare_means(aes(group = Treatment), paired=TRUE, label = "p.signif", label.y=6.5) +
   theme(axis.text=element_text(size=12), axis.title=element_text(size=14), axis.title.y=element_text(angle=90), legend.position="top") +
   guides(fill = guide_legend(title="Nuclear-fraction expression; treatment:"))

data = profiles[!grepl("Unmodified", Treatment),]
data = data[data[grepl("mAuxin", Treatment),], `:=`(Untreated = i.Nuclear, Ratio = Nuclear/i.Nuclear)]
data[!is.finite(Ratio) & is.finite(Nuclear) & Nuclear>0, Ratio := Nuclear]

data = data[!grepl("mAuxin", Treatment)][, Treatment := gsub("\\.", "-", Treatment)]
data[, `:=`(Unit = factor(Unit, levels=c("Protein_coding", "Long_non_coding", "UpstreamOfGene", "DownstreamOfGene", "Independent TU")))]

figure5D_bis = ggplot(data, aes(x=Unit, y=log2(Ratio), fill=Treatment)) +
   ggtitle("XRN2 auxin-treated over untreated expression (Nuclear level)") +
   geom_hline(yintercept=0, lty=2, lwd=0.5) +
   geom_boxplot(notch=TRUE, position=position_dodge(width=0.9), alpha=0.75, outlier.colour=NA, outlier.shape=20) +
   scale_x_discrete("", limits = rev(levels(data[,Unit]))) +
   scale_y_continuous("Expression ratio (log2)", breaks=seq(-8,8,2)) +
   # scale_colour_manual("Treatment", values=tableau_color_pal(palette = "Tableau 10")(7)[3:7]) +
   scale_fill_manual("Treatment", values=tableau_color_pal(palette = "Classic Blue-Red 6")(2)[2], guide=FALSE) +
   theme_bw() + theme(legend.position=c(0.1, 0.9), legend.box.margin=margin(0,0,0,0,"cm"),
                      axis.text=element_text(size=12), axis.title=element_text(size=14)) +
   coord_flip(ylim=c(-3,3)) # + theme(legend.title=element_blank(), legend.position=c(0.9, 0.9), legend.box.margin=margin(0,0,0,0,"cm")) + guides(col=guide_legend(ncol=2, byrow=TRUE), fill=guide_legend(ncol=2, byrow=TRUE))
save_and_plot(x=figure5D_bis, bname="Figures_and_Tables/figure5D_bis", width=8, height=6.5, plot=FALSE)
```

![](./Figures_and_Tables/Figure5D.png)

![](./Figures_and_Tables/Figure5D_box.png)

```{r figure5}
# figure5 = ggarrange(figure5A, figure5B_bis, figure5C_bis, figure5D, ncol=2, nrow=2, labels=c("A", "B", "C", "D"))
# save_and_plot(x=figure5, bname="Figures_and_Tables/Figure5", width=18, height=12, plot=FALSE)
# ggsave(figure5, filename="Figures_and_Tables/Figure5.eps", width=18, height=12, device=cairo_ps)
# figure5_bis = ggarrange(figure5A, figure5B_bis, figure5C_bis, figure5D, ncol=2, nrow=2, labels=c("A", "B", "C", "D"))
# save_and_plot(x=figure5_bis, bname="Figures_and_Tables/Figure5_bis", width=15, height=12, plot=FALSE)

figure5_box = ggarrange(figure5A_box, figure5B_box_bis, figure5C_box_bis, figure5D_box,
                        ncol=2, nrow=2, labels=c("A", "B", "C", "D"))
save_and_plot(x=figure5_box, bname="Figures_and_Tables/Figure5_box", width=18, height=12, plot=FALSE)

figure5_box.log = ggarrange(figure5A_box.log, figure5B_box_bis.log, figure5C_box_bis.log, figure5D_box.log,
                        ncol=2, nrow=2, labels=c("A", "B", "C", "D"))
save_and_plot(x=figure5_box.log, bname="Figures_and_Tables/Figure5_box.log", width=18, height=12, plot=FALSE)
```
